<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Interactively exploring high-dimensional data and models in R - 15&nbsp; Trees and forests</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./16-svm.html" rel="next">
<link href="./14-lda.html" rel="prev">
<link href="./cover.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./15-forests.html"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Trees and forests</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Interactively exploring high-dimensional data and models in R</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/dicook/mulgar_book" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./1-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Picturing high dimensions</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./2-notation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Notation conventions and R objects</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./3-intro-dimred.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Dimension reduction overview</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./4-pca.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Principal component analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./5-nldr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Non-linear dimension reduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./6-intro-clust.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Introduction to clustering</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./7-spin-and-brush.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Spin-and-brush approach</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./8-hierarchical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Hierarchical clustering</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./9-kmeans.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title"><span class="math inline">\(k\)</span>-means clustering</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10-model-based.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Model-based clustering</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11-som.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Self-organizing maps</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12-summary-clust.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Summarising and comparing clustering results</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./13-intro-class.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Introduction to supervised classification</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./14-lda.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Linear discriminant analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./15-forests.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Trees and forests</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./16-svm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Support vector machines</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./17-nn.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Neural networks and deep learning</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./18-summary-class.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Exploring misclassifications</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A1-toolbox.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Toolbox</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A2-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A3-book-code-and-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Links to Book Code and Additional Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A4-glossary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">D</span>&nbsp; <span class="chapter-title">Glossary</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#sec-trees" id="toc-sec-trees" class="nav-link active" data-scroll-target="#sec-trees"><span class="header-section-number">15.1</span> Trees</a></li>
  <li>
<a href="#random-forests" id="toc-random-forests" class="nav-link" data-scroll-target="#random-forests"><span class="header-section-number">15.2</span> Random forests</a>
  <ul class="collapse">
<li><a href="#sec-votes" id="toc-sec-votes" class="nav-link" data-scroll-target="#sec-votes"><span class="header-section-number">15.2.1</span> Examining the votes matrix</a></li>
  <li><a href="#sec-forest-var-imp" id="toc-sec-forest-var-imp" class="nav-link" data-scroll-target="#sec-forest-var-imp"><span class="header-section-number">15.2.2</span> Using variable importance</a></li>
  </ul>
</li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises">Exercises</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/dicook/mulgar_book/edit/master/15-forests.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/dicook/mulgar_book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-trees-forests" class="quarto-section-identifier"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Trees and forests</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><section id="sec-trees" class="level2" data-number="15.1"><h2 data-number="15.1" class="anchored" data-anchor-id="sec-trees">
<span class="header-section-number">15.1</span> Trees</h2>
<p></p>
<p>The tree algorithm <span class="citation" data-cites="BFOS84">(<a href="references.html#ref-BFOS84" role="doc-biblioref">Breiman et al., 1984</a>)</span> is a simple and versatile algorithmic method for supervised classification. The basic tree algorithm generates a classification rule by sequentially splitting the data into two buckets. Splits are made between sorted data values of individual variables, with the goal of obtaining pure classes on each side of the split. The inputs for a simple tree classifier commonly include (1) an impurity measure, an indication of the relative diversity among the cases in the terminal nodes; (2) a parameter that sets the minimum number of cases in a node, or the minimum number of observations in a terminal node of the tree; and (3) a complexity measure that controls the growth of a tree, balancing the use of a simple generalizable tree against a more accurate tree tailored to the sample. When applying tree methods, exploring the effects of the input parameters on the tree is instructive; for example, it helps us to assess the stability of the tree model.</p>
<p>Although algorithmic models do not depend on distributional assumptions, that does not mean that every algorithm is suitable for all data. For example, the tree model works best when all variables are independent within each class, because it does not take such dependencies into account. Visualization can help us to determine whether a particular model should be applied. In classification problems, it is useful to explore the cluster structure, comparing the clusters with the classes and looking for evidence of correlation within each class. The plots in <a href="14-lda.html#fig-lda-assumptions1" class="quarto-xref">Figure&nbsp;<span>14.2</span></a> and <a href="14-lda.html#fig-penguins-lda-ellipses-html" class="quarto-xref">Figure&nbsp;<span>14.4</span></a> shows a strong correlation between the variables within each species, which indicates that the tree model may not give good results for the penguins data. We’ll show how this is the case with two variables initially, and then extend to the four variables.</p>
<div class="cell">
<details class="code-fold"><summary>Draw tree and model boundaries</summary><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dicook.github.io/mulgar/">mulgar</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/bethatkinson/rpart">rpart</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.milbo.org/rpart-plot/index.html">rpart.plot</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://colorspace.R-Forge.R-project.org/">colorspace</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://had.co.nz/classifly">classifly</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/andrie/ggdendro">ggdendro</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jrnold/ggthemes">ggthemes</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/penguins_sub.rda"</span><span class="op">)</span></span>
<span><span class="va">p_bl_bd_tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rpart/man/rpart.html">rpart</a></span><span class="op">(</span><span class="va">species</span><span class="op">~</span><span class="va">bl</span><span class="op">+</span><span class="va">bd</span>, data<span class="op">=</span><span class="va">penguins_sub</span><span class="op">)</span></span>
<span><span class="co">#f1 &lt;- rpart.plot(p_bl_bd_tree, box.palette="Grays")</span></span>
<span><span class="va">d</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ggdendro/man/dendro_data.html">dendro_data</a></span><span class="op">(</span><span class="va">p_bl_bd_tree</span><span class="op">)</span></span>
<span><span class="va">f1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">d</span><span class="op">$</span><span class="va">segments</span>, </span>
<span>               <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, </span>
<span>                   xend <span class="op">=</span> <span class="va">xend</span>, yend <span class="op">=</span> <span class="va">yend</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">d</span><span class="op">$</span><span class="va">labels</span>, </span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, </span>
<span>                label <span class="op">=</span> <span class="va">label</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2.7</span>, </span>
<span>            vjust <span class="op">=</span> <span class="fl">1.2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">d</span><span class="op">$</span><span class="va">leaf_labels</span>,</span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, y <span class="op">=</span> <span class="va">y</span>, </span>
<span>                label <span class="op">=</span> <span class="va">label</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2.5</span>, </span>
<span>            vjust <span class="op">=</span> <span class="fl">2</span>, hjust<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0.5</span>,<span class="fl">0</span>,<span class="fl">0.5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/expand_limits.html">expand_limits</a></span><span class="op">(</span>x<span class="op">=</span><span class="fl">0.9</span>, y<span class="op">=</span><span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggdendro/man/theme_dendro.html">theme_dendro</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_bl_bd_tree_boundaries</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/classifly/man/explore.html">explore</a></span><span class="op">(</span><span class="va">p_bl_bd_tree</span>, <span class="va">penguins_sub</span><span class="op">)</span></span>
<span><span class="va">f2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">p_bl_bd_tree_boundaries</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">bl</span>, y<span class="op">=</span><span class="va">bd</span>, colour<span class="op">=</span><span class="va">species</span>, shape<span class="op">=</span><span class="va">.TYPE</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_divergingx.html">scale_color_discrete_divergingx</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Zissou 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_shape_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">46</span>, <span class="fl">16</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span>, legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">f1</span> <span class="op">+</span> <span class="va">f2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<div id="fig-p-bl-bd-tree" class="quarto-figure quarto-figure-center quarto-float anchored" alt="Tree diagram with top split bl<0.3004, leading to Adelie branch, second split at bd >= -0.4138, leading to Gentoo branch, and final split at bl< 0.1476, leading to Adelie and Chinstrap branches. The scatterplot at right shows bd vs bl, with three predictive region partitions, and the data is overplotted. The elliptical spreads of data points crosses the rectangular partitions in places.">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-p-bl-bd-tree-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="15-forests_files/figure-html/fig-p-bl-bd-tree-1.png" class="img-fluid figure-img" style="width:100.0%" alt="Tree diagram with top split bl<0.3004, leading to Adelie branch, second split at bd >= -0.4138, leading to Gentoo branch, and final split at bl< 0.1476, leading to Adelie and Chinstrap branches. The scatterplot at right shows bd vs bl, with three predictive region partitions, and the data is overplotted. The elliptical spreads of data points crosses the rectangular partitions in places.">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-p-bl-bd-tree-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;15.1: The association between variables in the penguins data causes problems for fitting a tree model. Although the model, computed using only bl and bd, is simple (left), the fit is poor (right) because it doesn’t adequately utilise combinations of variables.
</figcaption></figure>
</div>
</div>
</div>
<p>The plots in <a href="#fig-p-bl-bd-tree" class="quarto-xref">Figure&nbsp;<span>15.1</span></a> show the inadequacies of the tree fit. The background color indicates the class predictions, and thus boundaries produced by the tree fit. They can be seen to be boxy, and missing the elliptical nature of the penguin clusters. This produces errors in the classification of observations which are indefensible. One could always force the tree to fit the data more closely by adjusting the parameters, but the main problem persists: that one is trying to fit elliptical shapes using boxes.</p>
<div class="info">
<p>There are less strict assumptions for a non-parametric model but it is still important to understand the model fit relative to the data.</p>
</div>
<p>The boundaries for the tree model on all four variables of the penguins data can be viewed similarly, by predicting a set of points randomly generated in the 4D domain of observed values. <a href="#fig-penguins-lda-tree-html" class="quarto-xref">Figure&nbsp;<span>15.2</span></a> shows the prediction regions for LDA and a default tree in a slice tour <span class="citation" data-cites="slicetour">(<a href="references.html#ref-slicetour" role="doc-biblioref">Laa et al., 2020a</a>)</span>. The slice tour is used to help see into the middle of the 4D cube. It slices the cube through the centre of the data, where the boundaries of the regions should meet.</p>
<p>The prediction regions of the default fitted tree are shown in comparison to those from the LDA model. We don’t show the tree diagram here, but it makes only six splits of the tree model, which is delightfully simple. However, just like the model fitted to two variables, the result is not adequate for the penguins data. The tree model generates boxy boundaries, whereas the LDA model splits the 4D cube obliquely. The boxy regions don’t capture the differences between the elliptically-shaped clusters. Overlaying the observed data on this display would make this clearer, but the boundaries are easier to examine without them.</p>
<div class="cell">
<details class="code-fold"><summary>Code to make animated gifs of slice tour of boundaries</summary><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p_tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rpart/man/rpart.html">rpart</a></span><span class="op">(</span><span class="va">species</span><span class="op">~</span><span class="va">.</span>, data<span class="op">=</span><span class="va">penguins_sub</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/rpart.plot/man/rpart.plot.html">rpart.plot</a></span><span class="op">(</span><span class="va">p_tree</span>, box.palette<span class="op">=</span><span class="st">"Grays"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_tree_boundaries</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/classifly/man/explore.html">explore</a></span><span class="op">(</span><span class="va">p_tree</span>, <span class="va">penguins_sub</span><span class="op">)</span></span>
<span><span class="fu">animate_slice</span><span class="op">(</span><span class="va">p_tree_boundaries</span><span class="op">[</span><span class="va">p_tree_boundaries</span><span class="op">$</span><span class="va">.TYPE</span> <span class="op">==</span> <span class="st">"simulated"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, col<span class="op">=</span><span class="va">p_tree_boundaries</span><span class="op">[</span><span class="va">p_tree_boundaries</span><span class="op">$</span><span class="va">.TYPE</span> <span class="op">==</span> <span class="st">"simulated"</span>,<span class="fl">6</span><span class="op">]</span>, v_rel<span class="op">=</span><span class="fl">0.02</span>, axes<span class="op">=</span><span class="st">"bottomleft"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/penguins_tour_path.rda"</span><span class="op">)</span></span>
<span><span class="fu">render_gif</span><span class="op">(</span><span class="va">p_tree_boundaries</span><span class="op">[</span><span class="va">p_tree_boundaries</span><span class="op">$</span><span class="va">.TYPE</span> <span class="op">==</span> <span class="st">"simulated"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>           <span class="fu">planned_tour</span><span class="op">(</span><span class="va">pt1</span><span class="op">)</span>,</span>
<span>           <span class="fu">display_slice</span><span class="op">(</span>v_rel<span class="op">=</span><span class="fl">0.02</span>, </span>
<span>             col<span class="op">=</span><span class="va">p_tree_boundaries</span><span class="op">[</span><span class="va">p_tree_boundaries</span><span class="op">$</span><span class="va">.TYPE</span> <span class="op">==</span> <span class="st">"simulated"</span>,<span class="fl">6</span><span class="op">]</span>, </span>
<span>             axes<span class="op">=</span><span class="st">"bottomleft"</span><span class="op">)</span>,                     gif_file<span class="op">=</span><span class="st">"gifs/penguins_tree_boundaries.gif"</span>,</span>
<span>           frames<span class="op">=</span><span class="fl">500</span>,</span>
<span>           loop<span class="op">=</span><span class="cn">FALSE</span></span>
<span>           <span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p></p>
<div id="fig-penguins-lda-tree-html" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-penguins-lda-tree-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-penguins-lda-tree-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-lda-boundary" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-lda-boundary-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gifs/penguins_lda_boundaries.gif" class="img-fluid figure-img" data-ref-parent="fig-penguins-lda-tree-html" alt="FIX ME" width="300">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-lda-boundary-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) LDA model
</figcaption></figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-penguins-lda-tree-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-tree-boundary" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-tree-boundary-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gifs/penguins_tree_boundaries.gif" class="img-fluid figure-img" data-ref-parent="fig-penguins-lda-tree-html" alt="FIX ME" width="300">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-tree-boundary-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) Tree model
</figcaption></figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-penguins-lda-tree-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;15.2: Comparison of the boundaries produced by the LDA (a) and the tree (b) model, using a slice tour. The tree boundaries are more box-shaped than the LDA boundaries, which does not adequately capture the differences between the elliptically-shaped clusters of the penguins data.
</figcaption></figure>
</div>
<div class="insight">
<p>With the penguins data, a tree model may not be a good choice due to the strong correlation between variables. The best separation is in combinations of variables, not the single variable tree splits.</p>
</div>
</section><section id="random-forests" class="level2" data-number="15.2"><h2 data-number="15.2" class="anchored" data-anchor-id="random-forests">
<span class="header-section-number">15.2</span> Random forests</h2>
<p></p>
<p>A random forest <span class="citation" data-cites="Br01">(<a href="references.html#ref-Br01" role="doc-biblioref">Breiman, 2001</a>)</span> is a classifier that is built from multiple trees generated by randomly sampling the cases and the variables. The random sampling (with replacement) of cases has the fortunate effect of creating a training (“in-bag”) and a test (“out-of-bag”) sample for each tree computed. The class of each case in the out-of-bag sample for each tree is predicted, and the predictions for all trees are combined into a vote for the class identity.</p>
<p>A random forest is a computationally intensive method, a “black box” classifier, but it produces several diagnostics that make the outcome less mysterious. Some diagnostics that help us to assess the model are the votes, the measure of variable importance, and the proximity matrix.</p>
<section id="sec-votes" class="level3" data-number="15.2.1"><h3 data-number="15.2.1" class="anchored" data-anchor-id="sec-votes">
<span class="header-section-number">15.2.1</span> Examining the votes matrix</h3>
<p>Here we show how to use the <code>randomForest</code> <span class="citation" data-cites="randomForest2002">(<a href="references.html#ref-randomForest2002" role="doc-biblioref">Liaw &amp; Wiener, 2002</a>)</span> votes matrix for the penguins data to investigate confusion between classes, and observations which are problematic to classify. The votes matrix can be considered to be predictive probability distribution, where the values for each observation sum to 1. With only three classes the votes matrix is only a 2D object, and thus easy to examine. With four or more classes the votes matrix needs to be examined in a tour. </p>
<div class="cell">
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.stat.berkeley.edu/~breiman/RandomForests/">randomForest</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="va">penguins_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/randomForest/man/randomForest.html">randomForest</a></span><span class="op">(</span><span class="va">species</span><span class="op">~</span><span class="va">.</span>,</span>
<span>                             data<span class="op">=</span><span class="va">penguins_sub</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>                             importance<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">penguins_rf</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>
Call:
 randomForest(formula = species ~ ., data = penguins_sub[, 1:5],      importance = TRUE) 
               Type of random forest: classification
                     Number of trees: 500
No. of variables tried at each split: 2

        OOB estimate of  error rate: 2.4%
Confusion matrix:
          Adelie Chinstrap Gentoo class.error
Adelie       143         3      0 0.020547945
Chinstrap      4        64      0 0.058823529
Gentoo         0         1    118 0.008403361</code></pre>
</div>
</div>
<p>To examine the votes matrix, we extract the <code>votes</code> element from the random forest model object. The first five rows are:</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">penguins_rf</span><span class="op">$</span><span class="va">votes</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>  Adelie Chinstrap Gentoo
1 0.9949  0.005128      0
2 0.9836  0.016393      0
3 0.9786  0.021390      0
4 1.0000  0.000000      0
5 1.0000  0.000000      0</code></pre>
</div>
</div>
<p>This has three columns corresponding to the three species, but because each row is a set of proportions it is only a 2D object. To reduce the dimension from 3D to the 2D we use a Helmert matrix <span class="citation" data-cites="helmert">(<a href="references.html#ref-helmert" role="doc-biblioref">Lancaster, 1965</a>)</span>. A Helmert matrix has a first row of all 1’s. The remaining components of the matrix are 1’s in the lower triangle, and 0’s in the upper triangle and the diagonal elements are the negative row sum. The rows are usually normalised to have length 1. They are used to create contrasts to test combinations of factor levels for post-testing after Analysis of Variance (ANOVA). For compositional data, like the votes matrix, when the first row is removed a Helmert matrix can be used to reduce the dimension appropriately. For three classes, this will generate the common 2D ternary diagram, but for higher dimensions it will reduce to a <span class="math inline">\((g-1)\)</span>-dimensional simplex. For the penguins data, the Helmert matrix for 3D is </p>
<div class="cell">
<details class="code-fold"><summary>Code to compute Helmert matrix</summary><div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">geozoo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/geozoo/man/f_helmert.html">f_helmert</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>          [,1]    [,2]    [,3]
helmert 0.5774  0.5774  0.5774
x       0.7071 -0.7071  0.0000
x       0.4082  0.4082 -0.8165</code></pre>
</div>
</div>
<p>We drop the first row, transpose it, and use matrix multiplication with the votes matrix to get the ternary diagram.</p>
<div class="cell">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Project 4D into 3D</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://schloerke.github.io/geozoo/">geozoo</a></span><span class="op">)</span></span>
<span><span class="va">proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">geozoo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/geozoo/man/f_helmert.html">f_helmert</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="va">p_rf_v_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">penguins_rf</span><span class="op">$</span><span class="va">votes</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">proj</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">p_rf_v_p</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x1"</span>, <span class="st">"x2"</span><span class="op">)</span></span>
<span><span class="va">p_rf_v_p</span> <span class="op">&lt;-</span> <span class="va">p_rf_v_p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>species <span class="op">=</span> <span class="va">penguins_sub</span><span class="op">$</span><span class="va">species</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Add simplex</span></span>
<span><span class="va">simp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/geozoo/man/simplex.html">simplex</a></span><span class="op">(</span>p<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">simp</span><span class="op">$</span><span class="va">points</span><span class="op">)</span>, <span class="va">simp</span><span class="op">$</span><span class="va">points</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">1</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x1"</span>, <span class="st">"x2"</span>, <span class="st">"x3"</span>, <span class="st">"x4"</span><span class="op">)</span></span>
<span><span class="va">sp</span><span class="op">$</span><span class="va">species</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">penguins_sub</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p_ternary</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">sp</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x1</span>, y<span class="op">=</span><span class="va">x2</span>, xend<span class="op">=</span><span class="va">x3</span>, yend<span class="op">=</span><span class="va">x4</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">sp</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x1</span>, y<span class="op">=</span><span class="va">x2</span>, label<span class="op">=</span><span class="va">species</span><span class="op">)</span>,</span>
<span>            nudge_x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.06</span>, <span class="fl">0.07</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>            nudge_y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="fl">0.05</span>, <span class="op">-</span><span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">p_rf_v_p</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x1</span>, y<span class="op">=</span><span class="va">x2</span>, </span>
<span>                                colour<span class="op">=</span><span class="va">species</span><span class="op">)</span>, </span>
<span>             size<span class="op">=</span><span class="fl">2</span>, alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_divergingx.html">scale_color_discrete_divergingx</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Zissou 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/theme_map.html">theme_map</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>aspect.ratio<span class="op">=</span><span class="fl">1</span>, legend.position<span class="op">=</span><span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<details class="code-fold"><summary>Code to generate animated gifs</summary><div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Look at the votes matrix, in its 3D space</span></span>
<span><span class="fu">animate_xy</span><span class="op">(</span><span class="va">penguins_rf</span><span class="op">$</span><span class="va">votes</span>, col<span class="op">=</span><span class="va">penguins_sub</span><span class="op">$</span><span class="va">species</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Save an animated gif</span></span>
<span><span class="fu">render_gif</span><span class="op">(</span><span class="va">penguins_rf</span><span class="op">$</span><span class="va">votes</span>,</span>
<span>           <span class="fu">grand_tour</span><span class="op">(</span><span class="op">)</span>,</span>
<span>           <span class="fu">display_xy</span><span class="op">(</span>v_rel<span class="op">=</span><span class="fl">0.02</span>, </span>
<span>             col<span class="op">=</span><span class="va">penguins_sub</span><span class="op">$</span><span class="va">species</span>, </span>
<span>             axes<span class="op">=</span><span class="st">"bottomleft"</span><span class="op">)</span>, </span>
<span>           gif_file<span class="op">=</span><span class="st">"gifs/penguins_rf_votes.gif"</span>,</span>
<span>           frames<span class="op">=</span><span class="fl">500</span>,</span>
<span>           loop<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="fig-penguins-votes-html" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-penguins-votes-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-penguins-votes-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-p-votes-tour" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-p-votes-tour-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gifs/penguins_rf_votes.gif" class="img-fluid figure-img" data-ref-parent="fig-penguins-votes-html" alt="FIX ME" width="300">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-p-votes-tour-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) 3D
</figcaption></figure>
</div>
</div>
<div class="cell quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-penguins-votes-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div class="cell-output-display quarto-layout-cell-subref" data-ref-parent="fig-penguins-votes-html">
<div id="fig-p-votes-ggplot-html" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-p-votes-ggplot-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="15-forests_files/figure-html/fig-p-votes-ggplot-html-1.png" class="img-fluid figure-img" data-ref-parent="fig-penguins-votes-html" width="384">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-p-votes-ggplot-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) 2D ternary diagram
</figcaption></figure>
</div>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-penguins-votes-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;15.3: Examining the votes matrix from a random forest fit to the penguins: (a) from a tour of the 3D, (b) projected into 2D, to make a ternary diagram. In 3D the points can be seen to lie along a 2D plane, which is due to the constraint that the values sum to 1. From the ternary diagram, the classification can be seen to be reasonably well distinguished because points mostly lie at the vertex. There are a few penguins that are confused with a different species, as seen from the few points spread between vertices.
</figcaption></figure>
</div>
<p>We can use the <code>geozoo</code> package to generate the surrounding simplex, which for 2D is a triangle.</p>
<p>The votes matrix reports the proportion of trees each observation is classified as each class. From the tour of the votes matrix, as in <a href="#fig-penguins-votes-html" class="quarto-xref">Figure&nbsp;<span>15.3</span></a>(a), it can be seen to be 2D in 3D space. This is due to the constraint that the three proportions for each observation sum to 1. Using a Helmert matrix, this data can be projected into the 2D space, or more generally the <span class="math inline">\((g-1)\)</span>-dimensional space where it resides, shown in <a href="#fig-penguins-votes-html" class="quarto-xref">Figure&nbsp;<span>15.3</span></a>(b). In 2D this is called a ternary diagram, and in higher dimensions the bounding shapes might be considered to be a simplex. The vertices of this shape correspond to <span class="math inline">\((1,0,0), (0,1,0), (0,0,1)\)</span> (and analogously for higher dimensions), which represent perfect confidence, that an observation is classified into that group all the time.</p>
<p>What we can see here is a concentration of points in the corners of the triangle indicates that most of the penguins are confidently classified into their correct class. Then there is more separation between the Gentoo and the others, than between Chinstrap and Adelie. That means that as a group Gentoo are more distinguishable. Only one of the Gentoo penguins has substantial confusion, mostly confused as a Chinstrap, but occasionally confused as an Adelie – if it was only ever confused as a Chinstrap it would fall on the edge between Gentoo and Chinstrap. There are quite a few Chinstrap and Adelie penguins confused as each other, with a couple of each more confidently predicted to be the other class. This can be seen because there are points of the wrong colour close to those vertices.</p>
<p>The votes matrix is useful for investigating the fit, but one should remember that there are some structural elements of the penguins data that don’t lend themselves to tree models. Although a forest has the capacity to generate non-linear boundaries by combining predictions from multiple trees, it is still based on the boxy boundaries of trees. This makes it less suitable for the penguins data with elliptical classes. You could use the techniques from the previous section to explore the boundaries produced by the forest, and you will find that the are more boxy than the LDA models. </p>
<div class="info">
<p>By visualising the votes matrix we can understand which observations are harder to classify, which of the classes are more easily confused with each other.</p>
</div>
<p>To examine a vote matrix for a problem with more classes, we will examine the 10 class <code>fake_trees</code> data example. The full data has 100 variables, and we have seen from <a href="7-spin-and-brush.html" class="quarto-xref"><span>Chapter 7</span></a> that reducing to 10 principal components allows the linear branching structure in the data to be seen. Given that the branches correspond to the classes, it will be interesting to see how well the random forest model performs. </p>
<div class="cell">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dicook.github.io/mulgar/">mulgar</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sa-lee/liminal/">liminal</a></span><span class="op">)</span></span>
<span><span class="va">ft_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span><span class="va">fake_trees</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span>, </span>
<span>                 scale<span class="op">=</span><span class="cn">TRUE</span>, retx<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ft_pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">ft_pca</span><span class="op">$</span><span class="va">x</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">ft_pc</span><span class="op">$</span><span class="va">branches</span> <span class="op">&lt;-</span> <span class="va">fake_trees</span><span class="op">$</span><span class="va">branches</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.stat.berkeley.edu/~breiman/RandomForests/">randomForest</a></span><span class="op">)</span></span>
<span><span class="va">ft_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/randomForest/man/randomForest.html">randomForest</a></span><span class="op">(</span><span class="va">branches</span><span class="op">~</span><span class="va">.</span>, data<span class="op">=</span><span class="va">ft_pc</span>, </span>
<span>                            importance<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ft_rf</span><span class="op">$</span><span class="va">votes</span>, <span class="fl">5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>    0    1    2     3     4    5     6 7    8    9
1 0.8 0.00 0.02 0.000 0.022 0.13 0.000 0 0.02 0.00
2 0.7 0.00 0.03 0.000 0.006 0.26 0.000 0 0.00 0.00
3 0.8 0.00 0.05 0.010 0.116 0.01 0.000 0 0.00 0.02
4 0.9 0.00 0.03 0.000 0.011 0.08 0.000 0 0.00 0.00
5 0.6 0.01 0.05 0.005 0.005 0.28 0.005 0 0.01 0.00</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ft_rf_votes</span> <span class="op">&lt;-</span> <span class="va">ft_rf</span><span class="op">$</span><span class="va">votes</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>branches <span class="op">=</span> <span class="va">fake_trees</span><span class="op">$</span><span class="va">branches</span><span class="op">)</span></span>
<span></span>
<span><span class="va">proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">geozoo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/geozoo/man/f_helmert.html">f_helmert</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="va">f_rf_v_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">ft_rf_votes</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">proj</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">f_rf_v_p</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x1"</span>, <span class="st">"x2"</span>, <span class="st">"x3"</span>, <span class="st">"x4"</span>, <span class="st">"x5"</span>, <span class="st">"x6"</span>, <span class="st">"x7"</span>, <span class="st">"x8"</span>, <span class="st">"x9"</span><span class="op">)</span></span>
<span><span class="va">f_rf_v_p</span> <span class="op">&lt;-</span> <span class="va">f_rf_v_p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>branches <span class="op">=</span> <span class="va">fake_trees</span><span class="op">$</span><span class="va">branches</span><span class="op">)</span></span>
<span></span>
<span><span class="va">simp</span> <span class="op">&lt;-</span> <span class="fu">geozoo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/geozoo/man/simplex.html">simplex</a></span><span class="op">(</span>p<span class="op">=</span><span class="fl">9</span><span class="op">)</span></span>
<span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">simp</span><span class="op">$</span><span class="va">points</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x1"</span>, <span class="st">"x2"</span>, <span class="st">"x3"</span>, <span class="st">"x4"</span>, <span class="st">"x5"</span>, <span class="st">"x6"</span>, <span class="st">"x7"</span>, <span class="st">"x8"</span>, <span class="st">"x9"</span><span class="op">)</span></span>
<span><span class="va">sp</span><span class="op">$</span><span class="va">branches</span> <span class="op">=</span> <span class="st">""</span></span>
<span><span class="va">f_rf_v_p_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html">bind_rows</a></span><span class="op">(</span><span class="va">sp</span>, <span class="va">f_rf_v_p</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>branches <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">branches</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"0"</span> , <span class="st">"1"</span>, <span class="st">"2"</span>, <span class="st">"3"</span>, <span class="st">"4"</span>, <span class="st">"5"</span>, <span class="st">"6"</span>, <span class="st">"7"</span>, <span class="st">"8"</span>, <span class="st">"9"</span>,</span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="st">""</span>, <span class="fl">3000</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<details class="code-fold"><summary>Code to make animated gifs</summary><div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">animate_xy</span><span class="op">(</span><span class="va">f_rf_v_p_s</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>, col <span class="op">=</span> <span class="va">f_rf_v_p_s</span><span class="op">$</span><span class="va">branches</span>, </span>
<span>           axes <span class="op">=</span> <span class="st">"off"</span>, half_range <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>           edges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">simp</span><span class="op">$</span><span class="va">edges</span><span class="op">)</span>,</span>
<span>           obs_labels <span class="op">=</span> <span class="va">labels</span>, palette <span class="op">=</span> <span class="st">"Viridis"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">render_gif</span><span class="op">(</span><span class="va">f_rf_v_p_s</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>,</span>
<span>           <span class="fu">grand_tour</span><span class="op">(</span><span class="op">)</span>,</span>
<span>           <span class="fu">display_xy</span><span class="op">(</span>col <span class="op">=</span> <span class="va">f_rf_v_p_s</span><span class="op">$</span><span class="va">branches</span>, </span>
<span>           axes <span class="op">=</span> <span class="st">"off"</span>, half_range <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>           edges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">simp</span><span class="op">$</span><span class="va">edges</span><span class="op">)</span>,</span>
<span>           obs_labels <span class="op">=</span> <span class="va">labels</span>, palette<span class="op">=</span><span class="st">"Viridis"</span><span class="op">)</span>,</span>
<span>           gif_file<span class="op">=</span><span class="st">"gifs/ft_votes.gif"</span>,</span>
<span>           frames<span class="op">=</span><span class="fl">500</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="fig-ft-votes-html" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-ft-votes-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-ft-votes-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-ft-votes-tour" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-ft-votes-tour-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gifs/ft_votes.gif" class="img-fluid figure-img" data-ref-parent="fig-ft-votes-html" alt="FIX ME" width="300">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-ft-votes-tour-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) The 9D votes matrix for the 10 class fake_trees data in a tour.
</figcaption></figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-ft-votes-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-ft-votes-prj" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-ft-votes-prj-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/ft-votes.png" class="img-fluid figure-img" data-ref-parent="fig-ft-votes-html" alt="FIX ME" width="300">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-ft-votes-prj-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) Several static views from the tour revealing how clusters connect.
</figcaption></figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ft-votes-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;15.4: A tour and several static views of the votes matrix. Lines are the edges of the 8D simplex, which bounds the shape. Points mostly concentrate in the vertices, or spread along one of the edges, which means that most observations are clearly belonging to one group, or confused with a single other group. The exception to this is class 0, which spreads in many directions.
</figcaption></figure>
</div>
<p> The votes matrix is 9D, due to the 9 groups. With this many dimensions, if the cluster structure is weak, it will look messy in a tour. However, what we can see in <a href="#fig-ft-votes-html" class="quarto-xref">Figure&nbsp;<span>15.4</span></a> is that the structure is relatively simple, and very interesting in that it suggests a strong clustering of classes. Points are coloured by their true class. The lines represent the 8D simplex that bounds the observations, akin to the triangle in the ternary diagram.</p>
<p>Points concentrate at the vertices, which means that most are confidently predicted to be their true class. The most spread of points is along single edges, between pairs of vertices. This means that when there is confusion it is mostly with just one other group. One vertex (0) which has connections to all other vertexes. That is, there are points stretching from this vertex to every other. It means that some observations in every other class can be confused with class 0, and class 0 observations can be confused with every other class. This information suggests that cluster 0 is central to all the other clusters.</p>
<p>Some of this information could also be inferred from the confusion matrix for the model. However visualising the votes matrix provides more intricate details. Here we have seen that the points spread out from a vertex, with fewer and fewer the further one gets. It allows us to see the distribution of points, which is not possible from the confusion matrix alone. The same misclassification rate could be due to a variety of distributions. The visual pattern in the votes matrix is striking, and gives additional information about how the clustering distribution, and shapes of clusters, matches the class labels. It reinforces the clusters are linear extending into different dimensions in the 100D space, but really only into about 8D (as we’ll see from the variable importance explanation below). We also see that nine of the clusters are all connected to a single cluster.</p>
<div class="insight">
<p>The votes matrix for the fake trees has a striking geometric structure, with one central cluster connected to all other clusters, each of which is distinct from each other.</p>
</div>
</section><section id="sec-forest-var-imp" class="level3" data-number="15.2.2"><h3 data-number="15.2.2" class="anchored" data-anchor-id="sec-forest-var-imp">
<span class="header-section-number">15.2.2</span> Using variable importance</h3>
<p></p>
<p>The variable importance score across all classes, and for each class is useful for choosing variables to enter into a tour, to explore class differences. This is particularly so when there are many variables, as in the fake_trees data. We would also expect that this data will have a difference between importance for some classes.</p>
<div class="cell">
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://gt.rstudio.com/">gt</a></span><span class="op">)</span></span>
<span><span class="va">ft_rf</span><span class="op">$</span><span class="va">importance</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames<span class="op">=</span><span class="st">"Var"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>Acc<span class="op">=</span><span class="va">MeanDecreaseAccuracy</span>,</span>
<span>         Gini<span class="op">=</span><span class="va">MeanDecreaseGini</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co">#arrange(desc(Gini)) %&gt;%</span></span>
<span>  <span class="fu"><a href="https://gt.rstudio.com/reference/gt.html">gt</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://gt.rstudio.com/reference/fmt_number.html">fmt_number</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">`0`</span>,<span class="va">`1`</span>,<span class="va">`2`</span>,<span class="va">`3`</span>,<span class="va">`4`</span>,</span>
<span>                         <span class="va">`5`</span>,<span class="va">`6`</span>,<span class="va">`7`</span>,<span class="va">`8`</span>,<span class="va">`9`</span><span class="op">)</span>,</span>
<span>             decimals <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://gt.rstudio.com/reference/fmt_number.html">fmt_number</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="va">Acc</span>,</span>
<span>             decimals <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://gt.rstudio.com/reference/fmt_number.html">fmt_number</a></span><span class="op">(</span>columns <span class="op">=</span> <span class="va">Gini</span>,</span>
<span>             decimals <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="tbl-ft-importance" class="cell quarto-float anchored">
<figure class="quarto-float quarto-float-tbl figure"><figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-ft-importance-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;15.1: Variable importance from the random forest fit to the fake trees data, for each of the 9 classes, and using the accuracy and Gini metrics.
</figcaption><div aria-describedby="tbl-ft-importance-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<div id="ipjozlaatz" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#ipjozlaatz table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#ipjozlaatz thead, #ipjozlaatz tbody, #ipjozlaatz tfoot, #ipjozlaatz tr, #ipjozlaatz td, #ipjozlaatz th {
  border-style: none;
}

#ipjozlaatz p {
  margin: 0;
  padding: 0;
}

#ipjozlaatz .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#ipjozlaatz .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#ipjozlaatz .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#ipjozlaatz .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#ipjozlaatz .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#ipjozlaatz .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ipjozlaatz .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#ipjozlaatz .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#ipjozlaatz .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#ipjozlaatz .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#ipjozlaatz .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#ipjozlaatz .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#ipjozlaatz .gt_spanner_row {
  border-bottom-style: hidden;
}

#ipjozlaatz .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#ipjozlaatz .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#ipjozlaatz .gt_from_md > :first-child {
  margin-top: 0;
}

#ipjozlaatz .gt_from_md > :last-child {
  margin-bottom: 0;
}

#ipjozlaatz .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#ipjozlaatz .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#ipjozlaatz .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#ipjozlaatz .gt_row_group_first td {
  border-top-width: 2px;
}

#ipjozlaatz .gt_row_group_first th {
  border-top-width: 2px;
}

#ipjozlaatz .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#ipjozlaatz .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#ipjozlaatz .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#ipjozlaatz .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ipjozlaatz .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#ipjozlaatz .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#ipjozlaatz .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#ipjozlaatz .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#ipjozlaatz .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ipjozlaatz .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#ipjozlaatz .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#ipjozlaatz .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#ipjozlaatz .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#ipjozlaatz .gt_left {
  text-align: left;
}

#ipjozlaatz .gt_center {
  text-align: center;
}

#ipjozlaatz .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#ipjozlaatz .gt_font_normal {
  font-weight: normal;
}

#ipjozlaatz .gt_font_bold {
  font-weight: bold;
}

#ipjozlaatz .gt_font_italic {
  font-style: italic;
}

#ipjozlaatz .gt_super {
  font-size: 65%;
}

#ipjozlaatz .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#ipjozlaatz .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#ipjozlaatz .gt_indent_1 {
  text-indent: 5px;
}

#ipjozlaatz .gt_indent_2 {
  text-indent: 10px;
}

#ipjozlaatz .gt_indent_3 {
  text-indent: 15px;
}

#ipjozlaatz .gt_indent_4 {
  text-indent: 20px;
}

#ipjozlaatz .gt_indent_5 {
  text-indent: 25px;
}
</style>
<table class="gt_table do-not-create-environment cell table table-sm table-striped small" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead><tr class="header gt_col_headings">
<th id="Var" class="gt_col_heading gt_columns_bottom_border gt_left" data-quarto-table-cell-role="th" scope="col">Var</th>
<th id="0" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">0</th>
<th id="1" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">1</th>
<th id="2" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">2</th>
<th id="3" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">3</th>
<th id="4" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">4</th>
<th id="5" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">5</th>
<th id="6" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">6</th>
<th id="7" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">7</th>
<th id="8" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">8</th>
<th id="9" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">9</th>
<th id="Acc" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Acc</th>
<th id="Gini" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Gini</th>
</tr></thead>
<tbody class="gt_table_body">
<tr class="odd">
<td class="gt_row gt_left" headers="Var">PC1</td>
<td class="gt_row gt_right" headers="0">0.1</td>
<td class="gt_row gt_right" headers="1">0.4</td>
<td class="gt_row gt_right" headers="2">0.5</td>
<td class="gt_row gt_right" headers="3">0.3</td>
<td class="gt_row gt_right" headers="4">0.2</td>
<td class="gt_row gt_right" headers="5">0.5</td>
<td class="gt_row gt_right" headers="6">0.4</td>
<td class="gt_row gt_right" headers="7">0.2</td>
<td class="gt_row gt_right" headers="8">0.3</td>
<td class="gt_row gt_right" headers="9">0.3</td>
<td class="gt_row gt_right" headers="Acc">0.31</td>
<td class="gt_row gt_right" headers="Gini">490</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="Var">PC2</td>
<td class="gt_row gt_right" headers="0">0.1</td>
<td class="gt_row gt_right" headers="1">0.2</td>
<td class="gt_row gt_right" headers="2">0.2</td>
<td class="gt_row gt_right" headers="3">0.5</td>
<td class="gt_row gt_right" headers="4">0.3</td>
<td class="gt_row gt_right" headers="5">0.3</td>
<td class="gt_row gt_right" headers="6">0.2</td>
<td class="gt_row gt_right" headers="7">0.4</td>
<td class="gt_row gt_right" headers="8">0.2</td>
<td class="gt_row gt_right" headers="9">0.3</td>
<td class="gt_row gt_right" headers="Acc">0.28</td>
<td class="gt_row gt_right" headers="Gini">377</td>
</tr>
<tr class="odd">
<td class="gt_row gt_left" headers="Var">PC3</td>
<td class="gt_row gt_right" headers="0">0.1</td>
<td class="gt_row gt_right" headers="1">0.1</td>
<td class="gt_row gt_right" headers="2">0.1</td>
<td class="gt_row gt_right" headers="3">0.1</td>
<td class="gt_row gt_right" headers="4">0.5</td>
<td class="gt_row gt_right" headers="5">0.2</td>
<td class="gt_row gt_right" headers="6">0.1</td>
<td class="gt_row gt_right" headers="7">0.1</td>
<td class="gt_row gt_right" headers="8">0.2</td>
<td class="gt_row gt_right" headers="9">0.2</td>
<td class="gt_row gt_right" headers="Acc">0.17</td>
<td class="gt_row gt_right" headers="Gini">313</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="Var">PC4</td>
<td class="gt_row gt_right" headers="0">0.1</td>
<td class="gt_row gt_right" headers="1">0.5</td>
<td class="gt_row gt_right" headers="2">0.1</td>
<td class="gt_row gt_right" headers="3">0.0</td>
<td class="gt_row gt_right" headers="4">0.1</td>
<td class="gt_row gt_right" headers="5">0.0</td>
<td class="gt_row gt_right" headers="6">0.4</td>
<td class="gt_row gt_right" headers="7">0.2</td>
<td class="gt_row gt_right" headers="8">0.1</td>
<td class="gt_row gt_right" headers="9">0.1</td>
<td class="gt_row gt_right" headers="Acc">0.14</td>
<td class="gt_row gt_right" headers="Gini">328</td>
</tr>
<tr class="odd">
<td class="gt_row gt_left" headers="Var">PC5</td>
<td class="gt_row gt_right" headers="0">0.1</td>
<td class="gt_row gt_right" headers="1">0.1</td>
<td class="gt_row gt_right" headers="2">0.3</td>
<td class="gt_row gt_right" headers="3">0.1</td>
<td class="gt_row gt_right" headers="4">0.2</td>
<td class="gt_row gt_right" headers="5">0.2</td>
<td class="gt_row gt_right" headers="6">0.1</td>
<td class="gt_row gt_right" headers="7">0.1</td>
<td class="gt_row gt_right" headers="8">0.3</td>
<td class="gt_row gt_right" headers="9">0.2</td>
<td class="gt_row gt_right" headers="Acc">0.18</td>
<td class="gt_row gt_right" headers="Gini">329</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="Var">PC6</td>
<td class="gt_row gt_right" headers="0">0.1</td>
<td class="gt_row gt_right" headers="1">0.2</td>
<td class="gt_row gt_right" headers="2">0.3</td>
<td class="gt_row gt_right" headers="3">0.2</td>
<td class="gt_row gt_right" headers="4">0.0</td>
<td class="gt_row gt_right" headers="5">0.1</td>
<td class="gt_row gt_right" headers="6">0.0</td>
<td class="gt_row gt_right" headers="7">0.3</td>
<td class="gt_row gt_right" headers="8">0.1</td>
<td class="gt_row gt_right" headers="9">0.2</td>
<td class="gt_row gt_right" headers="Acc">0.16</td>
<td class="gt_row gt_right" headers="Gini">291</td>
</tr>
<tr class="odd">
<td class="gt_row gt_left" headers="Var">PC7</td>
<td class="gt_row gt_right" headers="0">0.1</td>
<td class="gt_row gt_right" headers="1">0.0</td>
<td class="gt_row gt_right" headers="2">0.2</td>
<td class="gt_row gt_right" headers="3">0.0</td>
<td class="gt_row gt_right" headers="4">0.0</td>
<td class="gt_row gt_right" headers="5">0.1</td>
<td class="gt_row gt_right" headers="6">0.1</td>
<td class="gt_row gt_right" headers="7">0.3</td>
<td class="gt_row gt_right" headers="8">0.1</td>
<td class="gt_row gt_right" headers="9">0.1</td>
<td class="gt_row gt_right" headers="Acc">0.11</td>
<td class="gt_row gt_right" headers="Gini">250</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="Var">PC8</td>
<td class="gt_row gt_right" headers="0">0.0</td>
<td class="gt_row gt_right" headers="1">0.1</td>
<td class="gt_row gt_right" headers="2">0.0</td>
<td class="gt_row gt_right" headers="3">0.2</td>
<td class="gt_row gt_right" headers="4">0.1</td>
<td class="gt_row gt_right" headers="5">0.1</td>
<td class="gt_row gt_right" headers="6">0.0</td>
<td class="gt_row gt_right" headers="7">0.0</td>
<td class="gt_row gt_right" headers="8">0.1</td>
<td class="gt_row gt_right" headers="9">0.3</td>
<td class="gt_row gt_right" headers="Acc">0.09</td>
<td class="gt_row gt_right" headers="Gini">218</td>
</tr>
<tr class="odd">
<td class="gt_row gt_left" headers="Var">PC9</td>
<td class="gt_row gt_right" headers="0">0.1</td>
<td class="gt_row gt_right" headers="1">0.0</td>
<td class="gt_row gt_right" headers="2">0.0</td>
<td class="gt_row gt_right" headers="3">0.0</td>
<td class="gt_row gt_right" headers="4">0.0</td>
<td class="gt_row gt_right" headers="5">0.0</td>
<td class="gt_row gt_right" headers="6">0.0</td>
<td class="gt_row gt_right" headers="7">0.0</td>
<td class="gt_row gt_right" headers="8">0.1</td>
<td class="gt_row gt_right" headers="9">0.0</td>
<td class="gt_row gt_right" headers="Acc">0.02</td>
<td class="gt_row gt_right" headers="Gini">60</td>
</tr>
<tr class="even">
<td class="gt_row gt_left" headers="Var">PC10</td>
<td class="gt_row gt_right" headers="0">0.0</td>
<td class="gt_row gt_right" headers="1">0.0</td>
<td class="gt_row gt_right" headers="2">0.0</td>
<td class="gt_row gt_right" headers="3">0.0</td>
<td class="gt_row gt_right" headers="4">0.0</td>
<td class="gt_row gt_right" headers="5">0.0</td>
<td class="gt_row gt_right" headers="6">0.0</td>
<td class="gt_row gt_right" headers="7">0.0</td>
<td class="gt_row gt_right" headers="8">0.0</td>
<td class="gt_row gt_right" headers="9">0.0</td>
<td class="gt_row gt_right" headers="Acc">0.01</td>
<td class="gt_row gt_right" headers="Gini">44</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</figure>
</div>
</div>
<p>From the variable importance (<a href="#tbl-ft-importance" class="quarto-xref">Table&nbsp;<span>15.1</span></a>), we can see that PC9 and PC10 do not substantially contribute. That means the 100D data can be reduced to 8D without losing the information about the cluster structure. PC1 is most important overall, and the order matches the PC order, as might be expected because highest variance corresponds to the most spread clusters. Each cluster has a different set of variables that are important. For example, the variables important for distinguishing cluster 1 are PC1 and PC4, and for cluster 2 they are PC1 and PC5.</p>
<div class="info">
<p>Class-wise variable importance helps to find a subspace on which to tour to examine how this class cluster differs from the others.</p>
</div>
<p>We can use the accuracy information to choose variables to provide to the tour. Overall, one would sequentially add the variables into a tour based on their accuracy or Gini value. Here it is simply starting with the first three PCs, and then sequentially adding the PCs to examine how distinct the clusters are with and without the extra variable. It can be helpful to focus on a single class against all the others. To do this create a new binary class variable, indicating that the observation belongs to class <span class="math inline">\(k\)</span> or not, as follows:</p>
<div class="cell">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ft_pc</span> <span class="op">&lt;-</span> <span class="va">ft_pc</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>cl1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html">case_when</a></span><span class="op">(</span></span>
<span>                 <span class="va">branches</span> <span class="op">==</span> <span class="st">"0"</span> <span class="op">~</span> <span class="st">"0"</span>,</span>
<span>                 <span class="va">branches</span> <span class="op">==</span> <span class="st">"1"</span> <span class="op">~</span> <span class="st">"1"</span>,</span>
<span>                 .default <span class="op">=</span> <span class="st">"other"</span></span>
<span>  <span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<details class="code-fold"><summary>Code to make animated gifs</summary><div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">animate_xy</span><span class="op">(</span><span class="va">ft_pc</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC4"</span>, <span class="st">"PC6"</span><span class="op">)</span><span class="op">]</span>, col<span class="op">=</span><span class="va">ft_pc</span><span class="op">$</span><span class="va">cl1</span>, palette<span class="op">=</span><span class="st">"Viridis"</span><span class="op">)</span></span>
<span><span class="fu">render_gif</span><span class="op">(</span><span class="va">ft_pc</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC4"</span>, <span class="st">"PC6"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>           <span class="fu">grand_tour</span><span class="op">(</span><span class="op">)</span>,</span>
<span>           <span class="fu">display_xy</span><span class="op">(</span>col<span class="op">=</span><span class="va">ft_pc</span><span class="op">$</span><span class="va">cl1</span>, palette<span class="op">=</span><span class="st">"Viridis"</span><span class="op">)</span>,</span>
<span>           gif_file<span class="op">=</span><span class="st">"gifs/ft_cl1.gif"</span>,</span>
<span>           frames<span class="op">=</span><span class="fl">500</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold"><summary>Code to make plot</summary><div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ft_pc_cl1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ft_pc</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">PC4</span>, y<span class="op">=</span><span class="va">PC1</span>, col<span class="op">=</span><span class="va">cl1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.7</span>, size<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_sequential.html">scale_color_discrete_sequential</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Viridis"</span>, rev<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>From <a href="#fig-ft-cl-html" class="quarto-xref">Figure&nbsp;<span>15.5</span></a> we can see how cluster 1 is distinct from all of the other observations, albeit with a close connection to the trunk of the tree (cluster 0). The distinction is visible whenever PC4 contributes to the projection, but can be seen clearly with only PC1 and PC4.</p>
<div id="fig-ft-cl-html" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-ft-cl-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-ft-cl-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-ft-cl1" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-ft-cl1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gifs/ft_cl1.gif" class="img-fluid figure-img" data-ref-parent="fig-ft-cl-html" alt="FIX ME" width="300">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-ft-cl1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Tour of most important variables for class 1.
</figcaption></figure>
</div>
</div>
<div class="cell quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-ft-cl-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div class="cell-output-display quarto-layout-cell-subref" data-ref-parent="fig-ft-cl-html">
<div id="fig-ft-cl1-pc" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-ft-cl1-pc-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="15-forests_files/figure-html/fig-ft-cl1-pc-1.png" class="img-fluid figure-img" style="width:100.0%" data-ref-parent="fig-ft-cl-html">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-ft-cl1-pc-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) PC1 and PC4 together reveal cluster 1.
</figcaption></figure>
</div>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ft-cl-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;15.5: Focusing on class 1 in the <code>fake_trees</code> data. The most important variables were PC1 and PC4. A combination of PC2 and PC4 reveals the difference between cluster 1 and all the other clusters.
</figcaption></figure>
</div>
<p>For a problem like this, it can be useful to several classes together. We’ve chosen to start with class 8 (light green), because from <a href="#fig-ft-votes-html" class="quarto-xref">Figure&nbsp;<span>15.4</span></a> it appears to have less connection with class 0, and closer connection with another class. This is class 6 (medium green). A good guess because it has one observation confused with class 8 according to the confusion matrix (printed below). </p>
<p>When we examine these two clusters in association with class 0, we can see that there is a third cluster that is connected with clusters 6 and 8. It turns out to be cluster 1. It’s confusing, because the confusion matrix would suggest that the overlap from all is with cluster 0, but not each other.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ft_rf</span><span class="op">$</span><span class="va">confusion</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>    0   1   2   3   4   5   6   7   8   9 class.error
0 266   5   1   3   3   4   5   2   7   4       0.113
1  13 287   0   0   0   0   0   0   0   0       0.043
2   9   0 290   0   1   0   0   0   0   0       0.033
3   4   0   0 289   0   0   0   5   0   2       0.037
4  11   0   0   0 289   0   0   0   0   0       0.037
5  11   0   0   0   0 289   0   0   0   0       0.037
6  13   0   0   0   0   0 286   0   1   0       0.047
7   6   0   0   3   0   0   0 291   0   0       0.030
8   8   0   0   0   0   0   0   0 292   0       0.027
9   5   0   0   0   0   0   0   0   0 295       0.017</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ft_pc</span> <span class="op">&lt;-</span> <span class="va">ft_pc</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>cl8 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html">case_when</a></span><span class="op">(</span></span>
<span>                 <span class="va">branches</span> <span class="op">==</span> <span class="st">"0"</span> <span class="op">~</span> <span class="st">"0"</span>,</span>
<span>                 <span class="va">branches</span> <span class="op">==</span> <span class="st">"6"</span> <span class="op">~</span> <span class="st">"6"</span>,</span>
<span>                 <span class="va">branches</span> <span class="op">==</span> <span class="st">"1"</span> <span class="op">~</span> <span class="st">"1"</span>,</span>
<span>                 <span class="va">branches</span> <span class="op">==</span> <span class="st">"8"</span> <span class="op">~</span> <span class="st">"8"</span>,</span>
<span>                 .default <span class="op">=</span> <span class="st">"other"</span></span>
<span>  <span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<details class="code-fold"><summary>Code to make animated gif</summary><div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">animate_xy</span><span class="op">(</span><span class="va">ft_pc</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC4"</span>, <span class="st">"PC5"</span>, <span class="st">"PC6"</span><span class="op">)</span><span class="op">]</span>, col<span class="op">=</span><span class="va">ft_pc</span><span class="op">$</span><span class="va">cl8</span>, palette<span class="op">=</span><span class="st">"Viridis"</span><span class="op">)</span></span>
<span><span class="fu">render_gif</span><span class="op">(</span><span class="va">ft_pc</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"PC1"</span>, <span class="st">"PC2"</span>, <span class="st">"PC4"</span>, <span class="st">"PC5"</span>, <span class="st">"PC6"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>           <span class="fu">grand_tour</span><span class="op">(</span><span class="op">)</span>,</span>
<span>           <span class="fu">display_xy</span><span class="op">(</span>col<span class="op">=</span><span class="va">ft_pc</span><span class="op">$</span><span class="va">cl8</span>, palette<span class="op">=</span><span class="st">"Viridis"</span><span class="op">)</span>,</span>
<span>           gif_file<span class="op">=</span><span class="st">"gifs/ft_cl8.gif"</span>,</span>
<span>           frames<span class="op">=</span><span class="fl">500</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold"><summary>Code to make plot</summary><div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ft_pc_cl8</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">ft_pc</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">PC1</span>, y<span class="op">=</span><span class="va">PC5</span>, col<span class="op">=</span><span class="va">cl8</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.7</span>, size<span class="op">=</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_sequential.html">scale_color_discrete_sequential</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Viridis"</span>, rev<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>From the tour in <a href="#fig-ft-cl2" class="quarto-xref">Figure&nbsp;<span>15.6</span></a> we can see that clusters 1, 6, and 8 share one end of the trunk (cluster 0). Cluster 8 is almost more closely connected with cluster 6, though, than cluster 0. PC1 and PC5 mostly show the distinction between cluster 8 and the rest of the points, but it is clearer if more variables are used.</p>
<div id="fig-ft-cl2" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-ft-cl2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-ft-cl2" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-ft-cl8" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-ft-cl8-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gifs/ft_cl8.gif" class="img-fluid figure-img" data-ref-parent="fig-ft-cl2" alt="FIX ME" width="300">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-ft-cl8-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Tour of most important variables for class 1.
</figcaption></figure>
</div>
</div>
<div class="cell quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-ft-cl2" style="flex-basis: 50.0%;justify-content: flex-start;">
<div class="cell-output-display quarto-layout-cell-subref" data-ref-parent="fig-ft-cl2">
<div id="fig-ft-cl8-pc" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-ft-cl8-pc-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="15-forests_files/figure-html/fig-ft-cl8-pc-1.png" class="img-fluid figure-img" style="width:100.0%" data-ref-parent="fig-ft-cl2">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-ft-cl8-pc-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) PC1 and PC5 together mostly reveal cluster 8.
</figcaption></figure>
</div>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ft-cl2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;15.6: Focusing on class 8 in the <code>fake_trees</code> data, relative to nearby clusters 1 and 6. The most important variables for cluster 8 are PC1, PC2, PC5, but to explore in association with clusters 1 and 6, we include PC4 and PC6. A combination of PC1 and PC5 reveals the difference between cluster 8, 6, 1 and 0.
</figcaption></figure>
</div>
<div class="insight">
<p>Although the confusion matrix suggests that class clusters are separated except for class 0, focusing on a few classes and using the variable importance to examine smaller subspaces, reveals they are connected in groups of three to class 0.</p>
</div>
</section></section><section id="exercises" class="level2 unnumbered"><h2 class="unnumbered anchored" data-anchor-id="exercises">Exercises</h2>
<ol type="1">
<li>Using a grand tour compare the boundaries from the random forest model on the <code>penguins</code> data to that of (a) a default tree model, (b) an LDA model. Is it less boxy than the tree model, but still more boxy than that of the LDA model?</li>
<li>Tinker with the parameters of the tree model to force it to fit a tree more closely to the data. Compare the boundaries from this with the default tree, and with the forest model. Is it less boxy than the default tree, but more boxy than the forest model?</li>
<li>Fit a random forest model to the <code>bushfires</code> data using the <code>cause</code> variable as the class. It is a highly imbalanced classification problem. What is the out-of-bag error rate for the forest? Are there some classes that have lower error rate than others? Examine the 4D votes matrix with a tour, and describe the confusion between classes. This is interesting because it is difficult to accurately classify the fire ignition cause, and only some groups are often confused with each other. You should be able to see this from the 3D votes matrix.</li>
<li>Fit a forest model to the first 21 PCs of the <code>sketches</code> data. Explore the 5D votes matrix. Why does it look star-shaped?</li>
<li>Choose a cluster (or group of clusters) from the <code>fake_trees</code> data (2, 3, 4, 5, 7, 9) to explore in detail like done in <a href="#sec-forest-var-imp" class="quarto-xref"><span>Section 15.2.2</span></a>. Be sure to choose which PCs are the most useful using a tour, and follow-up by making a scatterplot showing the best distinction between your chosen cluster and the other observations.</li>
</ol>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list" style="display: none">
<div id="ref-Ab1884" class="csl-entry" role="listitem">
Abbott, E. (1884). <em>Flatland: A romance of many dimensions</em>. Dover Publications.
</div>
<div id="ref-AWS91" class="csl-entry" role="listitem">
Ahlberg, C., Williamson, C., &amp; Shneiderman, B. (1991). Dynamic <span>Q</span>ueries for <span>I</span>nformation <span>E</span>xploration: <span>A</span>n <span>I</span>mplementation and <span>E</span>valuation. <em>ACM CHI ‘92 Conference Proceedings</em>, 619–626.
</div>
<div id="ref-keras" class="csl-entry" role="listitem">
Allaire, J., &amp; Chollet, F. (2023). <em>Keras: R interface to ’keras’</em>. <a href="https://CRAN.R-project.org/package=keras">https://CRAN.R-project.org/package=keras</a>
</div>
<div id="ref-An57" class="csl-entry" role="listitem">
Anderson, E. (1957). A <span>S</span>emigraphical <span>M</span>ethod for the <span>A</span>nalysis of <span>C</span>omplex <span>P</span>roblems. <em>Proceedings of the National Academy of Science, 13</em>, 923–927.
</div>
<div id="ref-An72" class="csl-entry" role="listitem">
Andrews, D. F. (1972). <span>P</span>lots of <span>H</span>igh-dimensional <span>D</span>ata. <em>Biometrics</em>, <em>28</em>, 125–136.
</div>
<div id="ref-AGW71" class="csl-entry" role="listitem">
Andrews, D. F., Gnanadesikan, R., &amp; Warner, J. L. (1971). <span>T</span>ransformations of <span>M</span>ultivariate <span>D</span>ata. <em>Biometrics</em>, <em>27</em>, 825–840.
</div>
<div id="ref-AB97" class="csl-entry" role="listitem">
Anselin, L., &amp; Bao, S. (1997). <span>E</span>xploratory <span>S</span>patial <span>D</span>ata <span>A</span>nalysis <span>L</span>inking <span>S</span>pace<span>S</span>tat and <span>A</span>rc<span>V</span>iew. In M. M. Fischer &amp; A. Getis (Eds.), <em><span>R</span>ecent <span>D</span>evelopments in <span>S</span>patial <span>A</span>nalysis</em> (pp. 35–59). Springer.
</div>
<div id="ref-R-ggthemes" class="csl-entry" role="listitem">
Arnold, J. B. (2021). <em>Ggthemes: Extra themes, scales and geoms for ggplot2</em>. <a href="https://github.com/jrnold/ggthemes">https://github.com/jrnold/ggthemes</a>
</div>
<div id="ref-ASA23" class="csl-entry" role="listitem">
ASA Statistical Graphics Section. (2023). <em>Video <span>L</span>ibrary</em>. <a href="https://community.amstat.org/jointscsg-section/media/videos" class="uri">https://community.amstat.org/jointscsg-section/media/videos</a>.
</div>
<div id="ref-As85" class="csl-entry" role="listitem">
Asimov, D. (1985). <span>T</span>he <span>G</span>rand <span>T</span>our: <span>A</span> <span>T</span>ool for <span>V</span>iewing <span>M</span>ultidimensional <span>D</span>ata. <em>SIAM Journal of Scientific and Statistical Computing</em>, <em>6</em>(1), 128–143.
</div>
<div id="ref-R-gridExtra" class="csl-entry" role="listitem">
Auguie, B. (2017). <em>gridExtra: Miscellaneous functions for "grid" graphics</em>. <a href="https://CRAN.R-project.org/package=gridExtra">https://CRAN.R-project.org/package=gridExtra</a>
</div>
<div id="ref-forest" class="csl-entry" role="listitem">
Australian Bureau of Agricultural and Resource Economics and Sciences. (2018). <em><span class="nocase">Forests of Australia</span></em>. <a href="https://www.agriculture.gov.au/abares/forestsaustralia/forest-data-maps-and-tools/spatial-data/forest-cover">https://www.agriculture.gov.au/abares/forestsaustralia/forest-data-maps-and-tools/spatial-data/forest-cover</a>
</div>
<div id="ref-BC84" class="csl-entry" role="listitem">
Becker, R. A., &amp; Chambers, J. M. (1984). <em><span>S</span>: An environment for data analysis and graphics</em>. Wadsworth.
</div>
<div id="ref-BC87" class="csl-entry" role="listitem">
Becker, R. A., &amp; Cleveland, W. S. (1988). <span>B</span>rushing <span>S</span>catterplots. In W. S. Cleveland &amp; M. E. McGill (Eds.), <em>Dynamic graphics for statistics</em> (pp. 201–224). Wadsworth.
</div>
<div id="ref-BeCS96" class="csl-entry" role="listitem">
Becker, R., Cleveland, W. S., &amp; Shyu, M.-J. (1996). <span>T</span>he <span>V</span>isual <span>D</span>esign and <span>C</span>ontrol of <span>T</span>rellis <span>D</span>isplays. <em>Journal of Computational and Graphical Statistics</em>, <em>6</em>(1), 123–155.
</div>
<div id="ref-BS03" class="csl-entry" role="listitem">
Bederson, B. B., &amp; Schneiderman, B. (2003). <em>The craft of information visualization: Readings and reflections</em>. Morgan Kaufmann.
</div>
<div id="ref-BellmanRichard1961" class="csl-entry" role="listitem">
Bellman, R. (1961). <em>Adaptive control processes : A guided tour</em>.
</div>
<div id="ref-bickel2018" class="csl-entry" role="listitem">
Bickel, P. J., Kur, G., &amp; Nadler, B. (2018). Projection pursuit in high dimensions. <em>Proceedings of the National Academy of Sciences</em>, <em>115</em>, 9151–9156. <a href="https://doi.org/10.1073/pnas.1801177115">https://doi.org/10.1073/pnas.1801177115</a>
</div>
<div id="ref-Bi06" class="csl-entry" role="listitem">
Bishop, C. M. (2006). <em>Pattern <span>R</span>ecognition and <span>M</span>achine <span>L</span>earning</em>. Springer.
</div>
<div id="ref-HOML" class="csl-entry" role="listitem">
Boehmke, B., &amp; Greenwell, B. M. (2019). <em>Hands-on machine learning with r (1st ed.)</em>. Chapman; Hall/CRC. <a href="https://doi.org/10.1201/9780367816377">https://doi.org/10.1201/9780367816377</a>
</div>
<div id="ref-R-aweSOM" class="csl-entry" role="listitem">
Boelaert, J., Ollion, E., &amp; Sodoge, J. (2022). <em>aweSOM: Interactive self-organizing maps</em>. <a href="https://CRAN.R-project.org/package=aweSOM">https://CRAN.R-project.org/package=aweSOM</a>
</div>
<div id="ref-Bonneau2006" class="csl-entry" role="listitem">
Bonneau, G.-P., Ertl, T., &amp; Nielson, G. M. (Eds.). (2006). <em>Scientific visualization: The visual extraction of knowledge from data</em>. Springer.
</div>
<div id="ref-BG05" class="csl-entry" role="listitem">
Borg, I., &amp; Groenen, P. J. F. (2005). <em>Modern <span>M</span>ultidimensional <span>S</span>caling</em>. Springer.
</div>
<div id="ref-Br01" class="csl-entry" role="listitem">
Breiman, L. (2001). Random <span>F</span>orests. <em>Machine Learning</em>, <em>45</em>(1), 5–32.
</div>
<div id="ref-R-randomForest" class="csl-entry" role="listitem">
Breiman, L., Cutler, A., Liaw, A., &amp; Wiener, M. (2022). <em>randomForest: Breiman and cutler’s random forests for classification and regression</em>. <a href="https://www.stat.berkeley.edu/~breiman/RandomForests/">https://www.stat.berkeley.edu/~breiman/RandomForests/</a>
</div>
<div id="ref-BFOS84" class="csl-entry" role="listitem">
Breiman, L., Friedman, J., Olshen, C., &amp; Stone, C. (1984). <em>Classification and <span>R</span>egression <span>T</span>rees</em>. Wadsworth; Brooks/Cole.
</div>
<div id="ref-BujaKS96" class="csl-entry" role="listitem">
Buja, A. (1996). Interactive <span>G</span>raphical <span>M</span>ethods in the <span>A</span>nalysis of <span>C</span>ustomer <span>P</span>anel <span>D</span>ata: <span>C</span>omment. <em>Journal of Business &amp; Economic Statistics</em>, <em>14</em>(1), 128–129.
</div>
<div id="ref-BA86b" class="csl-entry" role="listitem">
Buja, A., &amp; Asimov, D. (1986). <span>G</span>rand <span>T</span>our <span>M</span>ethods: <span>A</span>n <span>O</span>utline. <em>Computing Science and Statistics</em>, <em>17</em>, 63–67.
</div>
<div id="ref-BAHM88" class="csl-entry" role="listitem">
Buja, A., Asimov, D., Hurley, C., &amp; McDonald, J. A. (1988). <span>E</span>lements of a <span>V</span>iewing <span>P</span>ipeline for <span>D</span>ata <span>A</span>nalysis. In W. S. Cleveland &amp; M. E. McGill (Eds.), <em>Dynamic graphics for statistics</em> (pp. 277–308). Wadsworth.
</div>
<div id="ref-BCAH95c" class="csl-entry" role="listitem">
Buja, A., Cook, D., Asimov, D., &amp; Hurley, C. (1997). <em>Dynamic <span>P</span>rojections in <span>H</span>igh-<span>D</span>imensional <span>V</span>isualization: <span>T</span>heory and <span>C</span>omputational <span>M</span>ethods</em>. AT&amp;<span>T</span> Labs.
</div>
<div id="ref-BCAH05" class="csl-entry" role="listitem">
Buja, A., Cook, D., Asimov, D., &amp; Hurley, C. (2005). <span>C</span>omputational <span>M</span>ethods for <span>H</span>igh-<span>D</span>imensional <span>R</span>otations in <span>D</span>ata <span>V</span>isualization. In C. R. Rao, E. J. Wegman, &amp; J. L. Solka (Eds.), <em>Handbook of statistics: Data mining and visualization</em> (pp. 391–414). Elsevier/North-Holland.
</div>
<div id="ref-BCS95" class="csl-entry" role="listitem">
Buja, A., Cook, D., &amp; Swayne, D. (1996). <span>I</span>nteractive <span>H</span>igh-<span>D</span>imensional <span>D</span>ata <span>V</span>isualization. <em>Journal of Computational and Graphical Statistics</em>, <em>5</em>(1), 78–99.
</div>
<div id="ref-BHM86" class="csl-entry" role="listitem">
Buja, A., Hurley, C., &amp; McDonald, J. A. (1986). A <span>D</span>ata <span>V</span>iewer for <span>M</span>ultivariate <span>D</span>ata. <em>Computing Science and Statistics</em>, <em>17</em>(1), 171–174.
</div>
<div id="ref-xgvisjclass" class="csl-entry" role="listitem">
Buja, A., &amp; Swayne, D. F. (2002). Visualization <span>M</span>ethodology for <span>M</span>ultidimensional <span>S</span>caling. <em>Journal of Classification</em>, <em>19</em>(1), 7–43.
</div>
<div id="ref-ggvisjcgs" class="csl-entry" role="listitem">
Buja, A., Swayne, D. F., Littman, M. L., Dean, N., Hofmann, H., &amp; Chen, L. (2008). Data visualization with multidimensional scaling. <em>Journal of Computational and Graphical Statistics</em>, <em>17</em>(2), 444–472. <a href="https://doi.org/10.1198/106186008X318440">https://doi.org/10.1198/106186008X318440</a>
</div>
<div id="ref-BT91" class="csl-entry" role="listitem">
Buja, A., &amp; Tukey, P. (Eds.). (1991). <em>Computing and <span>G</span>raphics in <span>S</span>tatistics</em>. Springer-Verlag.
</div>
<div id="ref-CMS99" class="csl-entry" role="listitem">
Card, S. K., Mackinlay, J. D., &amp; Schneiderman, B. (1999). <em>Readings in information visualization</em>. Morgan Kaufmann Publishers.
</div>
<div id="ref-CWL96" class="csl-entry" role="listitem">
Carr, D. B., Wegman, E. J., &amp; Luo, Q. (1996). <em>Explor<span>N</span>: <span>D</span>esign <span>C</span>onsiderations <span>P</span>ast and <span>P</span>resent</em> (Technical Report No. 129). Center for Computational Statistics, George Mason University.
</div>
<div id="ref-Ch95" class="csl-entry" role="listitem">
Chatfield, C. (1995). <em>Problem solving: A statistician’s guide</em>. Chapman; Hall/CRC Press.
</div>
<div id="ref-haerdle" class="csl-entry" role="listitem">
Chen, C., Härdle, W., &amp; Unwin, A. (Eds.). (2006). <em>Handbook of computational statistics (volume III) data visualization</em>. Springer.
</div>
<div id="ref-UHC06" class="csl-entry" role="listitem">
Chen, C.-H., Härdle, W., &amp; Unwin, A. (Eds.). (2007). <em>Handbook of <span>D</span>ata <span>V</span>isualization</em>. Springer.
</div>
<div id="ref-CT94" class="csl-entry" role="listitem">
Cheng, B., &amp; Titterington, M. (1994). Neural <span>N</span>etworks: <span>A</span> <span>R</span>eview from a <span>S</span>tatistical <span>P</span>erspective. <em>Statistical Science</em>, <em>9</em>(1), 2–30.
</div>
<div id="ref-R-crosstalk" class="csl-entry" role="listitem">
Cheng, J., &amp; Sievert, C. (2021). <em>Crosstalk: Inter-widget interactivity for HTML widgets</em>. <a href="https://rstudio.github.io/crosstalk/">https://rstudio.github.io/crosstalk/</a>
</div>
<div id="ref-Ch73" class="csl-entry" role="listitem">
Chernoff, H. (1973). The <span>U</span>se of <span>F</span>aces to <span>R</span>epresent <span>P</span>oints in <span class="math inline">\(k\)</span>-dimensional <span>S</span>pace <span>G</span>raphically. <em>Journal of the American Statistical Association</em>, <em>68</em>, 361–368.
</div>
<div id="ref-Cl79" class="csl-entry" role="listitem">
Cleveland, W. S. (1979). Robust <span>L</span>ocaly <span>W</span>eighted <span>R</span>egression and <span>S</span>moothing <span>S</span>catterplots. <em>Journal of American Statistics Association</em>, <em>74</em>, 829–836.
</div>
<div id="ref-Cl93" class="csl-entry" role="listitem">
Cleveland, W. S. (1993). <em>Visualizing <span>D</span>ata</em>. Hobart Press.
</div>
<div id="ref-CM88" class="csl-entry" role="listitem">
Cleveland, W. S., &amp; McGill, M. E. (Eds.). (1988). <em>Dynamic graphics for statistics</em>. Wadsworth.
</div>
<div id="ref-CB95" class="csl-entry" role="listitem">
Cook, D., &amp; Buja, A. (1997). <span>M</span>anual <span>C</span>ontrols <span>F</span>or <span>H</span>igh-<span>D</span>imensional <span>D</span>ata <span>P</span>rojections. <em>Journal of Computational and Graphical Statistics</em>, <em>6</em>(4), 464–480.
</div>
<div id="ref-CBC93" class="csl-entry" role="listitem">
Cook, D., Buja, A., &amp; Cabrera, J. (1993). <span>P</span>rojection <span>P</span>ursuit <span>I</span>ndexes <span>B</span>ased on <span>O</span>rthonormal <span>F</span>unction <span>E</span>xpansions. <em>Journal of Computational and Graphical Statistics</em>, <em>2</em>(3), 225–250.
</div>
<div id="ref-CBCH95" class="csl-entry" role="listitem">
Cook, D., Buja, A., Cabrera, J., &amp; Hurley, C. (1995b). Grand <span>T</span>our and <span>P</span>rojection <span>P</span>ursuit. <em>Journal of Computational and Graphical Statistics</em>, <em>4</em>(3), 155–172.
</div>
<div id="ref-CBCH94" class="csl-entry" role="listitem">
Cook, D., Buja, A., Cabrera, J., &amp; Hurley, C. (1995a). Grand <span>T</span>our and <span>P</span>rojection <span>P</span>ursuit. <em>Journal of Computational and Graphical Statistics</em>, <em>4</em>(3), 155–172.
</div>
<div id="ref-CHLYNW06" class="csl-entry" role="listitem">
Cook, D., Hofmann, H., Lee, E.-K., Yang, H., Nikolau, B., &amp; Wurtele, E. (2007). Exploring <span>G</span>ene <span>E</span>xpression <span>D</span>ata, <span>U</span>sing <span>P</span>lots. <em>Journal of Data Science</em>, <em>5</em>(2), 151–182.
</div>
<div id="ref-R-mulgar" class="csl-entry" role="listitem">
Cook, D., &amp; Laa, U. (2023). <em>Mulgar: Functions for pre-processing data for multivariate data visualisation using tours</em>. <a href="https://dicook.github.io/mulgar/">https://dicook.github.io/mulgar/</a>
</div>
<div id="ref-CLBW06" class="csl-entry" role="listitem">
Cook, D., Lee, E.-K., Buja, A., &amp; Wickham, H. (2006). <span>G</span>rand <span>T</span>ours, <span>P</span>rojection <span>P</span>ursuit <span>G</span>uided <span>T</span>ours and <span>M</span>anual <span>C</span>ontrols. In C.-H. Chen, W. Härdle, &amp; A. Unwin (Eds.), <em>Handbook of <span>D</span>ata <span>V</span>isualization</em>. Springer.
</div>
<div id="ref-CMSC95" class="csl-entry" role="listitem">
Cook, D., Majure, J. J., Symanzik, J., &amp; Cressie, N. (1996). <span>D</span>ynamic <span>G</span>raphics in a <span>GIS</span>: <span>E</span>xploring and <span>A</span>nalyzing <span>M</span>ultivariate <span>S</span>patial <span>D</span>ata using <span>L</span>inked <span>S</span>oftware. <em>Computational Statistics: Special Issue on Computer Aided Analyses of Spatial Data</em>, <em>11</em>(4), 467–480.
</div>
<div id="ref-CS07" class="csl-entry" role="listitem">
Cook, D., &amp; Swayne, D. F. (2007). <em>Interactive and dynamic graphics for data analysis: With <span>R</span> and <span>GGobi</span></em>. Springer-Verlag. <a href="https://doi.org/10.1007/978-0-387-71762-3">https://doi.org/10.1007/978-0-387-71762-3</a>
</div>
<div id="ref-CPV2003" class="csl-entry" role="listitem">
Cortes, C., Pregibon, D., &amp; Volinsky, C. (2003). Computational <span>M</span>ethods for <span>D</span>ynamic <span>G</span>raphs. <em>Journal of Computational &amp; Graphical Statistics</em>, <em>12</em>(4), 950–970.
</div>
<div id="ref-CV95" class="csl-entry" role="listitem">
Cortes, C., &amp; Vapnik, V. N. (1995). Support-<span>V</span>ector <span>N</span>etworks. <em>Machine Learning</em>, <em>20</em>(3), 273–297.
</div>
<div id="ref-Oc1885" class="csl-entry" role="listitem">
d’Ocagne, M. (1885). <em><span>C</span>oordonnées <span>P</span>arallèles et <span>A</span>xiales: <span>M</span>éthode de transformation géométrique et procédé nouveau de calcul graphique déduits de la considération des coordonnées paralléles</em>. Gauthier-Villars.
</div>
<div id="ref-Dalgaard2002" class="csl-entry" role="listitem">
Dalgaard, P. (2002). <em>Introductory statistics with <span>R</span></em>. Springer.
</div>
<div id="ref-DSP05" class="csl-entry" role="listitem">
Dasu, T., Swayne, D. F., &amp; Poole, D. (2005). Grouping <span>M</span>ultivariate <span>T</span>ime <span>S</span>eries: A <span>C</span>ase <span>S</span>tudy. <em>Proceedings of the IEEE Workshop on <span>T</span>emporal <span>D</span>ata <span>M</span>ining: <span>A</span>lgorithms, <span>T</span>heory and <span>A</span>pplications, in Conjunction with the Conference on Data Mining, Houston, November 27, 2005</em>, 25–32.
</div>
<div id="ref-R-ggdendro" class="csl-entry" role="listitem">
de Vries, A., &amp; Ripley, B. D. (2022). <em>Ggdendro: Create dendrograms and tree diagrams using ggplot2</em>. <a href="https://github.com/andrie/ggdendro">https://github.com/andrie/ggdendro</a>
</div>
<div id="ref-fireorigin" class="csl-entry" role="listitem">
Department of Environment, Land, Water &amp; Planning. (2019). <em><span class="nocase">Fire Origins - Current and Historical</span></em>. <a href="https://discover.data.vic.gov.au/dataset/fire-origins-current-and-historical">https://discover.data.vic.gov.au/dataset/fire-origins-current-and-historical</a>
</div>
<div id="ref-cfa" class="csl-entry" role="listitem">
Department of Environment, Land, Water &amp; Planning. (2020a). <em><span>CFA - Fire Station</span></em>. <a href="https://discover.data.vic.gov.au/dataset/cfa-fire-station-vmfeat-geomark_point">https://discover.data.vic.gov.au/dataset/cfa-fire-station-vmfeat-geomark_point</a>
</div>
<div id="ref-recreation" class="csl-entry" role="listitem">
Department of Environment, Land, Water &amp; Planning. (2020b). <em><span>Recreation Sites</span></em>. <a href="https://discover.data.vic.gov.au/dataset/recreation-sites">https://discover.data.vic.gov.au/dataset/recreation-sites</a>
</div>
<div id="ref-DF84" class="csl-entry" role="listitem">
Diaconis, P., &amp; Freedman, D. (1984a). <span>A</span>symptotics of <span>G</span>raphical <span>P</span>rojection <span>P</span>ursuit. <em>Annals of Statistics</em>, <em>12</em>, 793–815.
</div>
<div id="ref-diaconis1984" class="csl-entry" role="listitem">
Diaconis, P., &amp; Freedman, D. (1984b). Asymptotics of graphical projection pursuit. <em>Annals of Statistics</em>, <em>12</em>(3), 793–815. <a href="https://doi.org/10.1214/aos/1176346703">https://doi.org/10.1214/aos/1176346703</a>
</div>
<div id="ref-DMK05" class="csl-entry" role="listitem">
Dykes, J., MacEachren, A. M., &amp; Kraak, M.-J. (2005). <em>Exploring geovisualization</em>. Elsevier.
</div>
<div id="ref-Ev01" class="csl-entry" role="listitem">
Everitt, B. S., Landau, S., &amp; Leese, M. (2001). <em>Cluster <span>A</span>nalysis (4th ed)</em>. Edward Arnold.
</div>
<div id="ref-Fi79" class="csl-entry" role="listitem">
Fienberg, S. E. (1979). Graphical <span>M</span>ethods in <span>S</span>tatistics. <em>Journal of American Statistical Association</em>, <em>33</em>(4), 165–178.
</div>
<div id="ref-Fi36" class="csl-entry" role="listitem">
Fisher, R. A. (1936a). The <span>U</span>se of <span>M</span>ultiple <span>M</span>easurements in <span>T</span>axonomic <span>P</span>roblems. <em>Annals of Eugenics</em>, <em>7</em>, 179–188.
</div>
<div id="ref-fisher1936" class="csl-entry" role="listitem">
Fisher, R. A. (1936b). The use of multiple measurements in taxonomic problems. <em>Annals of Eugenics</em>, <em>7</em>(2), 179–188. <a href="https://doi.org/10.1111/j.1469-1809.1936.tb02137.x">https://doi.org/10.1111/j.1469-1809.1936.tb02137.x</a>
</div>
<div id="ref-Fi38" class="csl-entry" role="listitem">
Fisher, R. A. (1938). The <span>S</span>tatistical <span>U</span>tilization of <span>M</span>ultiple <span>M</span>easurements. <em>Annals of Eugenics</em>, <em>8</em>, 376–386.
</div>
<div id="ref-PRIM9-video" class="csl-entry" role="listitem">
Fisherkeller, M. A., Friedman, J. H., &amp; Tukey, J. W. (1973). <em><span>PRIM-9</span>, an interactive multidimensional data display and analysis system</em>. <a href="https://www.youtube.com/watch?v=B7XoW2qiFUA">https://www.youtube.com/watch?v=B7XoW2qiFUA</a>
</div>
<div id="ref-tukey" class="csl-entry" role="listitem">
Fisherkeller, M. A., Friedman, J. H., &amp; Tukey, J. W. (1974). <span>PRIM-9</span>, an interactive multidimensional data display and analysis system. In W. S. Cleveland (Ed.), <em>The collected works of john w. Tukey: Graphics 1965-1985, volume v</em> (pp. 340–346).
</div>
<div id="ref-eechidna" class="csl-entry" role="listitem">
Forbes, J., Cook, D., &amp; Hyndman, R. J. (2020). Spatial modelling of the two-party preferred vote in australian federal elections: 2001–2016. <em>Australian &amp; New Zealand Journal of Statistics</em>, <em>62</em>(2), 168–185. https://doi.org/<a href="https://doi.org/10.1111/anzs.12292">https://doi.org/10.1111/anzs.12292</a>
</div>
<div id="ref-Ford92" class="csl-entry" role="listitem">
Ford, B. J. (1992). <em>Images of science: A history of scientific illustration</em>. The British Library.
</div>
<div id="ref-forgy65" class="csl-entry" role="listitem">
Forgy, E. (1965). Cluster analysis of multivariate data: Efficiency versus interpretability of classification. <em>Biometrics</em>, <em>21</em>(3), 768–769.
</div>
<div id="ref-FR02" class="csl-entry" role="listitem">
Fraley, C., &amp; Raftery, A. E. (2002). Model-based <span>C</span>lustering, <span>D</span>iscriminant <span>A</span>nalysis, <span>D</span>ensity <span>E</span>stimation. <em>Journal of the American Statistical Association</em>, <em>97</em>, 611–631.
</div>
<div id="ref-R-mclust" class="csl-entry" role="listitem">
Fraley, C., Raftery, A. E., &amp; Scrucca, L. (2022). <em>Mclust: Gaussian mixture modelling for model-based clustering, classification, and density estimation</em>. <a href="https://mclust-org.github.io/mclust/">https://mclust-org.github.io/mclust/</a>
</div>
<div id="ref-Fr87" class="csl-entry" role="listitem">
Friedman, J. H. (1987). <span>E</span>xploratory <span>P</span>rojection <span>P</span>ursuit. <em>Journal of American Statistical Association</em>, <em>82</em>, 249–266.
</div>
<div id="ref-FT74" class="csl-entry" role="listitem">
Friedman, J. H., &amp; Tukey, J. W. (1974). <span>A</span> <span>P</span>rojection <span>P</span>ursuit <span>A</span>lgorithm for <span>E</span>xploratory <span>D</span>ata <span>A</span>nalysis. <em>IEEE Transactions on Computing C</em>, <em>23</em>, 881–889.
</div>
<div id="ref-FD04" class="csl-entry" role="listitem">
Friendly, M., &amp; Denis, D. J. (2004). <em>Milestones in the history of thematic cartography, statistical graphics, and data visualization</em>. http://www.math.yorku.ca/SCS/Gallery/milestone/.
</div>
<div id="ref-neuralnet" class="csl-entry" role="listitem">
Fritsch, S., Guenther, F., &amp; Wright, M. N. (2019). <em>Neuralnet: Training of neural networks</em>. <a href="https://CRAN.R-project.org/package=neuralnet">https://CRAN.R-project.org/package=neuralnet</a>
</div>
<div id="ref-FB94" class="csl-entry" role="listitem">
Furnas, G. W., &amp; Buja, A. (1994). Prosection <span>V</span>iews: <span>D</span>imensional <span>I</span>nference <span>T</span>hrough <span>S</span>ections and <span>P</span>rojections. <em>Journal of Computational and Graphical Statistics</em>, <em>3</em>(4), 323–385.
</div>
<div id="ref-Ga71" class="csl-entry" role="listitem">
Gabriel, K. R. (1971). The <span>B</span>iplot <span>G</span>raphical <span>D</span>isplay of <span>M</span>atrices with <span>A</span>pplications to <span>P</span>rincipal <span>C</span>omponent <span>A</span>nalysis. <em>Biometrika</em>, <em>58</em>, 453–467.
</div>
<div id="ref-Gentle04" class="csl-entry" role="listitem">
Gentle, J. E., Härdle, W., &amp; Mori, Y. (Eds.). (2004). <em>Handbook of computational statistics: Concepts and methods</em>. Springer.
</div>
<div id="ref-giordani" class="csl-entry" role="listitem">
Giordani, P., Ferraro, M. B., &amp; Martella, F. (2020). <em>An introduction to clustering with r</em>. Springer Singapore. <a href="https://doi.org/10.1007/978-981-13-0553-5">https://doi.org/10.1007/978-981-13-0553-5</a>
</div>
<div id="ref-GH92" class="csl-entry" role="listitem">
Glover, D. M., &amp; Hopke, P. K. (1992). <span>E</span>xploration of <span>M</span>ultivariate <span>C</span>hemical <span>D</span>ata by <span>P</span>rojection <span>P</span>ursuit. <em>Chemometrics and Intelligent Laboratory Systems</em>, <em>16</em>, 45–59.
</div>
<div id="ref-Good05" class="csl-entry" role="listitem">
Good, P. (2005). <em>Permutation, <span>P</span>arametric, and <span>B</span>ootstrap <span>T</span>ests of <span>H</span>ypotheses</em>. Springer.
</div>
<div id="ref-GH96" class="csl-entry" role="listitem">
Gower, J. C., &amp; Hand, D. J. (1996). <em>Biplots</em>. Chapman; Hall.
</div>
<div id="ref-risk-survey" class="csl-entry" role="listitem">
Hajibaba, H., Karlsson, L., &amp; Dolnicar, S. (2016). Residents open their homes to tourists when disaster strikes. <em>Journal of Travel Research</em>, <em>56</em>(8), 1065–1078.
</div>
<div id="ref-Hansen2004" class="csl-entry" role="listitem">
Hansen, C., &amp; Johnson, C. R. (2004). <em>Visualization handbook</em>. Academic Press.
</div>
<div id="ref-R-langevitour" class="csl-entry" role="listitem">
Harrison, P. (2023a). <em>Langevitour: Langevin tour</em>. <a href="https://logarithmic.net/langevitour/">https://logarithmic.net/langevitour/</a>
</div>
<div id="ref-langevitour2023" class="csl-entry" role="listitem">
Harrison, P. (2023b). Langevitour: Smooth interactive touring of high dimensions, demonstrated with scRNA-seq data. <em>The R Journal</em>, <em>15</em>(2), 206–219. <a href="https://doi.org/10.32614/RJ-2023-046">https://doi.org/10.32614/RJ-2023-046</a>
</div>
<div id="ref-R-detourr" class="csl-entry" role="listitem">
Hart, C., &amp; Wang, E. (2022). <em>Detourr: Portable and performant tour animations</em>. <a href="https://casperhart.github.io/detourr/">https://casperhart.github.io/detourr/</a>
</div>
<div id="ref-HK81" class="csl-entry" role="listitem">
Hartigan, J. A., &amp; Kleiner, B. (1981). Mosaics for <span>C</span>ontingency <span>T</span>ables. <em>Computer Science and Statistics: Proceedings of the 13th Symposium on the Interface</em>, 268–273.
</div>
<div id="ref-HK84" class="csl-entry" role="listitem">
Hartigan, J., &amp; Kleiner, B. (1984). A <span>M</span>osaic of <span>T</span>elevision <span>R</span>atings. <em>The American Statistician</em>, <em>38</em>, 32–35.
</div>
<div id="ref-HBCUW91" class="csl-entry" role="listitem">
Haslett, J., Bradley, R., Craig, P., Unwin, A., &amp; Wills, G. (1991). <span>D</span>ynamic <span>G</span>raphics for <span>E</span>xploring <span>S</span>patial <span>D</span>ata with <span>A</span>pplication to <span>L</span>ocating <span>G</span>lobal and <span>L</span>ocal <span>A</span>nomalies. <em>The American Statistician</em>, <em>45</em>(3), 234–242.
</div>
<div id="ref-HTF01" class="csl-entry" role="listitem">
Hastie, T., Tibshirani, R., &amp; Friedman, J. (2001). <em>The <span>E</span>lements of <span>S</span>tatistical <span>L</span>earning</em>. Springer.
</div>
<div id="ref-hennig" class="csl-entry" role="listitem">
Hennig, C., Meila, M., Murtagh, F., &amp; Rocci, R. (Eds.). (2015). <em>Handbook of cluster analysis</em>. Chapman; Hall/<span>CRC</span>. <a href="https://doi.org/10.1201/b19706">https://doi.org/10.1201/b19706</a>
</div>
<div id="ref-Ho01" class="csl-entry" role="listitem">
Hofmann, H. (2001). <em>Graphical <span>T</span>ools for the <span>E</span>xploration of <span>M</span>ultivariate <span>C</span>ategorical <span>D</span>ata</em>. Books on Demand.
</div>
<div id="ref-Ho03" class="csl-entry" role="listitem">
Hofmann, H. (2003). Constructing and <span>R</span>eading <span>M</span>osaicplots. <em>Computational Statistics and Data Analysis</em>, <em>43</em>(4), 565–580.
</div>
<div id="ref-HT98" class="csl-entry" role="listitem">
Hofmann, H., &amp; Theus, M. (1998). Selection <span>S</span>equences in <span>MANET</span>. <em>Computational Statistics</em>, <em>13</em>(1), 77–87.
</div>
<div id="ref-ggfortify2018" class="csl-entry" role="listitem">
Horikoshi, M., &amp; Tang, Y. (2018). <em>Ggfortify: Data visualization tools for statistical analysis results</em>. <a href="https://CRAN.R-project.org/package=ggfortify">https://CRAN.R-project.org/package=ggfortify</a>
</div>
<div id="ref-R-ggfortify" class="csl-entry" role="listitem">
Horikoshi, M., &amp; Tang, Y. (2023). <em>Ggfortify: Data visualization tools for statistical analysis results</em>. <a href="https://github.com/sinhrks/ggfortify">https://github.com/sinhrks/ggfortify</a>
</div>
<div id="ref-R-palmerpenguins" class="csl-entry" role="listitem">
Horst, A., Hill, A., &amp; Gorman, K. (2022). <em>Palmerpenguins: Palmer archipelago (antarctica) penguin data</em>. <a href="https://CRAN.R-project.org/package=palmerpenguins">https://CRAN.R-project.org/package=palmerpenguins</a>
</div>
<div id="ref-hotelling-pca" class="csl-entry" role="listitem">
Hotelling, H. (1933). Analysis of a complex of statistical variables into principal components. <em>Journal of Educational Psychology</em>, <em>24</em>(6), 417--441. <a href="https://doi.org/10.1037/h0071325">https://doi.org/10.1037/h0071325</a>
</div>
<div id="ref-Hu85" class="csl-entry" role="listitem">
Huber, P. J. (1985). <span>P</span>rojection <span>P</span>ursuit (with discussion). <em>Annals of Statistics</em>, <em>13</em>, 435–525.
</div>
<div id="ref-Hurley87" class="csl-entry" role="listitem">
Hurley, C. (1987). <em>The data viewer: An interactive program for data analysis</em> [PhD thesis]. University of Washington.
</div>
<div id="ref-R-gt" class="csl-entry" role="listitem">
Iannone, R., Cheng, J., Schloerke, B., Hughes, E., Lauer, A., &amp; Seo, J. (2023). <em>Gt: Easily create presentation-ready display tables</em>. <a href="https://CRAN.R-project.org/package=gt">https://CRAN.R-project.org/package=gt</a>
</div>
<div id="ref-R1996" class="csl-entry" role="listitem">
Ihaka, R., &amp; Gentleman, R. (1996). R: <span>A</span> <span>L</span>anguage for <span>D</span>ata <span>A</span>nalysis and <span>G</span>raphics. <em>Journal of Computational and Graphical Statistics</em>, <em>5</em>, 299–314.
</div>
<div id="ref-R-colorspace" class="csl-entry" role="listitem">
Ihaka, R., Murrell, P., Hornik, K., Fisher, J. C., Stauffer, R., Wilke, C. O., McWhite, C. D., &amp; Zeileis, A. (2023). <em>Colorspace: A toolbox for manipulating and assessing colors and palettes</em>. <a href="https://CRAN.R-project.org/package=colorspace">https://CRAN.R-project.org/package=colorspace</a>
</div>
<div id="ref-In85" class="csl-entry" role="listitem">
Inselberg, A. (1985). <span>T</span>he <span>P</span>lane with <span>P</span>arallel <span>C</span>oordinates. <em>The Visual Computer</em>, <em>1</em>, 69–91.
</div>
<div id="ref-windasos" class="csl-entry" role="listitem">
Iowa State University. (2020). <em>ASOS-AWOS-METAR data download</em>. <a href="https://mesonet.agron.iastate.edu/request/download.phtml?network=AU__ASOS">https://mesonet.agron.iastate.edu/request/download.phtml?network=AU__ASOS</a>
</div>
<div id="ref-JT07" class="csl-entry" role="listitem">
Johnson, D., &amp; Travis, J. (2007). <em>Flatland: The movie</em>. https://round-drum-w7xh.squarespace.com/our-story.
</div>
<div id="ref-JW02" class="csl-entry" role="listitem">
Johnson, R. A., &amp; Wichern, D. W. (2002). <em>Applied multivariate statistical analysis (5th ed)</em>. Prentice-Hall.
</div>
<div id="ref-joliffe2016" class="csl-entry" role="listitem">
Jolliffe, I. T., &amp; Cadima, J. (2016). Principal component analysis: A review and recent developments. <em>Phil. Trans. R. Soc. A.</em>, <em>374</em>, 20150202. <a href="https://doi.org/10.1098/rsta.2015.0202">https://doi.org/10.1098/rsta.2015.0202</a>
</div>
<div id="ref-JS87" class="csl-entry" role="listitem">
Jones, M. C., &amp; Sibson, R. (1987). <span>W</span>hat is <span>P</span>rojection <span>P</span>ursuit? (With discussion). <em>Journal of the Royal Statistical Society, Series A</em>, <em>150</em>, 1–36.
</div>
<div id="ref-kassambara" class="csl-entry" role="listitem">
Kassambara, A. (2017). <em>Practical guide to cluster analysis in r: Unsupervised machine learning</em>. STHDA.
</div>
<div id="ref-R-ggpubr" class="csl-entry" role="listitem">
Kassambara, A. (2023). <em>Ggpubr: ggplot2 based publication ready plots</em>. <a href="https://rpkgs.datanovia.com/ggpubr/">https://rpkgs.datanovia.com/ggpubr/</a>
</div>
<div id="ref-Ko01" class="csl-entry" role="listitem">
Kohonen, T. (2001). <em>Self-<span>O</span>rganizing <span>M</span>aps (3rd ed)</em>. Springer.
</div>
<div id="ref-KS96" class="csl-entry" role="listitem">
Koschat, M. A., &amp; Swayne, D. F. (1996). Interactive <span>G</span>raphical <span>M</span>ethods in the <span>A</span>nalysis of <span>C</span>ustomer <span>P</span>anel <span>D</span>ata (with discussion). <em>Journal of Business and Economic Statistics</em>, <em>14</em>(1), 113–132.
</div>
<div id="ref-R-Rtsne" class="csl-entry" role="listitem">
Krijthe, J. (2022). <em>Rtsne: T-distributed stochastic neighbor embedding using a barnes-hut implementation</em>. <a href="https://github.com/jkrijthe/Rtsne">https://github.com/jkrijthe/Rtsne</a>
</div>
<div id="ref-Kr64a" class="csl-entry" role="listitem">
Kruskal, J. B. (1964a). Multidimensional <span>S</span>caling by <span>O</span>ptimizing <span>G</span>oodness of <span>F</span>it to a <span>N</span>onmetric <span>H</span>ypothesis. <em>Psychometrika</em>, <em>29</em>, 1–27.
</div>
<div id="ref-Kr64b" class="csl-entry" role="listitem">
Kruskal, J. B. (1964b). Nonmetric <span>M</span>ultidimensional <span>S</span>caling: A <span>N</span>umerical <span>M</span>ethod. <em>Psychometrika</em>, <em>29</em>, 115–129.
</div>
<div id="ref-KW78" class="csl-entry" role="listitem">
Kruskal, J. B., &amp; Wish, M. (1978). <em>Multidimensional <span>S</span>caling</em>. Sage Publications.
</div>
<div id="ref-tidymodels2020" class="csl-entry" role="listitem">
Kuhn, M., &amp; Wickham, H. (2020). <em>Tidymodels: A collection of packages for modeling and machine learning using tidyverse principles.</em> <a href="https://www.tidymodels.org">https://www.tidymodels.org</a>
</div>
<div id="ref-R-tidymodels" class="csl-entry" role="listitem">
Kuhn, M., &amp; Wickham, H. (2023). <em>Tidymodels: Easily install and load the tidymodels packages</em>. <a href="https://CRAN.R-project.org/package=tidymodels">https://CRAN.R-project.org/package=tidymodels</a>
</div>
<div id="ref-sagetour" class="csl-entry" role="listitem">
Laa, U., Cook, D., &amp; Lee, S. (2022). Burning sage: Reversing the curse of dimensionality in the visualization of high-dimensional data. <em>Journal of Computational and Graphical Statistics</em>, <em>31</em>(1), 40–49. <a href="https://doi.org/10.1080/10618600.2021.1963264">https://doi.org/10.1080/10618600.2021.1963264</a>
</div>
<div id="ref-slicetour" class="csl-entry" role="listitem">
Laa, U., Cook, D., &amp; Valencia, G. (2020a). A slice tour for finding hollowness in high-dimensional data. <em>Journal of Computational and Graphical Statistics</em>, <em>29</em>(3), 681–687. <a href="https://doi.org/10.1080/10618600.2020.1777140">https://doi.org/10.1080/10618600.2020.1777140</a>
</div>
<div id="ref-tourr2020" class="csl-entry" role="listitem">
Laa, U., Cook, D., &amp; Valencia, G. (2020b). A slice tour for finding hollowness in high-dimensional data. <em>Journal of Computational and Graphical Statistics</em>, <em>29</em>(3), 681–687. <a href="https://doi.org/10.1080/10618600.2020.1777140">https://doi.org/10.1080/10618600.2020.1777140</a>
</div>
<div id="ref-helmert" class="csl-entry" role="listitem">
Lancaster, H. O. (1965). The helmert matrices. <em>The American Mathematical Monthly</em>, <em>72</em>(1), 4–12.
</div>
<div id="ref-R-cxhull" class="csl-entry" role="listitem">
Laurent, S. (2023). <em>Cxhull: Convex hull</em>. <a href="https://github.com/stla/cxhull">https://github.com/stla/cxhull</a>
</div>
<div id="ref-Lee2018" class="csl-entry" role="listitem">
Lee, E.-K. (2018). PPtreeViz: An r package for visualizing projection pursuit classification trees. <em>Journal of Statistical Software</em>, <em>83</em>(8), 1–30. <a href="https://doi.org/10.18637/jss.v083.i08">https://doi.org/10.18637/jss.v083.i08</a>
</div>
<div id="ref-Lee2009" class="csl-entry" role="listitem">
Lee, E.-K., &amp; Cook, D. (2009). A projection pursuit index for large <span class="math inline">\(p\)</span> small <span class="math inline">\(n\)</span> data. <em>Statistics and Computing</em>, <em>20</em>, 381–392. <a href="https://doi.org/10.1007/s11222-009-9131-1">https://doi.org/10.1007/s11222-009-9131-1</a>
</div>
<div id="ref-LCKL05" class="csl-entry" role="listitem">
Lee, E.-K., Cook, D., Klinke, S., &amp; Lumley, T. (2005). <span>P</span>rojection <span>P</span>ursuit for <span>E</span>xploratory <span>S</span>upervised <span>C</span>lassification. <em>Journal of Computational and Graphical Statistics</em>, <em>14</em>(4), 831–846.
</div>
<div id="ref-R-liminal" class="csl-entry" role="listitem">
Lee, S. (2021). <em>Liminal: Multivariate data visualization with tours and embeddings</em>. <a href="https://CRAN.R-project.org/package=liminal">https://CRAN.R-project.org/package=liminal</a>
</div>
<div id="ref-tours2022" class="csl-entry" role="listitem">
Lee, S., Cook, D., Silva, N. da, Laa, U., Spyrison, N., Wang, E., &amp; Zhang, H. S. (2022). The state-of-the-art on tours for dynamic visualization of high-dimensional data. <em>WIREs Computational Statistics</em>, <em>14</em>(4), e1573. <a href="https://doi.org/10.1002/wics.1573">https://doi.org/10.1002/wics.1573</a>
</div>
<div id="ref-Lee2013" class="csl-entry" role="listitem">
Lee, Y. D., Cook, D., Park, J., &amp; Lee, E.-K. (2013). <span class="nocase">PPtree: Projection pursuit classification tree</span>. <em>Electronic Journal of Statistics</em>, <em>7</em>(none), 1369–1386. <a href="https://doi.org/10.1214/13-EJS810">https://doi.org/10.1214/13-EJS810</a>
</div>
<div id="ref-ctv-clustering" class="csl-entry" role="listitem">
Leisch, F., &amp; Gruen, B. (2023). <em>CRAN task view: Cluster analysis &amp; finite mixture models</em>. https://cran.r-project.org/web/views/Cluster.html.
</div>
<div id="ref-R-MSA" class="csl-entry" role="listitem">
Leisch, F., &amp; Grün, B. (2020). <em>MSA: Market segmentation analysis</em>.
</div>
<div id="ref-li2020visualizing" class="csl-entry" role="listitem">
Li, M., Zhao, Z., &amp; Scheidegger, C. (2020). Visualizing neural networks with the grand tour. <em>Distill</em>. <a href="https://doi.org/10.23915/distill.00025">https://doi.org/10.23915/distill.00025</a>
</div>
<div id="ref-randomForest2002" class="csl-entry" role="listitem">
Liaw, A., &amp; Wiener, M. (2002). Classification and regression by randomForest. <em>R News</em>, <em>2</em>(3), 18–22. <a href="https://CRAN.R-project.org/doc/Rnews/">https://CRAN.R-project.org/doc/Rnews/</a>
</div>
<div id="ref-LSDB92" class="csl-entry" role="listitem">
Littman, M. L., Swayne, D. F., Dean, N., &amp; Buja, A. (1992). Visualizing the <span>E</span>mbedding of <span>O</span>bjects in <span>E</span>uclidean <span>S</span>pace. <em>Computing Science and Statistics: Proceedings of the 24th Symposium on the Interface</em>, 208–217.
</div>
<div id="ref-Lloyd1982" class="csl-entry" role="listitem">
Lloyd, S. (1982). Least squares quantization in PCM. <em>IEEE Transactions on Information Theory</em>, <em>28</em>(2), 129–137. <a href="https://doi.org/10.1109/TIT.1982.1056489">https://doi.org/10.1109/TIT.1982.1056489</a>
</div>
<div id="ref-LMGR05" class="csl-entry" role="listitem">
Longley, P. A., Maguire, D. J., Goodchild, M. F., &amp; Rhind, D. W. (2005). <em>Geographic information systems and science</em>. John Wiley &amp; Sons.
</div>
<div id="ref-loperfido" class="csl-entry" role="listitem">
Loperfido, N. (2018). Skewness-based projection pursuit: A computational approach. <em>Computational Statistics &amp; Data Analysis</em>, <em>120</em>, 42–57. https://doi.org/<a href="https://doi.org/10.1016/j.csda.2017.11.001">https://doi.org/10.1016/j.csda.2017.11.001</a>
</div>
<div id="ref-Maaten2008" class="csl-entry" role="listitem">
Maaten, L. van der, &amp; Hinton, G. (2008). Visualizing data using <span class="nocase">t-SNE</span>. <em>J. Mach. Learn. Res.</em>, <em>9</em>(Nov), 2579–2605. <a href="http://www.jmlr.org/papers/v9/vandermaaten08a.html">http://www.jmlr.org/papers/v9/vandermaaten08a.html</a>
</div>
<div id="ref-MacQueen1967" class="csl-entry" role="listitem">
MacQueen, J. B. (1967). Some methods for classification and analysis of multivariate observations. In L. M. L. Cam &amp; J. Neyman (Eds.), <em>Proc. Of the fifth berkeley symposium on mathematical statistics and probability</em> (Vol. 1, pp. 281–297). University of California Press.
</div>
<div id="ref-MB03" class="csl-entry" role="listitem">
Maindonald, J., &amp; Braun, J. (2003). <em>Data analysis and graphics using r - an example-based approach</em>. Cambridge University Press.
</div>
<div id="ref-Ma65" class="csl-entry" role="listitem">
Martin, E. (1965). <em>Flatland</em>. http://www.der.org/films/flatland.html.
</div>
<div id="ref-kernelshap" class="csl-entry" role="listitem">
Mayer, M., &amp; Watson, D. (2023). <em>Kernelshap: Kernel SHAP</em>. <a href="https://CRAN.R-project.org/package=kernelshap">https://CRAN.R-project.org/package=kernelshap</a>
</div>
<div id="ref-MY94" class="csl-entry" role="listitem">
McFarlane, M., &amp; Young, F. W. (1994). Graphical <span>S</span>ensitivity <span>A</span>nalysis for <span>M</span>ultidimensional <span>S</span>caling. <em>Journal of Computational and Graphical Statistics</em>, <em>3</em>, 23–33.
</div>
<div id="ref-McInnes2018" class="csl-entry" role="listitem">
McInnes, L., Healy, J., &amp; Melville, J. (2018). <em><span>UMAP</span>: Uniform manifold approximation and projection for dimension reduction</em>. <a href="http://arxiv.org/abs/1802.03426">http://arxiv.org/abs/1802.03426</a>
</div>
<div id="ref-Mc77" class="csl-entry" role="listitem">
McNeil, D. (1977). <em>Interactive <span>D</span>ata <span>A</span>nalysis</em>. John Wiley &amp; Sons.
</div>
<div id="ref-wind" class="csl-entry" role="listitem">
McVicar, T. (2011). <em>Near-surface wind speed. v10. CSIRO. Data collection.</em> <a href="https://doi.org/10.25919/5c5106acbcb02">https://doi.org/10.25919/5c5106acbcb02</a>
</div>
<div id="ref-R-e1071" class="csl-entry" role="listitem">
Meyer, D., Dimitriadou, E., Hornik, K., Weingessel, A., &amp; Leisch, F. (2023). <em>e1071: Misc functions of the department of statistics, probability theory group (formerly: E1071), TU wien</em>. <a href="https://CRAN.R-project.org/package=e1071">https://CRAN.R-project.org/package=e1071</a>
</div>
<div id="ref-R-rpart.plot" class="csl-entry" role="listitem">
Milborrow, S. (2022). <em>Rpart.plot: Plot rpart models: An enhanced version of plot.rpart</em>. <a href="http://www.milbo.org/rpart-plot/index.html">http://www.milbo.org/rpart-plot/index.html</a>
</div>
<div id="ref-R-gtExtras" class="csl-entry" role="listitem">
Mock, T. (2022). <em>gtExtras: Extending gt for beautiful HTML tables</em>. <a href="https://CRAN.R-project.org/package=gtExtras">https://CRAN.R-project.org/package=gtExtras</a>
</div>
<div id="ref-iml" class="csl-entry" role="listitem">
Molnar, C. (2022). <em>Interpretable machine learning: A guide for making black box models explainable (2nd ed)</em>. https://christophm.github.io/interpretable-ml-book/.
</div>
<div id="ref-Moon2019" class="csl-entry" role="listitem">
Moon, K. R., Dijk, D. van, Wang, Z., Gigante, S., Burkhardt, D. B., Chen, W. S., Yim, K., Elzen, A. van den, Hirn, M. J., Coifman, R. R., Ivanova, N. B., Wolf, G., &amp; Krishnaswamy, S. (2019). Visualizing structure and transitions for biological data exploration. <em>Nature Biotechnology</em>, <em>37</em>, 1482–1492. <a href="https://doi.org/10.1038/s41587-019-0336-3">https://doi.org/10.1038/s41587-019-0336-3</a>
</div>
<div id="ref-Murrell2005" class="csl-entry" role="listitem">
Murrell, P. (2005). <em>R graphics</em>. Chapman &amp; Hall/CRC.
</div>
<div id="ref-OpenStreetMap" class="csl-entry" role="listitem">
OpenStreetMap contributors. (2020). <em><span class="nocase">Planet dump retrieved from https://planet.osm.org </span></em>.<a href="https://www.openstreetmap.org%20" class="uri"> https://www.openstreetmap.org</a>.
</div>
<div id="ref-pearson-pca" class="csl-entry" role="listitem">
Pearson, K. (1901). LIII. On lines and planes of closest fit to systems of points in space. <em>The London, Edinburgh, and Dublin Philosophical Magazine and Journal of Science</em>, <em>2</em>(11), 559–572. <a href="https://doi.org/10.1080/14786440109462720">https://doi.org/10.1080/14786440109462720</a>
</div>
<div id="ref-R-patchwork" class="csl-entry" role="listitem">
Pedersen, T. L. (2023). <em>Patchwork: The composer of plots</em>. <a href="https://CRAN.R-project.org/package=patchwork">https://CRAN.R-project.org/package=patchwork</a>
</div>
<div id="ref-perisic2009" class="csl-entry" role="listitem">
Perisic, I., &amp; Posse, C. (2005). Projection pursuit indices based on the empirical distribution function. <em>Journal of Computational and Graphical Statistics</em>, <em>14</em>(3), 700–715. <a href="https://doi.org/10.1198/106186005X69440">https://doi.org/10.1198/106186005X69440</a>
</div>
<div id="ref-Po95" class="csl-entry" role="listitem">
Polzehl, J. (1995). Projection <span>P</span>ursuit <span>D</span>iscriminant <span>A</span>nalysis. <em>Computational Statistics and Data Analysis</em>, <em>20</em>, 141–157.
</div>
<div id="ref-Po92" class="csl-entry" role="listitem">
Posse, C. (1992). Projection <span>P</span>ursuit <span>D</span>iscriminant <span>A</span>nalysis for <span>T</span>wo <span>G</span>roups. <em>Communications in Statistics, Part A – Theory and Methods</em>, <em>21</em>, 1–19.
</div>
<div id="ref-Po94" class="csl-entry" role="listitem">
Posse, C. (1995). Tools for <span>T</span>wo-dimensional <span>P</span>rojection <span>P</span>ursuit. <em>Journal of Computational and Graphical Statistics</em>, <em>4</em>(2), 83–100.
</div>
<div id="ref-jaxa" class="csl-entry" role="listitem">
P-Tree System. (2020). <em><span class="nocase">JAXA Himawari Monitor - User’s Guide</span></em>. <a href="https://www.eorc.jaxa.jp/ptree/userguide.html">https://www.eorc.jaxa.jp/ptree/userguide.html</a>
</div>
<div id="ref-R" class="csl-entry" role="listitem">
R Core Team. (2023). <em>R: A language and environment for statistical computing</em>. R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>
</div>
<div id="ref-Rao48" class="csl-entry" role="listitem">
Rao, C. R. (1948). The <span>U</span>tilization of <span>M</span>ultiple <span>M</span>easurements in <span>P</span>roblems of <span>B</span>iological <span>C</span>lassification (with discussion). <em>Journal of the Royal Statistical Society, Series B</em>, <em>10</em>, 159–203.
</div>
<div id="ref-Rao93" class="csl-entry" role="listitem">
Rao, C. R. (Ed.). (1993). <em><span>H</span>andbook of <span>S</span>tatistics, <span>V</span>ol.&nbsp;9</em>. Elsevier Science Publishers.
</div>
<div id="ref-W05" class="csl-entry" role="listitem">
Rao, C. R., Wegman, E. J., &amp; Solka, J. L. (Eds.). (2006). <em>Handbook of <span>S</span>tatistics: <span>D</span>ata <span>M</span>ining and <span>V</span>isualization</em>. Elsevier/North-Holland.
</div>
<div id="ref-Ri96" class="csl-entry" role="listitem">
Ripley, B. (1996). <em>Pattern <span>R</span>ecognition and <span>N</span>eural <span>N</span>etworks</em>. Cambridge University Press.
</div>
<div id="ref-R-MASS" class="csl-entry" role="listitem">
Ripley, B. (2023a). <em>MASS: Support functions and datasets for venables and ripley’s MASS</em>. <a href="http://www.stats.ox.ac.uk/pub/MASS4/">http://www.stats.ox.ac.uk/pub/MASS4/</a>
</div>
<div id="ref-R-nnet" class="csl-entry" role="listitem">
Ripley, B. (2023b). <em>Nnet: Feed-forward neural networks and multinomial log-linear models</em>. <a href="http://www.stats.ox.ac.uk/pub/MASS4/">http://www.stats.ox.ac.uk/pub/MASS4/</a>
</div>
<div id="ref-Ro1957" class="csl-entry" role="listitem">
Rothkopf, E. Z. (1957). A <span>M</span>easure of <span>S</span>timulus <span>S</span>imilarity and <span>E</span>rrors in <span>S</span>ome <span>P</span>aired-associate <span>L</span>earning <span>T</span>asks. <em>Journal of Experimental Psychology</em>, <em>53</em>, 94–101.
</div>
<div id="ref-R-geozoo" class="csl-entry" role="listitem">
Schloerke, B. (2016). <em>Geozoo: Zoo of geometric objects</em>. <a href="https://CRAN.R-project.org/package=geozoo">https://CRAN.R-project.org/package=geozoo</a>
</div>
<div id="ref-R-GGally" class="csl-entry" role="listitem">
Schloerke, B., Cook, D., Larmarange, J., Briatte, F., Marbach, M., Thoen, E., Elberg, A., &amp; Crowley, J. (2023). <em>GGally: Extension to ggplot2</em>. <a href="https://ggobi.github.io/ggally/">https://ggobi.github.io/ggally/</a>
</div>
<div id="ref-mclust2016" class="csl-entry" role="listitem">
Scrucca, L., Fop, M., Murphy, T. B., &amp; Raftery, A. E. (2016). <span class="nocase">mclust</span> 5: Clustering, classification and density estimation using <span>G</span>aussian finite mixture models. <em>The <span>R</span> Journal</em>, <em>8</em>(1), 289–317. <a href="https://doi.org/10.32614/RJ-2016-021">https://doi.org/10.32614/RJ-2016-021</a>
</div>
<div id="ref-Sh62" class="csl-entry" role="listitem">
Shepard, R. N. (1962). The <span>A</span>nalysis of <span>P</span>roximities: <span>M</span>ultidimensional <span>S</span>caling with an <span>U</span>nknown <span>D</span>istance <span>F</span>unction, <span>I</span> and <span>II</span>. <em>Psychometrika</em>, <em>27</em>, 125-139 and 219-246.
</div>
<div id="ref-plotly2020" class="csl-entry" role="listitem">
Sievert, C. (2020). <em>Interactive web-based data visualization with r, plotly, and shiny</em>. Chapman; Hall/CRC. <a href="https://plotly-r.com">https://plotly-r.com</a>
</div>
<div id="ref-R-plotly" class="csl-entry" role="listitem">
Sievert, C., Parmer, C., Hocking, T., Chamberlain, S., Ram, K., Corvellec, M., &amp; Despouy, P. (2023). <em>Plotly: Create interactive web graphics via plotly.js</em>. <a href="https://CRAN.R-project.org/package=plotly">https://CRAN.R-project.org/package=plotly</a>
</div>
<div id="ref-R-gtsummary" class="csl-entry" role="listitem">
Sjoberg, D. D., Larmarange, J., Curry, M., Lavery, J., Whiting, K., &amp; Zabor, E. C. (2023). <em>Gtsummary: Presentation-ready data summary and analytic result tables</em>. <a href="https://CRAN.R-project.org/package=gtsummary">https://CRAN.R-project.org/package=gtsummary</a>
</div>
<div id="ref-gtsummary" class="csl-entry" role="listitem">
Sjoberg, D. D., Whiting, K., Curry, M., Lavery, J. A., &amp; Larmarange, J. (2021). Reproducible summary tables with the gtsummary package. <em><span>The R Journal</span></em>, <em>13</em>, 570–580. <a href="https://doi.org/10.32614/RJ-2021-053">https://doi.org/10.32614/RJ-2021-053</a>
</div>
<div id="ref-R-ggrepel" class="csl-entry" role="listitem">
Slowikowski, K. (2023). <em>Ggrepel: Automatically position non-overlapping text labels with ggplot2</em>. <a href="https://github.com/slowkow/ggrepel">https://github.com/slowkow/ggrepel</a>
</div>
<div id="ref-R-bomrang" class="csl-entry" role="listitem">
Sparks, A. H., Carroll, J., Goldie, J., Marchiori, D., Melloy, P., Padgham, M., Parsonage, H., &amp; Pembleton, K. (2020). <em><span class="nocase">bomrang</span>: Australian government bureau of meteorology (BOM) data client</em>. <a href="https://CRAN.R-project.org/package=bomrang">https://CRAN.R-project.org/package=bomrang</a>
</div>
<div id="ref-Spence07" class="csl-entry" role="listitem">
Spence, R. (2007). <em>Information visualization: Design for interaction</em>. Prentice Hall.
</div>
<div id="ref-colorspace2009b" class="csl-entry" role="listitem">
Stauffer, R., Mayr, G. J., Dabernig, M., &amp; Zeileis, A. (2009). Somewhere over the rainbow: How to make effective use of colors in meteorological visualizations. <em>Bulletin of the American Meteorological Society</em>, <em>96</em>(2), 203–216. <a href="https://doi.org/10.1175/BAMS-D-13-00155.1">https://doi.org/10.1175/BAMS-D-13-00155.1</a>
</div>
<div id="ref-SRLLDCC00" class="csl-entry" role="listitem">
Sutherland, P., Rossini, A., Lumley, T., Lewin-Koh, N., Dickerson, J., Cox, Z., &amp; Cook, D. (2000a). Orca: <span>A</span> <span>V</span>isualization <span>T</span>oolkit for <span>H</span>igh-<span>D</span>imensional <span>D</span>ata. <em>Journal of Computational and Graphical Statistics</em>, <em>9</em>(3), 509–529.
</div>
<div id="ref-orca" class="csl-entry" role="listitem">
Sutherland, P., Rossini, A., Lumley, T., Lewin-Koh, N., Dickerson, J., Cox, Z., &amp; Cook, D. (2000b). Orca: A visualization toolkit for high-dimensional data. <em>Journal of Computational and Graphical Statistics</em>, <em>9</em>(3), 509–529. <a href="https://doi.org/10.1080/10618600.2000.10474896">https://doi.org/10.1080/10618600.2000.10474896</a>
</div>
<div id="ref-SBTL04" class="csl-entry" role="listitem">
Swayne, D. F., Buja, A., &amp; Temple Lang, D. (2004). Exploratory visual analysis of graphs in <span>GG</span>obi. In J. Antoch (Ed.), <em>CompStat: Proceedings in computational statistics, 16th symposium</em>. Physica-Verlag.
</div>
<div id="ref-SCB91" class="csl-entry" role="listitem">
Swayne, D. F., Cook, D., &amp; Buja, A. (1992). <span>XG</span>obi: <span>I</span>nteractive <span>D</span>ynamic <span>G</span>raphics in the <span>X</span> <span>W</span>indow <span>S</span>ystem with a <span>L</span>ink to <span>S</span>. <em>American Statistical Association 1991 Proceedings of the Section on Statistical Graphics</em>, 1–8.
</div>
<div id="ref-xgobi" class="csl-entry" role="listitem">
Swayne, D. F., Cook, D., &amp; Buja, A. (1998). XGobi: Interactive dynamic data visualization in the x window system. <em>Journal of Computational and Graphical Statistics</em>, <em>7</em>(1), 113–130. <a href="https://doi.org/10.1080/10618600.1998.10474764">https://doi.org/10.1080/10618600.1998.10474764</a>
</div>
<div id="ref-SK98" class="csl-entry" role="listitem">
Swayne, D. F., &amp; Klinke, S. (1998). Editorial commentary. <em>Computational Statistics: Special Issue on The Use of Interactive Graphics</em>, <em>14</em>(1).
</div>
<div id="ref-STLBC03" class="csl-entry" role="listitem">
Swayne, D. F., Temple Lang, D., Buja, A., &amp; Cook, D. (2003). <span>GG</span>obi: <span>E</span>volving from <span>XG</span>obi into an <span>E</span>xtensible <span>F</span>ramework for <span>I</span>nteractive <span>D</span>ata <span>V</span>isualization. <em>Computational Statistics &amp; Data Analysis</em>, <em>43</em>, 423–444.
</div>
<div id="ref-SB98" class="csl-entry" role="listitem">
Swayne, D., &amp; Buja, A. (1998). <span>M</span>issing <span>D</span>ata in <span>I</span>nteractive <span>H</span>igh-<span>D</span>imensional <span>D</span>ata <span>V</span>isualization. <em>Computational Statistics</em>, <em>13</em>(1), 15–26.
</div>
<div id="ref-Symanzik2002NewAO" class="csl-entry" role="listitem">
Symanzik, J. (2002). New applications of the image grand tour. <em>Computing Science and Statistics</em>, <em>34</em>, 500--512. <a href="https://math.usu.edu/symanzik/papers/2002_interface.pdf">https://math.usu.edu/symanzik/papers/2002_interface.pdf</a>
</div>
<div id="ref-Symanzik04" class="csl-entry" role="listitem">
Symanzik, J. (2004). Interactive and <span>D</span>ynamic <span>G</span>raphics. In J. E. Gentle, W. Härdle, &amp; Y. Mori (Eds.), <em>Handbook of computational statistics: Concepts and methods</em> (pp. 293–336). Springer.
</div>
<div id="ref-TaGa02" class="csl-entry" role="listitem">
Takatsuka, M., &amp; Gahegan, M. (2002). <span>GeoVISTA Studio</span>: A <span>C</span>odeless <span>V</span>isual <span>P</span>rogramming <span>E</span>nvironment for <span>G</span>eoscientific <span>D</span>ata <span>A</span>nalysis and <span>V</span>isualization. <em>The Journal of Computers and Geosciences</em>, <em>28</em>(10), 1131–1144.
</div>
<div id="ref-ggfortify2016" class="csl-entry" role="listitem">
Tang, Y., Horikoshi, M., &amp; Li, W. (2016). Ggfortify: Unified interface to visualize statistical result of popular r packages. <em>The R Journal</em>, <em>8</em>(2), 474–485. <a href="https://doi.org/10.32614/RJ-2016-060">https://doi.org/10.32614/RJ-2016-060</a>
</div>
<div id="ref-TF95" class="csl-entry" role="listitem">
Tarpey, T., Li, L., &amp; Flury, B. (1995). Principal points and self–consistent points of elliptical distributions. <em>The Annals of Statistics</em>, <em>23</em>, 103–112.
</div>
<div id="ref-LSWL06:rggobi" class="csl-entry" role="listitem">
Temple Lang, D., Swayne, D., Wickham, H., &amp; Lawrence, M. (2006). <em><span><code>rggobi</code></span>: <span>A</span>n <span>I</span>nterface between <span>R</span> and <span>GG</span>obi</em>.<span><code> http://www.R-project.org</code></span>.
</div>
<div id="ref-R-rpart" class="csl-entry" role="listitem">
Therneau, T., &amp; Atkinson, B. (2023). <em>Rpart: Recursive partitioning and regression trees</em>. <a href="https://CRAN.R-project.org/package=rpart">https://CRAN.R-project.org/package=rpart</a>
</div>
<div id="ref-Theus02" class="csl-entry" role="listitem">
Theus, M. (2002). Interactive <span>D</span>ata <span>V</span>isualization <span>U</span>sing <span>M</span>ondrian. <em>Journal of Statistical Software</em>, <em>7</em>(11), http://www.jstatsoft.org.
</div>
<div id="ref-THW98" class="csl-entry" role="listitem">
Theus, M., Hofmann, H., &amp; Wilhelm, A. F. X. (1998). Selection <span>S</span>equences – <span>I</span>nteractive <span>A</span>nalysis of <span>M</span>assive <span>D</span>ata <span>S</span>ets. <em>Computing Science and Statistics</em>, <em>29</em>(1), 439–444.
</div>
<div id="ref-GThom93" class="csl-entry" role="listitem">
Thompson, G. L. (1993). Generalized <span>P</span>ermutation <span>P</span>olytopes and <span>E</span>xploratory <span>G</span>raphical <span>M</span>ethods for <span>R</span>anked <span>D</span>ata. <em>The Annals of Statistics</em>, <em>21</em>, 1401–1430.
</div>
<div id="ref-Ti91" class="csl-entry" role="listitem">
Tierney, L. (1991). <em><span>L</span>isp<span>S</span>tat: <span>A</span>n <span>O</span>bject-<span>O</span>rientated <span>E</span>nvironment for <span>S</span>tatistical <span>C</span>omputing and <span>D</span>ynamic <span>G</span>raphics</em>. John Wiley &amp; Sons.
</div>
<div id="ref-naniar" class="csl-entry" role="listitem">
Tierney, N., &amp; Cook, D. (2023a). Expanding tidy data principles to facilitate missing data exploration, visualization and assessment of imputations. <em>Journal of Statistical Software</em>, <em>105</em>(7), 1–31. <a href="https://doi.org/10.18637/jss.v105.i07">https://doi.org/10.18637/jss.v105.i07</a>
</div>
<div id="ref-naniar2023" class="csl-entry" role="listitem">
Tierney, N., &amp; Cook, D. (2023b). Expanding tidy data principles to facilitate missing data exploration, visualization and assessment of imputations. <em>Journal of Statistical Software</em>, <em>105</em>(7), 1–31. <a href="https://doi.org/10.18637/jss.v105.i07">https://doi.org/10.18637/jss.v105.i07</a>
</div>
<div id="ref-R-naniar" class="csl-entry" role="listitem">
Tierney, N., Cook, D., McBain, M., &amp; Fay, C. (2023). <em>Naniar: Data structures, summaries, and visualisations for missing data</em>. <a href="https://github.com/njtierney/naniar">https://github.com/njtierney/naniar</a>
</div>
<div id="ref-To52" class="csl-entry" role="listitem">
Torgerson, W. S. (1952). Multidimensional <span>S</span>caling. 1. <span>T</span>heory and <span>M</span>ethod. <em>Psychometrika</em>, <em>17</em>, 401–419.
</div>
<div id="ref-Tufte83" class="csl-entry" role="listitem">
Tufte, E. (1983). <em>The visual display of quantitative information</em>. Graphics Press.
</div>
<div id="ref-Tufte90" class="csl-entry" role="listitem">
Tufte, E. (1990). <em>Envisioning information</em>. Graphics Press.
</div>
<div id="ref-Tu65" class="csl-entry" role="listitem">
Tukey, J. W. (1965). The <span>T</span>echnical <span>T</span>ools of <span>S</span>tatistics. <em>The American Statistician</em>, <em>19</em>, 23–28.
</div>
<div id="ref-UHS96" class="csl-entry" role="listitem">
Unwin, A. R., Hawkins, G., Hofmann, H., &amp; Siegl, B. (1996). <span>I</span>nteractive <span>G</span>raphics for <span>D</span>ata <span>S</span>ets with <span>M</span>issing <span>V</span>alues - <span>MANET</span>. <em>Journal of Computational and Graphical Statistics</em>, <em>5</em>(2), 113–122.
</div>
<div id="ref-UHW02" class="csl-entry" role="listitem">
Unwin, A., Hofmann, H., &amp; Wilhelm, A. (2002). Direct <span>M</span>anipulation <span>G</span>raphics for <span>D</span>ata <span>M</span>ining. <em>Journal of Image and Graphics</em>, <em>2</em>(1), 49–65.
</div>
<div id="ref-Unwin06" class="csl-entry" role="listitem">
Unwin, A., Theus, M., &amp; Hofmann, H. (2006). <em>Graphics of <span>L</span>arge <span>D</span>atasets: <span>V</span>isualizing a <span>M</span>illion</em>. Springer.
</div>
<div id="ref-UVW03" class="csl-entry" role="listitem">
Unwin, A., Volinsky, C., &amp; Winkler, S. (2003). Parallel <span>C</span>oordinates for <span>E</span>xploratory <span>M</span>odelling <span>A</span>nalysis. <em>Comput. Stat. Data Anal.</em>, <em>43</em>(4), 553–564. https://doi.org/<a href="{\tt http://dx.doi.org/10.1016/S0167-9473(02)00292-X}">{\tt http://dx.doi.org/10.1016/S0167-9473(02)00292-X}</a>
</div>
<div id="ref-iPlots03" class="csl-entry" role="listitem">
Urbanek, S., &amp; Theus, M. (2003). <span class="nocase">iPlots</span>: <span>H</span>igh <span>I</span>nteraction <span>G</span>raphics for <span>R</span>. In K. Hornik, F. Leisch, &amp; A. Zeileis (Eds.), <em>Proceedings of the 3rd international workshop on distributed statistical computing (DSC 2003)</em>.
</div>
<div id="ref-R-htmlwidgets" class="csl-entry" role="listitem">
Vaidyanathan, R., Xie, Y., Allaire, J., Cheng, J., Sievert, C., &amp; Russell, K. (2023). <em>Htmlwidgets: HTML widgets for r</em>. <a href="https://github.com/ramnathv/htmlwidgets">https://github.com/ramnathv/htmlwidgets</a>
</div>
<div id="ref-Rtsne2014" class="csl-entry" role="listitem">
van der Maaten, L. J. P. (2014). Accelerating t-SNE using tree-based algorithms. <em>Journal of Machine Learning Research</em>, <em>15</em>, 3221–3245.
</div>
<div id="ref-Rtsne2008" class="csl-entry" role="listitem">
van der Maaten, L. J. P., &amp; Hinton, G. E. (2008). Visualizing high-dimensional data using t-SNE. <em>Journal of Machine Learning Research</em>, <em>9</em>, 2579–2605.
</div>
<div id="ref-Va99" class="csl-entry" role="listitem">
Vapnik, V. N. (1999). <em>The <span>N</span>ature of <span>S</span>tatistical <span>L</span>earning <span>T</span>heory</em>. Springer.
</div>
<div id="ref-VeVe85" class="csl-entry" role="listitem">
Velleman, P. F., &amp; Velleman, A. Y. (1985). <em>Data desk handbook</em>. Data Description, Inc.
</div>
<div id="ref-VR02" class="csl-entry" role="listitem">
Venables, W. N., &amp; Ripley, B. (2002a). <em>Modern <span>A</span>pplied <span>S</span>tatistics with <span>S</span></em>. Springer-Verlag.
</div>
<div id="ref-MASS2002" class="csl-entry" role="listitem">
Venables, W. N., &amp; Ripley, B. D. (2002b). <em>Modern applied statistics with s</em> (Fourth). Springer. <a href="https://www.stats.ox.ac.uk/pub/MASS4/">https://www.stats.ox.ac.uk/pub/MASS4/</a>
</div>
<div id="ref-nnet2002" class="csl-entry" role="listitem">
Venables, W. N., &amp; Ripley, B. D. (2002c). <em>Modern applied statistics with s</em> (Fourth). Springer. <a href="https://www.stats.ox.ac.uk/pub/MASS4/">https://www.stats.ox.ac.uk/pub/MASS4/</a>
</div>
<div id="ref-Wa00" class="csl-entry" role="listitem">
Wainer, H. (2000). <em>Visual <span>R</span>evelations (2nd ed)</em>. LEA, Inc.
</div>
<div id="ref-WS05a" class="csl-entry" role="listitem">
Wainer, H., &amp; Spence, I. (eds). (2005a). <em>The <span>C</span>ommercial and <span>P</span>olitical <span>A</span>tlas, <span>R</span>epresenting, by means of <span>S</span>tained <span>C</span>opper-<span>P</span>late <span>C</span>harts, <span>T</span>he <span>P</span>rogress of the <span>C</span>ommerce, <span>R</span>evenues, <span>E</span>xpenditure, and <span>D</span>ebts of <span>E</span>ngland, during the whole of the <span>E</span>ighteenth <span>C</span>entury, by <span>W</span>illiam <span>P</span>layfair</em>. Cambridge University Press.
</div>
<div id="ref-WS05b" class="csl-entry" role="listitem">
Wainer, H., &amp; Spence, I. (eds). (2005b). <em>The <span>S</span>tatistical <span>B</span>reviary; <span>S</span>hewing on a <span>P</span>rinciple entirely new, the resources of every state and kingdom in <span>E</span>urope; illustrated with <span>S</span>tained <span>C</span>opper-<span>P</span>late <span>C</span>harts, representing the physical powers of each distinct nation with ease and perspicuity by <span>W</span>illiam <span>P</span>layfair</em>. Cambridge University Press.
</div>
<div id="ref-Wa78" class="csl-entry" role="listitem">
Wang, P. C. C. (Ed.). (1978). <em><span>G</span>raphical <span>R</span>epresentation of <span>M</span>ultivariate <span>D</span>ata</em>. Academic Press.
</div>
<div id="ref-We90" class="csl-entry" role="listitem">
Wegman, E. (1990). <span>H</span>yperdimensional <span>D</span>ata <span>A</span>nalysis <span>U</span>sing <span>P</span>arallel <span>C</span>oordinates. <em>Journal of American Statistics Association</em>, <em>85</em>, 664–675.
</div>
<div id="ref-We91" class="csl-entry" role="listitem">
Wegman, E. J. (1991). <em>The <span>G</span>rand <span>T</span>our in <span class="math inline">\(k\)</span>-<span>D</span>imensions</em> (Technical Report No. 68). Center for Computational Statistics, George Mason University.
</div>
<div id="ref-WC93" class="csl-entry" role="listitem">
Wegman, E. J., &amp; Carr, D. B. (1993). <em><span>S</span>tatistical <span>G</span>raphics and <span>V</span>isualization</em> (C. R. Rao, Ed.; pp. 857–958). Elsevier Science Publishers.
</div>
<div id="ref-WPS98" class="csl-entry" role="listitem">
Wegman, E. J., Poston, W. L., &amp; Solka, J. L. (1998). Image <span>G</span>rand <span>T</span>our. <em>Automatic Target Recognition VIII - Proceedings of SPIE, 3371</em>, 286–294.
</div>
<div id="ref-kohonen2007" class="csl-entry" role="listitem">
Wehrens, R., &amp; Buydens, L. M. C. (2007). Self- and super-organizing maps in <span>R</span>: The <span class="nocase">kohonen</span> package. <em>Journal of Statistical Software</em>, <em>21</em>(5), 1–19. <a href="https://doi.org/10.18637/jss.v021.i05">https://doi.org/10.18637/jss.v021.i05</a>
</div>
<div id="ref-kohonen2018" class="csl-entry" role="listitem">
Wehrens, R., &amp; Kruisselbrink, J. (2018). Flexible self-organizing maps in <span class="nocase">kohonen</span> 3.0. <em>Journal of Statistical Software</em>, <em>87</em>(7), 1–18. <a href="https://doi.org/10.18637/jss.v087.i07">https://doi.org/10.18637/jss.v087.i07</a>
</div>
<div id="ref-R-kohonen" class="csl-entry" role="listitem">
Wehrens, R., &amp; Kruisselbrink, J. (2023). <em>Kohonen: Supervised and unsupervised self-organising maps</em>. <a href="https://CRAN.R-project.org/package=kohonen">https://CRAN.R-project.org/package=kohonen</a>
</div>
<div id="ref-ggplot22016" class="csl-entry" role="listitem">
Wickham, H. (2016). <em>ggplot2: Elegant graphics for data analysis</em>. Springer-Verlag New York. <a href="https://ggplot2.tidyverse.org">https://ggplot2.tidyverse.org</a>
</div>
<div id="ref-R-classifly" class="csl-entry" role="listitem">
Wickham, H. (2022). <em>Classifly: Explore classification models in high dimensions</em>. <a href="http://had.co.nz/classifly">http://had.co.nz/classifly</a>
</div>
<div id="ref-R-ggplot2" class="csl-entry" role="listitem">
Wickham, H., Chang, W., Henry, L., Pedersen, T. L., Takahashi, K., Wilke, C., Woo, K., Yutani, H., &amp; Dunnington, D. (2023). <em>ggplot2: Create elegant data visualisations using the grammar of graphics</em>. <a href="https://CRAN.R-project.org/package=ggplot2">https://CRAN.R-project.org/package=ggplot2</a>
</div>
<div id="ref-R-tourr" class="csl-entry" role="listitem">
Wickham, H., &amp; Cook, D. (2024). <em>Tourr: Tour methods for multivariate data visualisation</em>. <a href="https://github.com/ggobi/tourr">https://github.com/ggobi/tourr</a>
</div>
<div id="ref-wickham2015" class="csl-entry" role="listitem">
Wickham, H., Cook, D., &amp; Hofmann, H. (2015). Visualizing statistical models: Removing the blindfold. <em>Statistical Analysis and Data Mining: The ASA Data Science Journal</em>, <em>8</em>(4), 203–225. <a href="https://doi.org/10.1002/sam.11271">https://doi.org/10.1002/sam.11271</a>
</div>
<div id="ref-tourr" class="csl-entry" role="listitem">
Wickham, H., Cook, D., Hofmann, H., &amp; Buja, A. (2011a). Tourr: <span>An</span> <span>R</span> <span>Package</span> for <span>Exploring</span> <span>Multivariate</span> <span>Data</span> with <span>Projections</span>. <em>Journal of Statistical Software</em>, <em>40</em>(2). <a href="https://doi.org/10.18637/jss.v040.i02">https://doi.org/10.18637/jss.v040.i02</a>
</div>
<div id="ref-tourr2011" class="csl-entry" role="listitem">
Wickham, H., Cook, D., Hofmann, H., &amp; Buja, A. (2011b). <span class="nocase">tourr</span>: An <span>R</span> package for exploring multivariate data with projections. <em>Journal of Statistical Software</em>, <em>40</em>(2), 1–18. <a href="https://doi.org/10.18637/jss.v040.i02">https://doi.org/10.18637/jss.v040.i02</a>
</div>
<div id="ref-R-dplyr" class="csl-entry" role="listitem">
Wickham, H., François, R., Henry, L., Müller, K., &amp; Vaughan, D. (2023). <em>Dplyr: A grammar of data manipulation</em>. <a href="https://CRAN.R-project.org/package=dplyr">https://CRAN.R-project.org/package=dplyr</a>
</div>
<div id="ref-R-readr" class="csl-entry" role="listitem">
Wickham, H., Hester, J., &amp; Bryan, J. (2023). <em>Readr: Read rectangular text data</em>. <a href="https://CRAN.R-project.org/package=readr">https://CRAN.R-project.org/package=readr</a>
</div>
<div id="ref-WWS99" class="csl-entry" role="listitem">
Wilhelm, A. F. X., Wegman, E. J., &amp; Symanzik, J. (1999). Visual <span>C</span>lustering and <span>C</span>lassification: <span>T</span>he <span>O</span>ronsay <span>P</span>article <span>S</span>ize <span>D</span>ata <span>S</span>et <span>R</span>evisited. <em>Computational Statistics: Special Issue on Interactive Graphical Data Analysis</em>, <em>14</em>(1), 109–146.
</div>
<div id="ref-Wilkinson05" class="csl-entry" role="listitem">
Wilkinson, L. (2005). <em>The grammar of graphics</em>. Springer.
</div>
<div id="ref-Wills99" class="csl-entry" role="listitem">
Wills, G. (1999). NicheWorks – <span>I</span>nteractive <span>V</span>isualization of <span>V</span>ery <span>L</span>arge <span>G</span>raphs. <em>Journal of Computational and Graphical Statistics</em>, <em>8</em>(2), 190–212.
</div>
<div id="ref-cranvas" class="csl-entry" role="listitem">
Xie, Y., Hofmann, H., &amp; Cheng, X. (2014). <span class="nocase">Reactive Programming for Interactive Graphics</span>. <em>Statistical Science</em>, <em>29</em>(2), 201–213. <a href="https://doi.org/10.1214/14-STS477">https://doi.org/10.1214/14-STS477</a>
</div>
<div id="ref-Young06" class="csl-entry" role="listitem">
Young, F. W., Valero-Mora, P. M., &amp; Friendly, M. (2006). <em>Visual <span>S</span>tatistics: <span>S</span>eeing <span>D</span>ata with <span>D</span>ynamic <span>I</span>nteractive <span>G</span>raphics</em>. John Wiley &amp; Sons.
</div>
<div id="ref-colorspace2020a" class="csl-entry" role="listitem">
Zeileis, A., Fisher, J. C., Hornik, K., Ihaka, R., McWhite, C. D., Murrell, P., Stauffer, R., &amp; Wilke, C. O. (2020). <span class="nocase">colorspace</span>: A toolbox for manipulating and assessing colors and palettes. <em>Journal of Statistical Software</em>, <em>96</em>(1), 1–49. <a href="https://doi.org/10.18637/jss.v096.i01">https://doi.org/10.18637/jss.v096.i01</a>
</div>
<div id="ref-colorspace2009a" class="csl-entry" role="listitem">
Zeileis, A., Hornik, K., &amp; Murrell, P. (2009). Escaping <span>RGB</span>land: Selecting colors for statistical graphics. <em>Computational Statistics &amp; Data Analysis</em>, <em>53</em>(9), 3259–3270. <a href="https://doi.org/10.1016/j.csda.2008.11.033">https://doi.org/10.1016/j.csda.2008.11.033</a>
</div>
<div id="ref-zhang2023" class="csl-entry" role="listitem">
Zhang, C., Ye, J., &amp; Wang, X. (2023). A computational perspective on projection pursuit in high dimensions: Feasible or infeasible feature extraction. <em>International Statistical Review</em>, <em>91</em>(1), 140–161. <a href="https://doi.org/10.1111/insr.12517">https://doi.org/10.1111/insr.12517</a>
</div>
<div id="ref-ferrn2021" class="csl-entry" role="listitem">
Zhang, H. S., Cook, D., Laa, U., Langrené, N., &amp; Menéndez, P. (2021). Visual diagnostics for constrained optimisation with application to guided tours. <em>The R Journal</em>, <em>13</em>(2), 624–641. <a href="https://doi.org/10.32614/RJ-2021-105">https://doi.org/10.32614/RJ-2021-105</a>
</div>
<div id="ref-R-ferrn" class="csl-entry" role="listitem">
Zhang, H. S., Cook, D., Laa, U., Langrené, N., &amp; Menéndez, P. (2022). <em>Ferrn: Facilitate exploration of touRR optimisatioN</em>. <a href="https://github.com/huizezhang-sherry/ferrn/">https://github.com/huizezhang-sherry/ferrn/</a>
</div>
<div id="ref-R-kableExtra" class="csl-entry" role="listitem">
Zhu, H. (2021). <em>kableExtra: Construct complex table with kable and pipe syntax</em>. <a href="https://CRAN.R-project.org/package=kableExtra">https://CRAN.R-project.org/package=kableExtra</a>
</div>
</div>
</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./14-lda.html" class="pagination-link  aria-label=" discriminant="" analysis="">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Linear discriminant analysis</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./16-svm.html" class="pagination-link" aria-label="<span class='chapter-number'>16</span>&nbsp; <span class='chapter-title'>Support vector machines</span>">
        <span class="nav-page-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Support vector machines</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>Interactively exploring high-dimensional data and models in R was written by Dianne Cook and Ursula Laa.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/dicook/mulgar_book/edit/master/15-forests.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/dicook/mulgar_book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>


</body></html>