<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.549">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Interactively exploring high-dimensional data and models in R - 17&nbsp; Neural networks and deep learning</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./18-summary-class.html" rel="next">
<link href="./16-svm.html" rel="prev">
<link href="./cover.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script><script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./17-nn.html"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Neural networks and deep learning</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Interactively exploring high-dimensional data and models in R</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/dicook/mulgar_book" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./1-intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Picturing high dimensions</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./2-notation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Notation conventions and R objects</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./3-intro-dimred.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Dimension reduction overview</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./4-pca.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Principal component analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./5-nldr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Non-linear dimension reduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./6-intro-clust.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Introduction to clustering</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./7-spin-and-brush.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Spin-and-brush approach</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./8-hierarchical.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Hierarchical clustering</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./9-kmeans.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title"><span class="math inline">\(k\)</span>-means clustering</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10-model-based.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Model-based clustering</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11-som.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Self-organizing maps</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12-summary-clust.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Summarising and comparing clustering results</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./13-intro-class.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Introduction to supervised classification</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./14-lda.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Linear discriminant analysis</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./15-forests.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Trees and forests</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./16-svm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Support vector machines</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./17-nn.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Neural networks and deep learning</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./18-summary-class.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Exploring misclassifications</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A1-toolbox.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Toolbox</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A2-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A3-book-code-and-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">C</span>&nbsp; <span class="chapter-title">Links to Book Code and Additional Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./A4-glossary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">D</span>&nbsp; <span class="chapter-title">Glossary</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#setting-up-the-model" id="toc-setting-up-the-model" class="nav-link active" data-scroll-target="#setting-up-the-model"><span class="header-section-number">17.1</span> Setting up the model</a></li>
  <li><a href="#checking-the-trainingtest-split" id="toc-checking-the-trainingtest-split" class="nav-link" data-scroll-target="#checking-the-trainingtest-split"><span class="header-section-number">17.2</span> Checking the training/test split</a></li>
  <li><a href="#fit-the-model" id="toc-fit-the-model" class="nav-link" data-scroll-target="#fit-the-model"><span class="header-section-number">17.3</span> Fit the model</a></li>
  <li><a href="#extracting-model-components" id="toc-extracting-model-components" class="nav-link" data-scroll-target="#extracting-model-components"><span class="header-section-number">17.4</span> Extracting model components</a></li>
  <li><a href="#examining-predictive-probabilities" id="toc-examining-predictive-probabilities" class="nav-link" data-scroll-target="#examining-predictive-probabilities"><span class="header-section-number">17.5</span> Examining predictive probabilities</a></li>
  <li><a href="#local-explanations" id="toc-local-explanations" class="nav-link" data-scroll-target="#local-explanations"><span class="header-section-number">17.6</span> Local explanations</a></li>
  <li><a href="#examining-boundaries" id="toc-examining-boundaries" class="nav-link" data-scroll-target="#examining-boundaries"><span class="header-section-number">17.7</span> Examining boundaries</a></li>
  <li><a href="#application-to-a-large-dataset" id="toc-application-to-a-large-dataset" class="nav-link" data-scroll-target="#application-to-a-large-dataset"><span class="header-section-number">17.8</span> Application to a large dataset</a></li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises">Exercises</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/dicook/mulgar_book/edit/master/17-nn.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/dicook/mulgar_book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">
<span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Neural networks and deep learning</span>
</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><p></p>
<p>Neural networks (NN) can be considered to be nested additive (or even ensemble) models where explanatory variables are combined, and transformed through an activation function like a logistic. These transformed combinations are added recursively to yield class predictions. They are considered to be black box models, but there is a growing demand for interpretability. Although interpretability is possible, it can be unappealing to understand a complex model constructed to tackle a difficult classification task. Nevertheless, this is the motivation for the explanation of visualisation for NN models in this chapter.</p>
<p>In the simplest form, we might write the equation for a NN as</p>
<p><span class="math display">\[
\hat{y} = f(x) = a_0+\sum_{h=1}^{s}
w_{0h}\phi(a_h+\sum_{i=1}^{p} w_{ih}x_i)
\]</span> where <span class="math inline">\(s\)</span> indicates the number of nodes in the hidden (middle layer), and <span class="math inline">\(\phi\)</span> is a choice of activation function. In a simple situation where <span class="math inline">\(p=3\)</span>, <span class="math inline">\(s=2\)</span>, and linear output layer, the model could be written as:</p>
<p><span class="math display">\[
\begin{aligned}
\hat{y} = a_0+ &amp; w_{01}\phi(a_1+w_{11}x_1+w_{21}x_2+w_{31}x_3) +\\
&amp; w_{02}\phi(a_2+w_{12}x_1+w_{22}x_2+w_{32}x_3)
\end{aligned}
\]</span> which is a combination of two (linear) models, each of which could be examined for their role in making predictions.</p>
<p>In practice, a model may have many nodes, and several hidden layers, a variety of activation functions, and regularisation modifications. One should keep in mind the principle of parsimony is important when applying NNs, because it is tempting to make an overly complex, and thus over-parameterised, construction. Fitting NNs is still problematic. One would hope that fitting produces a stable result, whatever the starting seed the same parameter estimates are returned. However, this is not the case, and different, sometimes radically different, results are routinely obtained after each attempted fit <span class="citation" data-cites="wickham2015">(<a href="references.html#ref-wickham2015" role="doc-biblioref">Wickham et al., 2015</a>)</span>.</p>
<p>For these examples we use the software <code>keras</code> <span class="citation" data-cites="keras">(<a href="references.html#ref-keras" role="doc-biblioref">Allaire &amp; Chollet, 2023</a>)</span> following the installation and tutorial details at <a href="https://tensorflow.rstudio.com/tutorials/">https://tensorflow.rstudio.com/tutorials/</a>. Because it is an interface to python it can be tricky to install. If this is a problem, the example code should be possible to convert to use <code>nnet</code> <span class="citation" data-cites="VR02">(<a href="references.html#ref-VR02" role="doc-biblioref">Venables &amp; Ripley, 2002a</a>)</span> or <code>neuralnet</code> <span class="citation" data-cites="neuralnet">(<a href="references.html#ref-neuralnet" role="doc-biblioref">Fritsch et al., 2019</a>)</span>. We will use the penguins data to illustrate the fitting, because it makes it easier to understand the procedures and the fit. However, a NN is like using a jackhammer instead of a trowel to plant a seedling, more complicated than necessary to build a good classification model for this data.</p>
<section id="setting-up-the-model" class="level2" data-number="17.1"><h2 data-number="17.1" class="anchored" data-anchor-id="setting-up-the-model">
<span class="header-section-number">17.1</span> Setting up the model</h2>
<p></p>
<p>A first step is to decide how many nodes the NN architecture should have, and what activation function should be used. To make these decisions, ideally you already have some knowledge of the shapes of class clusters. For the penguins classification, we have seen that it contains three elliptically shaped clusters of roughly the same size. This suggests two nodes in the hidden layer would be sufficient to separate three clusters (<a href="#fig-nn-diagram" class="quarto-xref">Figure&nbsp;<span>17.1</span></a>). Because the shapes of the clusters are convex, using linear activation (“relu”) will also be sufficient. The model specification is as follows:</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tensorflow.rstudio.com/">keras</a></span><span class="op">)</span></span>
<span><span class="fu">tensorflow</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tensorflow/man/set_random_seed.html">set_random_seed</a></span><span class="op">(</span><span class="fl">211</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define model</span></span>
<span><span class="va">p_nn_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/keras/man/keras_model_sequential.html">keras_model_sequential</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p_nn_model</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/keras/man/layer_dense.html">layer_dense</a></span><span class="op">(</span>units <span class="op">=</span> <span class="fl">2</span>, activation <span class="op">=</span> <span class="st">'relu'</span>, </span>
<span>              input_shape <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/keras/man/layer_dense.html">layer_dense</a></span><span class="op">(</span>units <span class="op">=</span> <span class="fl">3</span>, activation <span class="op">=</span> <span class="st">'softmax'</span><span class="op">)</span></span>
<span><span class="va">p_nn_model</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="va">summary</span></span>
<span></span>
<span><span class="va">loss_fn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/keras/man/loss-functions.html">loss_sparse_categorical_crossentropy</a></span><span class="op">(</span></span>
<span>  from_logits <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p_nn_model</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://generics.r-lib.org/reference/compile.html">compile</a></span><span class="op">(</span></span>
<span>  optimizer <span class="op">=</span> <span class="st">"adam"</span>,</span>
<span>  loss      <span class="op">=</span> <span class="va">loss_fn</span>,</span>
<span>  metrics   <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'accuracy'</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Note that <code>tensorflow::set_random_seed(211)</code> sets the seed for the model fitting so that we can obtain the same result to discuss later. It needs to be set before the model is defined in the code. The model will also be saved in order to diagnose and make predictions.</p>
<div id="fig-nn-diagram" class="quarto-figure quarto-figure-center quarto-float anchored" data-align="center">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-nn-diagram-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/nn-diagram.png" class="img-fluid figure-img" data-align="center">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-nn-diagram-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;17.1: Network architecture for the model on the penguins data. The round nodes indicate original or transformed variables, and each arrow connecting these is represented as one of the weights <span class="math inline">\(w_{ih}\)</span> in the definition. The boxes indicate the additive constant entering the nodes, and the corresponding arrows represent the terms <span class="math inline">\(a_h\)</span>.
</figcaption></figure>
</div>
</section><section id="checking-the-trainingtest-split" class="level2" data-number="17.2"><h2 data-number="17.2" class="anchored" data-anchor-id="checking-the-trainingtest-split">
<span class="header-section-number">17.2</span> Checking the training/test split</h2>
<p></p>
<p>Splitting the data into training and test is an essential way to protect against overfitting, for most classifiers, but especially so for the copiously parameterised NNs. The model specified for the penguins data with only two nodes is unlikely to be overfitted, but it is nevertheless good practice to use a training set for building and a test set for evaluation.</p>
<p><a href="#fig-p-split-html" class="quarto-xref">Figure&nbsp;<span>17.2</span></a> shows the tour being used to examine the split into training and test samples for the penguins data. Using random sampling, particularly stratified by group, should result the two sets being very similar, as can be seen here. It does happen that several observations in the test set are on the extremes of their class cluster, so it could be that the model makes errors in the neighbourhoods of these points.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Split the data intro training and testing</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jrnold/ggthemes">ggthemes</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://rsample.tidymodels.org">rsample</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/eclarke/ggbeeswarm">ggbeeswarm</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/penguins_sub.rda"</span><span class="op">)</span> <span class="co"># from mulgar book</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">821</span><span class="op">)</span></span>
<span><span class="va">p_split</span> <span class="op">&lt;-</span> <span class="va">penguins_sub</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">bl</span><span class="op">:</span><span class="va">species</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rsample.tidymodels.org/reference/initial_split.html">initial_split</a></span><span class="op">(</span>prop <span class="op">=</span> <span class="fl">2</span><span class="op">/</span><span class="fl">3</span>, </span>
<span>                strata<span class="op">=</span><span class="va">species</span><span class="op">)</span></span>
<span><span class="va">p_train</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rsample.tidymodels.org/reference/initial_split.html">training</a></span><span class="op">(</span><span class="va">p_split</span><span class="op">)</span></span>
<span><span class="va">p_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rsample.tidymodels.org/reference/initial_split.html">testing</a></span><span class="op">(</span><span class="va">p_split</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Check training and test split</span></span>
<span><span class="va">p_split_check</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html">bind_rows</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html">bind_cols</a></span><span class="op">(</span><span class="va">p_train</span>, type <span class="op">=</span> <span class="st">"train"</span><span class="op">)</span>, </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html">bind_cols</a></span><span class="op">(</span><span class="va">p_test</span>, type <span class="op">=</span> <span class="st">"test"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">type</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold"><summary>Code to run tours</summary><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">animate_xy</span><span class="op">(</span><span class="va">p_split_check</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, </span>
<span>           col<span class="op">=</span><span class="va">p_split_check</span><span class="op">$</span><span class="va">species</span>,</span>
<span>           pch<span class="op">=</span><span class="va">p_split_check</span><span class="op">$</span><span class="va">type</span><span class="op">)</span></span>
<span><span class="fu">animate_xy</span><span class="op">(</span><span class="va">p_split_check</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, </span>
<span>           <span class="fu">guided_tour</span><span class="op">(</span><span class="fu">lda_pp</span><span class="op">(</span><span class="va">p_split_check</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           col<span class="op">=</span><span class="va">p_split_check</span><span class="op">$</span><span class="va">species</span>,</span>
<span>           pch<span class="op">=</span><span class="va">p_split_check</span><span class="op">$</span><span class="va">type</span><span class="op">)</span></span>
<span><span class="fu">render_gif</span><span class="op">(</span><span class="va">p_split_check</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>           <span class="fu">grand_tour</span><span class="op">(</span><span class="op">)</span>,</span>
<span>           <span class="fu">display_xy</span><span class="op">(</span> </span>
<span>             col<span class="op">=</span><span class="va">p_split_check</span><span class="op">$</span><span class="va">species</span>, </span>
<span>             pch<span class="op">=</span><span class="va">p_split_check</span><span class="op">$</span><span class="va">type</span>,</span>
<span>             cex<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>             axes<span class="op">=</span><span class="st">"bottomleft"</span><span class="op">)</span>, </span>
<span>           gif_file<span class="op">=</span><span class="st">"gifs/p_split.gif"</span>,</span>
<span>           frames<span class="op">=</span><span class="fl">500</span>,</span>
<span>           loop<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">render_gif</span><span class="op">(</span><span class="va">p_split_check</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>,</span>
<span>           <span class="fu">guided_tour</span><span class="op">(</span><span class="fu">lda_pp</span><span class="op">(</span><span class="va">p_split_check</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           <span class="fu">display_xy</span><span class="op">(</span> </span>
<span>             col<span class="op">=</span><span class="va">p_split_check</span><span class="op">$</span><span class="va">species</span>, </span>
<span>             pch<span class="op">=</span><span class="va">p_split_check</span><span class="op">$</span><span class="va">type</span>,</span>
<span>             cex<span class="op">=</span><span class="fl">1.5</span>,</span>
<span>             axes<span class="op">=</span><span class="st">"bottomleft"</span><span class="op">)</span>, </span>
<span>           gif_file<span class="op">=</span><span class="st">"gifs/p_split_guided.gif"</span>,</span>
<span>           frames<span class="op">=</span><span class="fl">500</span>,</span>
<span>           loop<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="fig-p-split-html" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-p-split-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-p-split-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-split-grand" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-split-grand-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gifs/p_split.gif" class="img-fluid figure-img" data-ref-parent="fig-p-split-html" alt="FIX ME" width="300">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-split-grand-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Grand tour
</figcaption></figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-p-split-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-split-guided" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-split-guided-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gifs/p_split_guided.gif" class="img-fluid figure-img" data-ref-parent="fig-p-split-html" alt="FIX ME" width="300">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-split-guided-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) Guided tour
</figcaption></figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-p-split-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;17.2: Evaluating the training/test split, where we expect that the two samples should roughly match. There are a few observations in the test set that are on the outer edges of the clusters, which will likely result in the model making an error in these regions. However, the two samples roughly match.
</figcaption></figure>
</div>
</section><section id="fit-the-model" class="level2" data-number="17.3"><h2 data-number="17.3" class="anchored" data-anchor-id="fit-the-model">
<span class="header-section-number">17.3</span> Fit the model</h2>
<p></p>
<p>The data needs to be specially formatted for the model fitted using <code>keras</code>. The explanatory variables need to be provided as a <code>matrix</code>, and the categorical response needs to be separate, and specified as a <code>numeric</code> variable, beginning with 0.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Data needs to be matrix, and response needs to be numeric</span></span>
<span><span class="va">p_train_x</span> <span class="op">&lt;-</span> <span class="va">p_train</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">bl</span><span class="op">:</span><span class="va">bm</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p_train_y</span> <span class="op">&lt;-</span> <span class="va">p_train</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">species</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="va">p_train_y</span> <span class="op">&lt;-</span> <span class="va">p_train_y</span><span class="op">-</span><span class="fl">1</span> <span class="co"># Needs to be 0, 1, 2</span></span>
<span><span class="va">p_test_x</span> <span class="op">&lt;-</span> <span class="va">p_test</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">bl</span><span class="op">:</span><span class="va">bm</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">p_test_y</span> <span class="op">&lt;-</span> <span class="va">p_test</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html">pull</a></span><span class="op">(</span><span class="va">species</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="va">p_test_y</span> <span class="op">&lt;-</span> <span class="va">p_test_y</span><span class="op">-</span><span class="fl">1</span> <span class="co"># Needs to be 0, 1, 2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The specified model is reasonably simple, four input variables, two nodes in the hidden layer and a three column binary matrix for output. This corresponds to 5+5+3+3+3=19 parameters.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>Model: "sequential"
________________________________________________________________________________
 Layer (type)                       Output Shape                    Param #     
================================================================================
 dense_1 (Dense)                    (None, 2)                       10          
 dense (Dense)                      (None, 3)                       9           
================================================================================
Total params: 19 (76.00 Byte)
Trainable params: 19 (76.00 Byte)
Non-trainable params: 0 (0.00 Byte)
________________________________________________________________________________</code></pre>
</div>
</div>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Fit model</span></span>
<span><span class="va">p_nn_fit</span> <span class="op">&lt;-</span> <span class="va">p_nn_model</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu">keras</span><span class="fu">::</span><span class="fu"><a href="https://generics.r-lib.org/reference/fit.html">fit</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="va">p_train_x</span>, </span>
<span>  y <span class="op">=</span> <span class="va">p_train_y</span>,</span>
<span>  epochs <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  verbose <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Because we set the random number seed we will get the same fit each time the code provided here is run. However, if the model is re-fit without setting the seed, you will see that there is a surprising amount of variability in the fits. Setting <code>epochs = 200</code> helps to usually get a good fit. One expects that <code>keras</code> is reasonably stable so one would not expect the huge array of fits as observed in <span class="citation" data-cites="wickham2015">Wickham et al. (<a href="references.html#ref-wickham2015" role="doc-biblioref">2015</a>)</span> using <code>nnet</code>. That this can happen with the simple model used here reinforces the notion that fitting of NN models is fiddly, and great care needs to be taken to validate and diagnose the fit.</p>
<div class="info">
<p>Fitting NN models is fiddly, and very different fitted models can result from restarts, parameter choices, and architecture.</p>
</div>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tensorflow.rstudio.com/">keras</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://colorspace.R-Forge.R-project.org/">colorspace</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load fitted model</span></span>
<span><span class="va">p_nn_model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/keras/man/save_model_tf.html">load_model_tf</a></span><span class="op">(</span><span class="st">"data/penguins_cnn"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The fitted model that we have chosen as the final one has reasonably small loss and high accuracy. Plots of loss and accuracy across epochs showing the change during fitting can be plotted, but we don’t show them here, because they are generally not very interesting.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p_nn_model</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/tensorflow/man/evaluate.html">evaluate</a></span><span class="op">(</span><span class="va">p_test_x</span>, <span class="va">p_test_y</span>, verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>     loss  accuracy 
0.2563850 0.9553571 </code></pre>
</div>
</div>
<p>The model object can be saved for later use with:</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/keras/man/save_model_tf.html">save_model_tf</a></span><span class="op">(</span><span class="va">p_nn_model</span>, <span class="st">"data/penguins_cnn"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section><section id="extracting-model-components" class="level2" data-number="17.4"><h2 data-number="17.4" class="anchored" data-anchor-id="extracting-model-components">
<span class="header-section-number">17.4</span> Extracting model components</h2>
<p></p>
<div class="info">
<p>View the individual node models to understand how they combine to produce the overall model.</p>
</div>
<p>Because nodes in the hidden layers of NNs are themselves (relatively simple regression) models, it can be interesting to examine these to understand how the model is making it’s predictions. Although it’s rarely easy, most software will allow the coefficients for the models at these nodes to be extracted. With the penguins NN model there are two nodes, so we can extract the coefficients and plot the resulting two linear combinations to examine the separation between classes.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Extract hidden layer model weights</span></span>
<span><span class="va">p_nn_wgts</span> <span class="op">&lt;-</span> <span class="fu">keras</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/keras/man/get_weights.html">get_weights</a></span><span class="op">(</span><span class="va">p_nn_model</span>, trainable<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">p_nn_wgts</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>[[1]]
           [,1]        [,2]
[1,]  0.6216676  1.33304155
[2,]  0.1851478 -0.01596385
[3,] -0.1680396 -0.30432791
[4,] -0.8867414 -0.36627045

[[2]]
[1]  0.12708087 -0.09466381

[[3]]
           [,1]     [,2]       [,3]
[1,] -0.1646167 1.527644 -1.9215064
[2,] -0.7547278 1.555889  0.3210194

[[4]]
[1]  0.4554813 -0.9371488  0.3577386</code></pre>
</div>
</div>
<p>The linear coefficients for the first node in the model are 0.62, 0.19, -0.17, -0.89, and the second node in the model are 1.33, -0.02, -0.3, -0.37. We can use these like we used the linear discriminants in LDA to make a 2D view of the data, where the model is separating the three species. The constants 0.13, -0.09 are not important for this. They are only useful for drawing the location of the boundaries between classes produced by the model.</p>
<p>These two sets of model coefficients provide linear combinations of the original variables. Together, they define a plane on which the data is projected to view the classification produced by the model. Ideally, though this plane should be defined using an orthonormal basis otherwise the shape of the data distribution might be warped. So we orthonormalise this matrix before computing the data projection.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Orthonormalise the weights to make 2D projection</span></span>
<span><span class="va">p_nn_wgts_on</span> <span class="op">&lt;-</span> <span class="fu">tourr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tourr/man/orthonormalise.html">orthonormalise</a></span><span class="op">(</span><span class="va">p_nn_wgts</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">p_nn_wgts_on</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>           [,1]       [,2]
[1,]  0.5593355  0.7969849
[2,]  0.1665838 -0.2145664
[3,] -0.1511909 -0.1541475
[4,] -0.7978314  0.5431528</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Hidden layer</span></span>
<span><span class="va">p_train_m</span> <span class="op">&lt;-</span> <span class="va">p_train</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>nn1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">p_train</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span></span>
<span>           <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">p_nn_wgts_on</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>         nn2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">p_train</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span></span>
<span>           <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="va">p_nn_wgts_on</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Now add the test points on.</span></span>
<span><span class="va">p_test_m</span> <span class="op">&lt;-</span> <span class="va">p_test</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>nn1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">p_test</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span></span>
<span>           <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">p_nn_wgts_on</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span>,</span>
<span>         nn2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">p_test</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span></span>
<span>           <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="va">p_nn_wgts_on</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, ncol<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p_train_m</span> <span class="op">&lt;-</span> <span class="va">p_train_m</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>set <span class="op">=</span> <span class="st">"train"</span><span class="op">)</span></span>
<span><span class="va">p_test_m</span> <span class="op">&lt;-</span> <span class="va">p_test_m</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>set <span class="op">=</span> <span class="st">"test"</span><span class="op">)</span></span>
<span><span class="va">p_all_m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html">bind_rows</a></span><span class="op">(</span><span class="va">p_train_m</span>, <span class="va">p_test_m</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">p_all_m</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">nn1</span>, y<span class="op">=</span><span class="va">nn2</span>, </span>
<span>                     colour<span class="op">=</span><span class="va">species</span>, shape<span class="op">=</span><span class="va">set</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_divergingx.html">scale_colour_discrete_divergingx</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Zissou 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_shape_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">16</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>aspect.ratio<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-hidden-layer" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIXME">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-hidden-layer-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="17-nn_files/figure-html/fig-hidden-layer-1.png" class="img-fluid figure-img" style="width:80.0%" alt="FIXME">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-hidden-layer-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;17.3: Plot of the data in the linear combinations from the two nodes in the hidden layer. The three species are clearly different, although with some overlap between all three. A main issue to notice is that there isn’t a big gap between Gentoo and the other species, which we know is there based on our data exploration done in other chapters. This suggests this fitted model is sub-optimal.
</figcaption></figure>
</div>
</div>
</div>
<p><a href="#fig-hidden-layer" class="quarto-xref">Figure&nbsp;<span>17.3</span></a> shows the data projected into the plane determined by the two linear combinations of the two nodes in the hidden layer. Training and test sets are indicated by empty and solid circles. The three species are clearly different but there is some overlap or confusion for a few penguins. The most interesting aspect to learn is that there is no big gap between the Gentoo and other species, which we know exists in the data. The model has not found this gap, and thus is likely to unfortunately and erroneously confuse some Gentoo penguins, particularly with Adelie.</p>
<p>What we have shown here is a process to use the models at the nodes of the hidden layer to produce a reduced dimensional space where the classes are best separated, at least as determined by the model. The process will work in higher dimensions also.</p>
<p>When there are more nodes in the hidden layer than the number of original variables it means that the space is extended to achieve useful classifications that need more complicated non-linear boundaries. The extra nodes describe the non-linearity. <span class="citation" data-cites="wickham2015">Wickham et al. (<a href="references.html#ref-wickham2015" role="doc-biblioref">2015</a>)</span> provides a good illustration of this in 2D. The process of examining each of the node models can be useful for understanding this non-linear separation, also in high dimensions.</p>
</section><section id="examining-predictive-probabilities" class="level2" data-number="17.5"><h2 data-number="17.5" class="anchored" data-anchor-id="examining-predictive-probabilities">
<span class="header-section-number">17.5</span> Examining predictive probabilities</h2>
<p></p>
<p>When the predictive probabilities are returned by a model, as is done by this NN, we can use a ternary diagram for three class problems, or high-dimensional simplex when there are more classes to examine the strength of the classification. This done in the same way that was used for the votes matrix from a random forest in <a href="15-forests.html#sec-votes" class="quarto-xref"><span>Section 15.2.1</span></a>.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Predict training and test set</span></span>
<span><span class="va">p_train_pred</span> <span class="op">&lt;-</span> <span class="va">p_nn_model</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">p_train_x</span>, verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">p_train_pred_cat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">p_train</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">[</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">p_train_pred</span>, <span class="fl">1</span>,</span>
<span>        <span class="va">which.max</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">p_train_pred_cat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span></span>
<span>  <span class="va">p_train_pred_cat</span>,</span>
<span>  levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">p_train</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">p_train</span><span class="op">$</span><span class="va">species</span>, <span class="va">p_train_pred_cat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>           p_train_pred_cat
            Adelie Chinstrap Gentoo
  Adelie        92         4      1
  Chinstrap      0        45      0
  Gentoo         1         0     78</code></pre>
</div>
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p_test_pred</span> <span class="op">&lt;-</span> <span class="va">p_nn_model</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">p_test_x</span>, verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">p_test_pred_cat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">p_test</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">[</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">p_test_pred</span>, <span class="fl">1</span>, </span>
<span>        <span class="va">which.max</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">p_test_pred_cat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span></span>
<span>  <span class="va">p_test_pred_cat</span>,</span>
<span>  levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">p_test</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">p_test</span><span class="op">$</span><span class="va">species</span>, <span class="va">p_test_pred_cat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>           p_test_pred_cat
            Adelie Chinstrap Gentoo
  Adelie        45         3      1
  Chinstrap      0        23      0
  Gentoo         1         0     39</code></pre>
</div>
</div>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Set up the data to make the ternary diagram</span></span>
<span><span class="co"># Join data sets</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">p_train_pred</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Adelie"</span>, <span class="st">"Chinstrap"</span>, <span class="st">"Gentoo"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">p_test_pred</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Adelie"</span>, <span class="st">"Chinstrap"</span>, <span class="st">"Gentoo"</span><span class="op">)</span></span>
<span><span class="va">p_train_pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">p_train_pred</span><span class="op">)</span></span>
<span><span class="va">p_train_m</span> <span class="op">&lt;-</span> <span class="va">p_train_m</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>pspecies <span class="op">=</span> <span class="va">p_train_pred_cat</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html">bind_cols</a></span><span class="op">(</span><span class="va">p_train_pred</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>set <span class="op">=</span> <span class="st">"train"</span><span class="op">)</span></span>
<span><span class="va">p_test_pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">p_test_pred</span><span class="op">)</span></span>
<span><span class="va">p_test_m</span> <span class="op">&lt;-</span> <span class="va">p_test_m</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>pspecies <span class="op">=</span> <span class="va">p_test_pred_cat</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html">bind_cols</a></span><span class="op">(</span><span class="va">p_test_pred</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>set <span class="op">=</span> <span class="st">"test"</span><span class="op">)</span></span>
<span><span class="va">p_all_m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html">bind_rows</a></span><span class="op">(</span><span class="va">p_train_m</span>, <span class="va">p_test_m</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add simplex to make ternary</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://schloerke.github.io/geozoo/">geozoo</a></span><span class="op">)</span></span>
<span><span class="va">proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">geozoo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/geozoo/man/f_helmert.html">f_helmert</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="va">p_nn_v_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">p_all_m</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Adelie"</span>, <span class="st">"Chinstrap"</span>, <span class="st">"Gentoo"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">proj</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">p_nn_v_p</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x1"</span>, <span class="st">"x2"</span><span class="op">)</span></span>
<span><span class="va">p_nn_v_p</span> <span class="op">&lt;-</span> <span class="va">p_nn_v_p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>species <span class="op">=</span> <span class="va">p_all_m</span><span class="op">$</span><span class="va">species</span>,</span>
<span>         set <span class="op">=</span> <span class="va">p_all_m</span><span class="op">$</span><span class="va">set</span><span class="op">)</span></span>
<span></span>
<span><span class="va">simp</span> <span class="op">&lt;-</span> <span class="fu">geozoo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/geozoo/man/simplex.html">simplex</a></span><span class="op">(</span>p<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">simp</span><span class="op">$</span><span class="va">points</span><span class="op">)</span>, <span class="va">simp</span><span class="op">$</span><span class="va">points</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">1</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x1"</span>, <span class="st">"x2"</span>, <span class="st">"x3"</span>, <span class="st">"x4"</span><span class="op">)</span></span>
<span><span class="va">sp</span><span class="op">$</span><span class="va">species</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">penguins_sub</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot it</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_segment.html">geom_segment</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">sp</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x1</span>, y<span class="op">=</span><span class="va">x2</span>, xend<span class="op">=</span><span class="va">x3</span>, yend<span class="op">=</span><span class="va">x4</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">sp</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x1</span>, y<span class="op">=</span><span class="va">x2</span>, label<span class="op">=</span><span class="va">species</span><span class="op">)</span>,</span>
<span>            nudge_x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.1</span>, <span class="fl">0.15</span>, <span class="fl">0</span><span class="op">)</span>,</span>
<span>            nudge_y<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="fl">0.05</span>, <span class="op">-</span><span class="fl">0.05</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">p_nn_v_p</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x1</span>, y<span class="op">=</span><span class="va">x2</span>, </span>
<span>                                colour<span class="op">=</span><span class="va">species</span>,</span>
<span>                                shape<span class="op">=</span><span class="va">set</span><span class="op">)</span>, </span>
<span>             size<span class="op">=</span><span class="fl">2</span>, alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_divergingx.html">scale_color_discrete_divergingx</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Zissou 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_shape_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">19</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggthemes/man/theme_map.html">theme_map</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>aspect.ratio<span class="op">=</span><span class="fl">1</span>, legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="17-nn_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption>Ternary diagram for the three groups of the predictive probabilities of both training ans test sets. From what we already know about the penguins data this fit is not good. Both Chinstrap and Gentoo penguins are confused with Adelie, or at risk of it. Gentoo is very well-separated from the other two species when several variables are used, and this fitted model is blind to it. One useful finding is that there is little difference between training and test sets, so the model has not been over-fitted.</figcaption></figure>
</div>
</div>
</div>
<div class="info">
<p>If the training and test sets look similar when plotted in the model space then the model is not suffering from over-fitting.</p>
</div>
</section><section id="local-explanations" class="level2" data-number="17.6"><h2 data-number="17.6" class="anchored" data-anchor-id="local-explanations">
<span class="header-section-number">17.6</span> Local explanations</h2>
<p> </p>
<p>It especially important to be able to interpret or explain a model, even more so when the model is complex or black-box’y. A good resource for learning about the range of methods is <span class="citation" data-cites="iml">Molnar (<a href="references.html#ref-iml" role="doc-biblioref">2022</a>)</span>. Local explanations provide some information about variables that are important for making the prediction for a particular observation. The method that we use here is Shapley value, as computed using the <code>kernelshap</code> package <span class="citation" data-cites="kernelshap">(<a href="references.html#ref-kernelshap" role="doc-biblioref">Mayer &amp; Watson, 2023</a>)</span>.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Explanations</span></span>
<span><span class="co"># https://www.r-bloggers.com/2022/08/kernel-shap/</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ModelOriented/kernelshap">kernelshap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/ModelOriented/shapviz">shapviz</a></span><span class="op">)</span></span>
<span><span class="va">p_explain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/kernelshap/man/kernelshap.html">kernelshap</a></span><span class="op">(</span></span>
<span>    <span class="va">p_nn_model</span>,</span>
<span>    <span class="va">p_train_x</span>, </span>
<span>    bg_X <span class="op">=</span> <span class="va">p_train_x</span>,</span>
<span>    verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">p_exp_sv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/shapviz/man/shapviz.html">shapviz</a></span><span class="op">(</span><span class="va">p_explain</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">p_exp_sv</span>, file<span class="op">=</span><span class="st">"data/p_exp_sv.rda"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>A Shapley value for an observation indicates how the variable contributes to the model prediction for that observation, relative to other variables. It is an average, computed from the change in prediction when all combinations of presence or absence of other variables. In the computation, for each combination, the prediction is computed by substituting absent variables with their average value, like one might do when imputing missing values.</p>
<p><a href="#fig-shapley-pcp" class="quarto-xref">Figure&nbsp;<span>17.4</span></a> shows the Shapley values for Gentoo observations (both training and test sets) in the penguins data, as a parallel coordinate plot. The values for the single misclassified Gentoo penguin (in the training set) is coloured orange. Overall, the Shapley values don’t vary much on <code>bl</code>, <code>bd</code> and <code>fl</code> but they do on <code>bm</code>. The effect of other variables is seems to be only important for <code>bm</code>.</p>
<p>For the misclassified penguin, the effect of <code>bm</code> for all combinations of other variables leads to a decline in predicted value, thus less confidence in it being a Gentoo. In contrast, for this same penguin when considering the effect of <code>bl</code> the predicted value increases on average.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="st">"data/p_exp_sv.rda"</span><span class="op">)</span></span>
<span><span class="va">p_exp_gentoo</span> <span class="op">&lt;-</span> <span class="va">p_exp_sv</span><span class="op">$</span><span class="va">Class_3</span><span class="op">$</span><span class="va">S</span></span>
<span><span class="va">p_exp_gentoo</span> <span class="op">&lt;-</span> <span class="va">p_exp_gentoo</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>species <span class="op">=</span> <span class="va">p_train</span><span class="op">$</span><span class="va">species</span>,</span>
<span>         pspecies <span class="op">=</span> <span class="va">p_train_pred_cat</span>,</span>
<span>  <span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>error <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">species</span> <span class="op">==</span> <span class="va">pspecies</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p_exp_gentoo</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">species</span> <span class="op">==</span> <span class="st">"Gentoo"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span><span class="va">bl</span><span class="op">:</span><span class="va">bm</span>, names_to<span class="op">=</span><span class="st">"var"</span>, values_to<span class="op">=</span><span class="st">"shap"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>var <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">var</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"bl"</span>, <span class="st">"bd"</span>, <span class="st">"fl"</span>, <span class="st">"bm"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">var</span>, y<span class="op">=</span><span class="va">shap</span>, colour<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">error</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_quasirandom.html">geom_quasirandom</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_divergingx.html">scale_colour_discrete_divergingx</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Geyser"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co">#facet_wrap(~var) +</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"SHAP"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb25"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/heike/ggpcp">ggpcp</a></span><span class="op">)</span></span>
<span><span class="va">p_exp_gentoo</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">species</span> <span class="op">==</span> <span class="st">"Gentoo"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/ggpcp/man/pcp_select.html">pcp_select</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ggpcp/man/aes_pcp.html">aes_pcp</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggpcp/man/geom_pcp_axes.html">geom_pcp_axes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggpcp/man/geom_pcp_boxes.html">geom_pcp_boxes</a></span><span class="op">(</span>fill<span class="op">=</span><span class="st">"grey80"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggpcp/man/geom_pcp.html">geom_pcp</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">error</span><span class="op">)</span><span class="op">)</span>, </span>
<span>             linewidth <span class="op">=</span> <span class="fl">2</span>, alpha<span class="op">=</span><span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_divergingx.html">scale_colour_discrete_divergingx</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Geyser"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"SHAP"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<div id="fig-shapley-pcp" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIXME">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-shapley-pcp-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="17-nn_files/figure-html/fig-shapley-pcp-1.png" class="img-fluid figure-img" style="width:80.0%" alt="FIXME">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-shapley-pcp-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;17.4: SHAP values focused on Gentoo class, for each variable. The one misclassified penguin (orange) has a much lower value for body mass, suggesting that this variable is used differently for the prediction than for other penguins.
</figcaption></figure>
</div>
</div>
</div>
<p>If we examine the data <a href="#fig-penguins-bl-bm-bd" class="quarto-xref">Figure&nbsp;<span>17.5</span></a> the explanation makes some sense. The misclassified penguin has an unusually small value on <code>bm</code>. That the SHAP value for <code>bm</code> was quite different pointed to this being a potential issue with the model, particularly for this penguin. This penguin’s prediction is negatively impacted by <code>bm</code> being in the model.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com">patchwork</a></span><span class="op">)</span></span>
<span><span class="co"># Check position on bm</span></span>
<span><span class="va">shap_proj</span> <span class="op">&lt;-</span> <span class="va">p_exp_gentoo</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">species</span> <span class="op">==</span> <span class="st">"Gentoo"</span>, <span class="va">error</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">bl</span><span class="op">:</span><span class="va">bm</span><span class="op">)</span></span>
<span><span class="va">shap_proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">shap_proj</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">sqrt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">shap_proj</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p_exp_gentoo_proj</span> <span class="op">&lt;-</span> <span class="va">p_exp_gentoo</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>shap_bl <span class="op">=</span> <span class="va">bl</span>, </span>
<span>         shap_bd <span class="op">=</span> <span class="va">bd</span>,</span>
<span>         shap_fl <span class="op">=</span> <span class="va">fl</span>, </span>
<span>         shap_bm <span class="op">=</span> <span class="va">bm</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html">bind_cols</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="va">p_train_x</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>shap1 <span class="op">=</span> <span class="va">shap_proj</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">*</span><span class="va">bl</span><span class="op">+</span></span>
<span>           <span class="va">shap_proj</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">*</span><span class="va">bd</span><span class="op">+</span></span>
<span>           <span class="va">shap_proj</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">*</span><span class="va">fl</span><span class="op">+</span></span>
<span>           <span class="va">shap_proj</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span><span class="op">*</span><span class="va">bm</span><span class="op">)</span></span>
<span><span class="va">sp1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">p_exp_gentoo_proj</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">bm</span>, y<span class="op">=</span><span class="va">bl</span>, </span>
<span>             colour<span class="op">=</span><span class="va">species</span>, </span>
<span>             shape<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">error</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_divergingx.html">scale_colour_discrete_divergingx</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Zissou 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_shape_manual</a></span><span class="op">(</span><span class="st">"error"</span>, values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">19</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>aspect.ratio<span class="op">=</span><span class="fl">1</span>, legend.position<span class="op">=</span><span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">sp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">p_exp_gentoo_proj</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">bm</span>, y<span class="op">=</span><span class="va">shap1</span>, </span>
<span>             colour<span class="op">=</span><span class="va">species</span>, </span>
<span>             shape<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">error</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_divergingx.html">scale_colour_discrete_divergingx</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Zissou 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_shape_manual</a></span><span class="op">(</span><span class="st">"error"</span>, values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">19</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"SHAP"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>aspect.ratio<span class="op">=</span><span class="fl">1</span>, legend.position<span class="op">=</span><span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">sp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">p_exp_gentoo_proj</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">shap1</span>, </span>
<span>             fill<span class="op">=</span><span class="va">species</span>, colour<span class="op">=</span><span class="va">species</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="va">p_exp_gentoo_proj</span><span class="op">$</span><span class="va">shap1</span><span class="op">[</span></span>
<span>    <span class="va">p_exp_gentoo_proj</span><span class="op">$</span><span class="va">species</span><span class="op">==</span><span class="st">"Gentoo"</span> <span class="op">&amp;</span></span>
<span>    <span class="va">p_exp_gentoo_proj</span><span class="op">$</span><span class="va">error</span><span class="op">==</span><span class="fl">1</span><span class="op">]</span>, colour<span class="op">=</span><span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_divergingx.html">scale_fill_discrete_divergingx</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Zissou 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_divergingx.html">scale_colour_discrete_divergingx</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Zissou 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>aspect.ratio<span class="op">=</span><span class="fl">1</span>, legend.position<span class="op">=</span><span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">sp2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">p_exp_gentoo_proj</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">bm</span>, y<span class="op">=</span><span class="va">bd</span>, </span>
<span>             colour<span class="op">=</span><span class="va">species</span>, </span>
<span>             shape<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">error</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_divergingx.html">scale_colour_discrete_divergingx</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Zissou 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_shape_manual</a></span><span class="op">(</span><span class="st">"error"</span>, values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">19</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>aspect.ratio<span class="op">=</span><span class="fl">1</span>, legend.position<span class="op">=</span><span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">sp1</span> <span class="op">+</span> <span class="va">sp2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html">plot_layout</a></span><span class="op">(</span>ncol<span class="op">=</span><span class="fl">2</span>, guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position<span class="op">=</span><span class="st">"bottom"</span>,</span>
<span>        legend.direction<span class="op">=</span><span class="st">"vertical"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<div id="fig-penguins-bl-bm-bd" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIXME">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-penguins-bl-bm-bd-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="17-nn_files/figure-html/fig-penguins-bl-bm-bd-1.png" class="img-fluid figure-img" style="width:100.0%" alt="FIXME">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-penguins-bl-bm-bd-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;17.5: Plots of the data to help understand what the SHAP values indicate. The misclassified Gentoo penguin has an unusually low body mass value which makes it appear to be more like an Adelie penguin, particularly when considered in relation to it’s bill length.
</figcaption></figure>
</div>
</div>
</div>
</section><section id="examining-boundaries" class="level2" data-number="17.7"><h2 data-number="17.7" class="anchored" data-anchor-id="examining-boundaries">
<span class="header-section-number">17.7</span> Examining boundaries</h2>
<!-- Check against LDA, suspect that `bm` is used too much in CNN model.-->
<p><a href="#fig-penguins-lda-nn" class="quarto-xref">Figure&nbsp;<span>17.7</span></a> shows the boundaries for this NN model along with those of the LDA model.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Generate grid over explanatory variables</span></span>
<span><span class="va">p_grid</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span></span>
<span>  bl <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">10000</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">penguins_sub</span><span class="op">$</span><span class="va">bl</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">penguins_sub</span><span class="op">$</span><span class="va">bl</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  bd <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">10000</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">penguins_sub</span><span class="op">$</span><span class="va">bd</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">penguins_sub</span><span class="op">$</span><span class="va">bd</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  fl <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">10000</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">penguins_sub</span><span class="op">$</span><span class="va">fl</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">penguins_sub</span><span class="op">$</span><span class="va">fl</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  bm <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">10000</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">penguins_sub</span><span class="op">$</span><span class="va">bm</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">penguins_sub</span><span class="op">$</span><span class="va">bm</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Predict grid</span></span>
<span><span class="va">p_grid_pred</span> <span class="op">&lt;-</span> <span class="va">p_nn_model</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">p_grid</span><span class="op">)</span>, verbose<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">p_grid_pred_cat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">p_train</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">p_grid_pred</span>, <span class="fl">1</span>, <span class="va">which.max</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">p_grid_pred_cat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">p_grid_pred_cat</span>,</span>
<span>                          levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">p_train</span><span class="op">$</span><span class="va">species</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Project into weights from the two nodes</span></span>
<span><span class="va">p_grid_proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">p_grid</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">p_nn_wgts_on</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">p_grid_proj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nn1"</span>, <span class="st">"nn2"</span><span class="op">)</span></span>
<span><span class="va">p_grid_proj</span> <span class="op">&lt;-</span> <span class="va">p_grid_proj</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>species <span class="op">=</span> <span class="va">p_grid_pred_cat</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">p_grid_proj</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">nn1</span>, y<span class="op">=</span><span class="va">nn2</span>, </span>
<span>                     colour<span class="op">=</span><span class="va">species</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>data<span class="op">=</span><span class="va">p_all_m</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">nn1</span>, </span>
<span>                               y<span class="op">=</span><span class="va">nn2</span>, </span>
<span>                               shape<span class="op">=</span><span class="va">species</span><span class="op">)</span>,</span>
<span>             inherit.aes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_divergingx.html">scale_colour_discrete_divergingx</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Zissou 1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_shape_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>aspect.ratio<span class="op">=</span><span class="fl">1</span>, </span>
<span>        legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>        legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="fig-penguins-lda-nn-html" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-penguins-lda-nn-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-penguins-lda-nn-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-lda-boundary" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-lda-boundary-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gifs/penguins_lda_boundaries.gif" class="img-fluid figure-img" data-ref-parent="fig-penguins-lda-nn-html" alt="FIX ME" width="300">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-lda-boundary-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) LDA model
</figcaption></figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-penguins-lda-nn-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-tree-boundary" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-tree-boundary-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gifs/penguins_nn_boundaries.gif" class="img-fluid figure-img" data-ref-parent="fig-penguins-lda-nn-html" alt="FIX ME" width="300">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-tree-boundary-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) NN model
</figcaption></figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-penguins-lda-nn-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;17.6: Comparison of the boundaries produced by the LDA (a) and the NN (b) model, using a slice tour.
</figcaption></figure>
</div>
<div id="fig-penguins-lda-nn" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-penguins-lda-nn-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-penguins-lda-nn" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-lda-boundary2" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-lda-boundary2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/fig-lda-2D-boundaries-1.png" class="img-fluid figure-img" data-ref-parent="fig-penguins-lda-nn" alt="FIX ME" width="200">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-lda-boundary2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) LDA model
</figcaption></figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-penguins-lda-nn" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-nn-boundary" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-nn-boundary-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="images/penguins-nn-boundaries-1.png" class="img-fluid figure-img" data-ref-parent="fig-penguins-lda-nn" alt="FIX ME" width="290">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-nn-boundary-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) NN model
</figcaption></figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-penguins-lda-nn-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;17.7: Comparison of the boundaries produced by the LDA (a) and the NN (b) model, using a slice tour.
</figcaption></figure>
</div>
<p></p>
</section><section id="application-to-a-large-dataset" class="level2" data-number="17.8"><h2 data-number="17.8" class="anchored" data-anchor-id="application-to-a-large-dataset">
<span class="header-section-number">17.8</span> Application to a large dataset</h2>
<p>To see how these methods apply in the setting where we have a large number of variables, observations and classes we will look at a neural network that predicts the category for the fashion MNIST data. The code for designing and fitting the model is following the tutorial available from https://tensorflow.rstudio.com/tutorials/keras/classification and you can find additional information there. Below we only replicate the steps needed to build the model from scratch. We also note that a similar investigation was presented in <span class="citation" data-cites="li2020visualizing">Li et al. (<a href="references.html#ref-li2020visualizing" role="doc-biblioref">2020</a>)</span>, with a focus on investigating the model at different epochs during the training. </p>
<p>The first step is to download and prepare the data. Here we scale the observations to range between zero and one, and we define the label names.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tensorflow.rstudio.com/">keras</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># download the data</span></span>
<span><span class="va">fashion_mnist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/keras/man/dataset_fashion_mnist.html">dataset_fashion_mnist</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># split into input variables and response</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">train_images</span>, <span class="va">train_labels</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/zeallot/man/operator.html">%&lt;-%</a></span> <span class="va">fashion_mnist</span><span class="op">$</span><span class="va">train</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">test_images</span>, <span class="va">test_labels</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/zeallot/man/operator.html">%&lt;-%</a></span> <span class="va">fashion_mnist</span><span class="op">$</span><span class="va">test</span></span>
<span></span>
<span><span class="co"># for interpretation we also define the category names</span></span>
<span><span class="va">class_names</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'T-shirt/top'</span>,</span>
<span>                <span class="st">'Trouser'</span>,</span>
<span>                <span class="st">'Pullover'</span>,</span>
<span>                <span class="st">'Dress'</span>,</span>
<span>                <span class="st">'Coat'</span>,</span>
<span>                <span class="st">'Sandal'</span>,</span>
<span>                <span class="st">'Shirt'</span>,</span>
<span>                <span class="st">'Sneaker'</span>,</span>
<span>                <span class="st">'Bag'</span>,</span>
<span>                <span class="st">'Ankle boot'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># rescaling to the range (0,1)</span></span>
<span><span class="va">train_images</span> <span class="op">&lt;-</span> <span class="va">train_images</span> <span class="op">/</span> <span class="fl">255</span></span>
<span><span class="va">test_images</span> <span class="op">&lt;-</span> <span class="va">test_images</span> <span class="op">/</span> <span class="fl">255</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>In the next step we define the neural network and train the model. Note that because we have many observations, even a very simple structure returns a good model. And because this example is well-known, we do not need to tune the model or check the validation accuracy.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># defining the model</span></span>
<span><span class="va">model_fashion_mnist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/keras/man/keras_model_sequential.html">keras_model_sequential</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">model_fashion_mnist</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># flatten the image data into a long vector</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/keras/man/layer_flatten.html">layer_flatten</a></span><span class="op">(</span>input_shape <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">28</span>, <span class="fl">28</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># hidden layer with 128 units</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/keras/man/layer_dense.html">layer_dense</a></span><span class="op">(</span>units <span class="op">=</span> <span class="fl">128</span>, activation <span class="op">=</span> <span class="st">'relu'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="co"># output layer for 10 categories</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/keras/man/layer_dense.html">layer_dense</a></span><span class="op">(</span>units <span class="op">=</span> <span class="fl">10</span>, activation <span class="op">=</span> <span class="st">'softmax'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">model_fashion_mnist</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://generics.r-lib.org/reference/compile.html">compile</a></span><span class="op">(</span></span>
<span>  optimizer <span class="op">=</span> <span class="st">'adam'</span>,</span>
<span>  loss <span class="op">=</span> <span class="st">'sparse_categorical_crossentropy'</span>,</span>
<span>  metrics <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'accuracy'</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># fitting the model, if we did not know the model yet we</span></span>
<span><span class="co"># would add a validation split to diagnose the training</span></span>
<span><span class="va">model_fashion_mnist</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://generics.r-lib.org/reference/fit.html">fit</a></span><span class="op">(</span><span class="va">train_images</span>,</span>
<span>              <span class="va">train_labels</span>,</span>
<span>              epochs <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/keras/man/save_model_tf.html">save_model_tf</a></span><span class="op">(</span><span class="va">model_fashion_mnist</span>, <span class="st">"data/fashion_nn"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We have defined a flat neural network with a single hidden layer with 128 nodes. To investigate the model we can start by comparing the activations to the original input data distribution. Since both the input space and the space of activations is large, and they are of different dimensionality, we will first use principal component analysis. This simplifies the analysis, and in general we do not need the original pixel or hidden node information for the interpretation here. The comparison is using the test-subset of the data.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># get the fitted model</span></span>
<span><span class="va">model_fashion_mnist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/keras/man/save_model_tf.html">load_model_tf</a></span><span class="op">(</span><span class="st">"data/fashion_nn"</span><span class="op">)</span></span>
<span><span class="co"># observed response labels in the test set</span></span>
<span><span class="va">test_tags</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">class_names</span><span class="op">[</span><span class="va">test_labels</span> <span class="op">+</span> <span class="fl">1</span><span class="op">]</span>,</span>
<span>                    levels <span class="op">=</span> <span class="va">class_names</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># calculate activation for the hidden layer, this can be done</span></span>
<span><span class="co"># within the keras framework</span></span>
<span><span class="va">activations_model_fashion</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/keras/man/keras_model.html">keras_model</a></span><span class="op">(</span></span>
<span>  inputs <span class="op">=</span> <span class="va">model_fashion_mnist</span><span class="op">$</span><span class="va">input</span>,</span>
<span>  outputs <span class="op">=</span> <span class="va">model_fashion_mnist</span><span class="op">$</span><span class="va">layers</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">output</span></span>
<span><span class="op">)</span></span>
<span><span class="va">activations_fashion</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span></span>
<span>  <span class="va">activations_model_fashion</span>,</span>
<span>  <span class="va">test_images</span>, verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># PCA for activations</span></span>
<span><span class="va">activations_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span><span class="va">activations_fashion</span><span class="op">)</span></span>
<span><span class="va">activations_pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">activations_pca</span><span class="op">$</span><span class="va">x</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># PCA on the original data</span></span>
<span><span class="co"># we first need to flatten the image input</span></span>
<span><span class="va">test_images_flat</span> <span class="op">&lt;-</span> <span class="va">test_images</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">test_images_flat</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">test_images_flat</span><span class="op">)</span>, <span class="fl">784</span><span class="op">)</span></span>
<span><span class="va">images_pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">test_images_flat</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">images_pc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="va">images_pca</span><span class="op">$</span><span class="va">x</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold"><summary>Code to run tours</summary><div class="sourceCode" id="cb31"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">activations_pc</span>,</span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">PC1</span>, <span class="va">PC2</span>, color <span class="op">=</span> <span class="va">test_tags</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Activations"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_qualitative.html">scale_color_discrete_qualitative</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dynamic"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span>, aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">images_pc</span>,</span>
<span>       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">PC1</span>, <span class="va">PC2</span>, color <span class="op">=</span> <span class="va">test_tags</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Input space"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://colorspace.R-Forge.R-project.org/reference/scale_colour_discrete_qualitative.html">scale_color_discrete_qualitative</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dynamic"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span>, aspect.ratio <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="va">legend_labels</span> <span class="op">&lt;-</span> <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/get_legend.html">get_legend</a></span><span class="op">(</span></span>
<span>  <span class="va">p1</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>          legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="op">)</span></span>
<span><span class="co"># hide plotting code</span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span>, <span class="va">legend_labels</span>,</span>
<span>                   rel_heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">.3</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output-display">
<div>
<figure class="figure"><p><img src="17-nn_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Looking only at the first two principal components we note some clear differences from the transformation in the hidden layer. The observations seem to be more evenly spread in the input space, while in the activations space we notice grouping along specific directions. In particular the category “Bag” appears to be most different from all other classes, and the non-linear transformation in the activations space shows that they are clearly different from the shoe categories, while in the input space we could note some overlap in the linear projection. To better identify differences between other groups we will use the tour on the first five principal components.</p>
<div class="cell">
<details class="code-fold"><summary>Code to run tours</summary><div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">animate_xy</span><span class="op">(</span><span class="va">images_pc</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, col <span class="op">=</span> <span class="va">test_tags</span>,</span>
<span>        cex<span class="op">=</span><span class="fl">0.2</span>, palette <span class="op">=</span> <span class="st">"Dynamic"</span><span class="op">)</span></span>
<span><span class="fu">animate_xy</span><span class="op">(</span><span class="va">activations_pc</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>, col <span class="op">=</span> <span class="va">test_tags</span>,</span>
<span>        cex<span class="op">=</span><span class="fl">0.2</span>, palette <span class="op">=</span> <span class="st">"Dynamic"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">render_gif</span><span class="op">(</span><span class="va">images_pc</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>           <span class="fu">grand_tour</span><span class="op">(</span><span class="op">)</span>,</span>
<span>           <span class="fu">display_xy</span><span class="op">(</span> </span>
<span>             col<span class="op">=</span><span class="va">test_tags</span>, </span>
<span>             cex<span class="op">=</span><span class="fl">0.2</span>,</span>
<span>             palette <span class="op">=</span> <span class="st">"Dynamic"</span>,</span>
<span>             axes<span class="op">=</span><span class="st">"bottomleft"</span><span class="op">)</span>, </span>
<span>           gif_file<span class="op">=</span><span class="st">"gifs/fashion_images_gt.gif"</span>,</span>
<span>           frames<span class="op">=</span><span class="fl">500</span>,</span>
<span>           loop<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="fu">render_gif</span><span class="op">(</span><span class="va">activations_pc</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>,</span>
<span>           <span class="fu">grand_tour</span><span class="op">(</span><span class="op">)</span>,</span>
<span>           <span class="fu">display_xy</span><span class="op">(</span> </span>
<span>             col<span class="op">=</span><span class="va">test_tags</span>, </span>
<span>             cex<span class="op">=</span><span class="fl">0.2</span>,</span>
<span>             palette <span class="op">=</span> <span class="st">"Dynamic"</span>,</span>
<span>             axes<span class="op">=</span><span class="st">"bottomleft"</span><span class="op">)</span>, </span>
<span>           gif_file<span class="op">=</span><span class="st">"gifs/fashion_activations_gt.gif"</span>,</span>
<span>           frames<span class="op">=</span><span class="fl">500</span>,</span>
<span>           loop<span class="op">=</span><span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="fig-fashion-gt-html" class="quarto-layout-panel">
<figure class="quarto-float quarto-float-fig figure"><div aria-describedby="fig-fashion-gt-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="quarto-layout-row">
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-fashion-gt-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-fashion-input" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-fashion-input-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gifs/fashion_images_gt.gif" class="img-fluid figure-img" data-ref-parent="fig-fashion-gt-html" alt="FIX ME" width="200">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-fashion-input-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(a) Input space
</figcaption></figure>
</div>
</div>
<div class="quarto-layout-cell-subref quarto-layout-cell" data-ref-parent="fig-fashion-gt-html" style="flex-basis: 50.0%;justify-content: flex-start;">
<div id="fig-fashion-activation" class="quarto-figure quarto-figure-center quarto-float anchored" alt="FIX ME">
<figure class="quarto-float quarto-subfloat-fig figure"><div aria-describedby="fig-fashion-activation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="gifs/fashion_activations_gt.gif" class="img-fluid figure-img" data-ref-parent="fig-fashion-gt-html" alt="FIX ME" width="200">
</div>
<figcaption class="quarto-float-caption-bottom quarto-subfloat-caption quarto-subfloat-fig" id="fig-fashion-activation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
(b) Activations
</figcaption></figure>
</div>
</div>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-fashion-gt-html-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;17.8: Comparison of the test observations in the first five principal components of the input space (left) and in the hidden layer activations (right). The activation function results in more clearly defined grouping of the different classes.
</figcaption></figure>
</div>
<p>As with the first two principal components we get a much more spread out distribution in the original space. Nevertheless we can see differences between the classes, and that some groups are varying along specific directions in that space. Overall the activations space shows tighter clusters as expected after including the ReLU activation function, but the picture is not as neat as the first two principal components would suggest. While certain groups appear very compact even in this larger subspace, others vary quite a bit within part of the space. For example we can clearly see the “Bag” observations as different from all other images, but also notice that there is a large variation within this class along certain directions.</p>
<p>Finally we will investigate the model performance through the misclassifications and uncertainty between classes. We start with the error matrix for the test observations. To fit the error matrix we use the numeric labels, the ordering is as defined above for the labels.</p>
<div class="cell">
<details class="code-fold"><summary>Code</summary><div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">fashion_test_pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">model_fashion_mnist</span>,</span>
<span>                             <span class="va">test_images</span>, verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">fashion_test_pred_cat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">test_tags</span><span class="op">)</span><span class="op">[</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">fashion_test_pred</span>, <span class="fl">1</span>,</span>
<span>        <span class="va">which.max</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">predicted</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span></span>
<span>  <span class="va">fashion_test_pred_cat</span>,</span>
<span>  levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">test_tags</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span></span>
<span><span class="va">observed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">test_tags</span><span class="op">)</span> <span class="op">-</span><span class="fl">1</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">observed</span>, <span class="va">predicted</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details><div class="cell-output cell-output-stdout">
<pre><code>        predicted
observed    0    1    2    3    4    5    6    7    8    9
       0  128    0   84  719   10    0   51    7    1    0
       1    0   42   15  941    0    0    0    2    0    0
       2    6    0  824   38  121    0    3    8    0    0
       3    1    0   14  976    8    0    1    0    0    0
       4    1    0  205  181  605    0    0    8    0    0
       5    1    0    0    2    0   77    0  902    1   17
       6   24    0  231  282  405    0   44   12    2    0
       7    0    0    0    0    0    0    0 1000    0    0
       8   17    1   78  102   49    0    5  730   16    2
       9    0    0    0    2    0    0    0  947    0   51</code></pre>
</div>
</div>
<p>From this we see that the model mainly confuses certain categories with each other, and within expected groups (e.g.&nbsp;different types of shoes can be confused with each other, or different types of shirts). We can further investigate this by visualizing the full probability matrix for the test observations, to see which categories the model is uncertain about.</p>
<div class="cell">
<details class="code-fold"><summary>Code to visualize probabilities</summary><div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># getting the probabilities from the output layer</span></span>
<span><span class="va">fashion_test_pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span><span class="op">(</span><span class="va">model_fashion_mnist</span>,</span>
<span>                             <span class="va">test_images</span>, verbose <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># copying this from RF fake tree vote matrix</span></span>
<span><span class="va">proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu">geozoo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/geozoo/man/f_helmert.html">f_helmert</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="va">f_nn_v_p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">fashion_test_pred</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/matmult.html">%*%</a></span> <span class="va">proj</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">f_nn_v_p</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x1"</span>, <span class="st">"x2"</span>, <span class="st">"x3"</span>, <span class="st">"x4"</span>, <span class="st">"x5"</span>, <span class="st">"x6"</span>, <span class="st">"x7"</span>, <span class="st">"x8"</span>, <span class="st">"x9"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">f_nn_v_p</span> <span class="op">&lt;-</span> <span class="va">f_nn_v_p</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>class <span class="op">=</span> <span class="va">test_tags</span><span class="op">)</span></span>
<span></span>
<span><span class="va">simp</span> <span class="op">&lt;-</span> <span class="fu">geozoo</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/geozoo/man/simplex.html">simplex</a></span><span class="op">(</span>p<span class="op">=</span><span class="fl">9</span><span class="op">)</span></span>
<span><span class="va">sp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">simp</span><span class="op">$</span><span class="va">points</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">sp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"x1"</span>, <span class="st">"x2"</span>, <span class="st">"x3"</span>, <span class="st">"x4"</span>, <span class="st">"x5"</span>, <span class="st">"x6"</span>, <span class="st">"x7"</span>, <span class="st">"x8"</span>, <span class="st">"x9"</span><span class="op">)</span></span>
<span><span class="va">sp</span><span class="op">$</span><span class="va">class</span> <span class="op">=</span> <span class="st">""</span></span>
<span><span class="va">f_nn_v_p_s</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html">bind_rows</a></span><span class="op">(</span><span class="va">sp</span>, <span class="va">f_nn_v_p</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">class</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"T-shirt/top"</span>,</span>
<span>                                     <span class="st">"Pullover"</span>,</span>
<span>                                     <span class="st">"Shirt"</span>,</span>
<span>                                     <span class="st">"Coat"</span><span class="op">)</span>, <span class="va">class</span>, <span class="st">"Other"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>class <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">class</span>, levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Other"</span>,</span>
<span>                                        <span class="st">"T-shirt/top"</span>,</span>
<span>                                        <span class="st">"Pullover"</span>,</span>
<span>                                        <span class="st">"Shirt"</span>,</span>
<span>                                        <span class="st">"Coat"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="co"># nicely shows confusion between certain classes is common</span></span>
<span><span class="fu">animate_xy</span><span class="op">(</span><span class="va">f_nn_v_p_s</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">9</span><span class="op">]</span>, col <span class="op">=</span> <span class="va">f_nn_v_p_s</span><span class="op">$</span><span class="va">class</span>, </span>
<span>           axes <span class="op">=</span> <span class="st">"off"</span>, pch <span class="op">=</span> <span class="st">"."</span>,</span>
<span>           edges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="va">simp</span><span class="op">$</span><span class="va">edges</span><span class="op">)</span>,</span>
<span>           edges.width <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>           palette <span class="op">=</span> <span class="st">"Lajolla"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>For this data using explainers like SHAP is not so interesting, since the individual pixel contribution to a prediction are typically not of interest. With image classification a next step might be to further investigate which part of the image is important for a prediction, and this can be visualized as a heat map placed over the original image. This is especially interesting in the case of difficult or misclassified images. This however is beyond the scope of this book.</p>
<!--
This chapter will likely include:

- Models at nodes or epochs 
- Predictions (like vote matrix) for training and test. It provides a visual guide to overfitting.
- Classification boundaries comparison with other methods
- Explainability and interpretability

(This paper https://distill.pub/2020/grand-tour/ has good examples)

NOTE: Results might vary with different knits 

References keras/tensorflow book, tidymodels, interpretable machine learning, and removing the blindfold


Outline for chapter:

- Penguins data
    - Setting up with the NN with keras
    - Splitting training and test, checking
    - Specifying model, choices of layers
    - Making predictions
    - Extracting weights
    - Building your diagnostic data set
    - Examine predictive probabilities with simplex
    - Examining the nodes - like the discriminant space when its 2
    - Misclassifications
    - 

Bushfires
    - Model fitting when regularisation needed
    - Overfitting
    
Sketches as an example  
-->
</section><section id="exercises" class="level2 unnumbered"><h2 class="unnumbered anchored" data-anchor-id="exercises">Exercises</h2>
<ol type="1">
<li>The problem with the NN model fitted to the penguins is that the Gentoo are poorly classified, when they should be perfectly predictable due to the big gap between class clusters. Re-fit the NN to the penguins data, to find a better model that appropriately perfectly predicts Gentoo penguins. Support this by plotting the model (using the hidden layer), and the predictive probabilities as a ternary plot. Do the SHAP values also support that <code>bd</code> plays a stronger role in your best model? (<code>bd</code> is the main variable for distinguishing Gentoo’s from the other species, particularly when used with <code>fl</code> or <code>bl</code>.)</li>
<li>For the fashion MNIST data we have seen that certain categories are more likely to be confused with each other. Select a subset of the data including only the categories Ankle boot, Sneaker and Sandal and see if you can reproduce the analysis of the penguins data in this chapter with this subset.</li>
<li>XXX fake trees, can we think about the number of nodes and make it work with a simple NN similar to penguins data?</li>
<li>The sketches data could also be considered a classic image classification problem, and we have seen that we can get a reasonable accuracy with a random forest model. Because we only have a smaller number of observations (compared to the fashion MNIST data) when fitting a neural network we need to be very careful not to overfit the training data. Try fitting a flat neural network (similar to what we did for the fashion MNIST data) and check the test accuracy of the model.</li>
<li>Challenge: try to design a more accurate neural network for the sketches data. Here you can investigate using a convolutional neural network in combination with data augmentation. In addition, using batch normalization should improve the model performance.</li>
</ol>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" data-line-spacing="2" role="list" style="display: none">
<div id="ref-Ab1884" class="csl-entry" role="listitem">
Abbott, E. (1884). <em>Flatland: A romance of many dimensions</em>. Dover Publications.
</div>
<div id="ref-AWS91" class="csl-entry" role="listitem">
Ahlberg, C., Williamson, C., &amp; Shneiderman, B. (1991). Dynamic <span>Q</span>ueries for <span>I</span>nformation <span>E</span>xploration: <span>A</span>n <span>I</span>mplementation and <span>E</span>valuation. <em>ACM CHI ‘92 Conference Proceedings</em>, 619–626.
</div>
<div id="ref-keras" class="csl-entry" role="listitem">
Allaire, J., &amp; Chollet, F. (2023). <em>Keras: R interface to ’keras’</em>. <a href="https://CRAN.R-project.org/package=keras">https://CRAN.R-project.org/package=keras</a>
</div>
<div id="ref-An57" class="csl-entry" role="listitem">
Anderson, E. (1957). A <span>S</span>emigraphical <span>M</span>ethod for the <span>A</span>nalysis of <span>C</span>omplex <span>P</span>roblems. <em>Proceedings of the National Academy of Science, 13</em>, 923–927.
</div>
<div id="ref-An72" class="csl-entry" role="listitem">
Andrews, D. F. (1972). <span>P</span>lots of <span>H</span>igh-dimensional <span>D</span>ata. <em>Biometrics</em>, <em>28</em>, 125–136.
</div>
<div id="ref-AGW71" class="csl-entry" role="listitem">
Andrews, D. F., Gnanadesikan, R., &amp; Warner, J. L. (1971). <span>T</span>ransformations of <span>M</span>ultivariate <span>D</span>ata. <em>Biometrics</em>, <em>27</em>, 825–840.
</div>
<div id="ref-AB97" class="csl-entry" role="listitem">
Anselin, L., &amp; Bao, S. (1997). <span>E</span>xploratory <span>S</span>patial <span>D</span>ata <span>A</span>nalysis <span>L</span>inking <span>S</span>pace<span>S</span>tat and <span>A</span>rc<span>V</span>iew. In M. M. Fischer &amp; A. Getis (Eds.), <em><span>R</span>ecent <span>D</span>evelopments in <span>S</span>patial <span>A</span>nalysis</em> (pp. 35–59). Springer.
</div>
<div id="ref-R-ggthemes" class="csl-entry" role="listitem">
Arnold, J. B. (2021). <em>Ggthemes: Extra themes, scales and geoms for ggplot2</em>. <a href="https://github.com/jrnold/ggthemes">https://github.com/jrnold/ggthemes</a>
</div>
<div id="ref-ASA23" class="csl-entry" role="listitem">
ASA Statistical Graphics Section. (2023). <em>Video <span>L</span>ibrary</em>. <a href="https://community.amstat.org/jointscsg-section/media/videos" class="uri">https://community.amstat.org/jointscsg-section/media/videos</a>.
</div>
<div id="ref-As85" class="csl-entry" role="listitem">
Asimov, D. (1985). <span>T</span>he <span>G</span>rand <span>T</span>our: <span>A</span> <span>T</span>ool for <span>V</span>iewing <span>M</span>ultidimensional <span>D</span>ata. <em>SIAM Journal of Scientific and Statistical Computing</em>, <em>6</em>(1), 128–143.
</div>
<div id="ref-R-gridExtra" class="csl-entry" role="listitem">
Auguie, B. (2017). <em>gridExtra: Miscellaneous functions for "grid" graphics</em>. <a href="https://CRAN.R-project.org/package=gridExtra">https://CRAN.R-project.org/package=gridExtra</a>
</div>
<div id="ref-forest" class="csl-entry" role="listitem">
Australian Bureau of Agricultural and Resource Economics and Sciences. (2018). <em><span class="nocase">Forests of Australia</span></em>. <a href="https://www.agriculture.gov.au/abares/forestsaustralia/forest-data-maps-and-tools/spatial-data/forest-cover">https://www.agriculture.gov.au/abares/forestsaustralia/forest-data-maps-and-tools/spatial-data/forest-cover</a>
</div>
<div id="ref-BC84" class="csl-entry" role="listitem">
Becker, R. A., &amp; Chambers, J. M. (1984). <em><span>S</span>: An environment for data analysis and graphics</em>. Wadsworth.
</div>
<div id="ref-BC87" class="csl-entry" role="listitem">
Becker, R. A., &amp; Cleveland, W. S. (1988). <span>B</span>rushing <span>S</span>catterplots. In W. S. Cleveland &amp; M. E. McGill (Eds.), <em>Dynamic graphics for statistics</em> (pp. 201–224). Wadsworth.
</div>
<div id="ref-BeCS96" class="csl-entry" role="listitem">
Becker, R., Cleveland, W. S., &amp; Shyu, M.-J. (1996). <span>T</span>he <span>V</span>isual <span>D</span>esign and <span>C</span>ontrol of <span>T</span>rellis <span>D</span>isplays. <em>Journal of Computational and Graphical Statistics</em>, <em>6</em>(1), 123–155.
</div>
<div id="ref-BS03" class="csl-entry" role="listitem">
Bederson, B. B., &amp; Schneiderman, B. (2003). <em>The craft of information visualization: Readings and reflections</em>. Morgan Kaufmann.
</div>
<div id="ref-BellmanRichard1961" class="csl-entry" role="listitem">
Bellman, R. (1961). <em>Adaptive control processes : A guided tour</em>.
</div>
<div id="ref-bickel2018" class="csl-entry" role="listitem">
Bickel, P. J., Kur, G., &amp; Nadler, B. (2018). Projection pursuit in high dimensions. <em>Proceedings of the National Academy of Sciences</em>, <em>115</em>, 9151–9156. <a href="https://doi.org/10.1073/pnas.1801177115">https://doi.org/10.1073/pnas.1801177115</a>
</div>
<div id="ref-Bi06" class="csl-entry" role="listitem">
Bishop, C. M. (2006). <em>Pattern <span>R</span>ecognition and <span>M</span>achine <span>L</span>earning</em>. Springer.
</div>
<div id="ref-HOML" class="csl-entry" role="listitem">
Boehmke, B., &amp; Greenwell, B. M. (2019). <em>Hands-on machine learning with r (1st ed.)</em>. Chapman; Hall/CRC. <a href="https://doi.org/10.1201/9780367816377">https://doi.org/10.1201/9780367816377</a>
</div>
<div id="ref-R-aweSOM" class="csl-entry" role="listitem">
Boelaert, J., Ollion, E., &amp; Sodoge, J. (2022). <em>aweSOM: Interactive self-organizing maps</em>. <a href="https://CRAN.R-project.org/package=aweSOM">https://CRAN.R-project.org/package=aweSOM</a>
</div>
<div id="ref-Bonneau2006" class="csl-entry" role="listitem">
Bonneau, G.-P., Ertl, T., &amp; Nielson, G. M. (Eds.). (2006). <em>Scientific visualization: The visual extraction of knowledge from data</em>. Springer.
</div>
<div id="ref-BG05" class="csl-entry" role="listitem">
Borg, I., &amp; Groenen, P. J. F. (2005). <em>Modern <span>M</span>ultidimensional <span>S</span>caling</em>. Springer.
</div>
<div id="ref-Br01" class="csl-entry" role="listitem">
Breiman, L. (2001). Random <span>F</span>orests. <em>Machine Learning</em>, <em>45</em>(1), 5–32.
</div>
<div id="ref-R-randomForest" class="csl-entry" role="listitem">
Breiman, L., Cutler, A., Liaw, A., &amp; Wiener, M. (2022). <em>randomForest: Breiman and cutler’s random forests for classification and regression</em>. <a href="https://www.stat.berkeley.edu/~breiman/RandomForests/">https://www.stat.berkeley.edu/~breiman/RandomForests/</a>
</div>
<div id="ref-BFOS84" class="csl-entry" role="listitem">
Breiman, L., Friedman, J., Olshen, C., &amp; Stone, C. (1984). <em>Classification and <span>R</span>egression <span>T</span>rees</em>. Wadsworth; Brooks/Cole.
</div>
<div id="ref-BujaKS96" class="csl-entry" role="listitem">
Buja, A. (1996). Interactive <span>G</span>raphical <span>M</span>ethods in the <span>A</span>nalysis of <span>C</span>ustomer <span>P</span>anel <span>D</span>ata: <span>C</span>omment. <em>Journal of Business &amp; Economic Statistics</em>, <em>14</em>(1), 128–129.
</div>
<div id="ref-BA86b" class="csl-entry" role="listitem">
Buja, A., &amp; Asimov, D. (1986). <span>G</span>rand <span>T</span>our <span>M</span>ethods: <span>A</span>n <span>O</span>utline. <em>Computing Science and Statistics</em>, <em>17</em>, 63–67.
</div>
<div id="ref-BAHM88" class="csl-entry" role="listitem">
Buja, A., Asimov, D., Hurley, C., &amp; McDonald, J. A. (1988). <span>E</span>lements of a <span>V</span>iewing <span>P</span>ipeline for <span>D</span>ata <span>A</span>nalysis. In W. S. Cleveland &amp; M. E. McGill (Eds.), <em>Dynamic graphics for statistics</em> (pp. 277–308). Wadsworth.
</div>
<div id="ref-BCAH95c" class="csl-entry" role="listitem">
Buja, A., Cook, D., Asimov, D., &amp; Hurley, C. (1997). <em>Dynamic <span>P</span>rojections in <span>H</span>igh-<span>D</span>imensional <span>V</span>isualization: <span>T</span>heory and <span>C</span>omputational <span>M</span>ethods</em>. AT&amp;<span>T</span> Labs.
</div>
<div id="ref-BCAH05" class="csl-entry" role="listitem">
Buja, A., Cook, D., Asimov, D., &amp; Hurley, C. (2005). <span>C</span>omputational <span>M</span>ethods for <span>H</span>igh-<span>D</span>imensional <span>R</span>otations in <span>D</span>ata <span>V</span>isualization. In C. R. Rao, E. J. Wegman, &amp; J. L. Solka (Eds.), <em>Handbook of statistics: Data mining and visualization</em> (pp. 391–414). Elsevier/North-Holland.
</div>
<div id="ref-BCS95" class="csl-entry" role="listitem">
Buja, A., Cook, D., &amp; Swayne, D. (1996). <span>I</span>nteractive <span>H</span>igh-<span>D</span>imensional <span>D</span>ata <span>V</span>isualization. <em>Journal of Computational and Graphical Statistics</em>, <em>5</em>(1), 78–99.
</div>
<div id="ref-BHM86" class="csl-entry" role="listitem">
Buja, A., Hurley, C., &amp; McDonald, J. A. (1986). A <span>D</span>ata <span>V</span>iewer for <span>M</span>ultivariate <span>D</span>ata. <em>Computing Science and Statistics</em>, <em>17</em>(1), 171–174.
</div>
<div id="ref-xgvisjclass" class="csl-entry" role="listitem">
Buja, A., &amp; Swayne, D. F. (2002). Visualization <span>M</span>ethodology for <span>M</span>ultidimensional <span>S</span>caling. <em>Journal of Classification</em>, <em>19</em>(1), 7–43.
</div>
<div id="ref-ggvisjcgs" class="csl-entry" role="listitem">
Buja, A., Swayne, D. F., Littman, M. L., Dean, N., Hofmann, H., &amp; Chen, L. (2008). Data visualization with multidimensional scaling. <em>Journal of Computational and Graphical Statistics</em>, <em>17</em>(2), 444–472. <a href="https://doi.org/10.1198/106186008X318440">https://doi.org/10.1198/106186008X318440</a>
</div>
<div id="ref-BT91" class="csl-entry" role="listitem">
Buja, A., &amp; Tukey, P. (Eds.). (1991). <em>Computing and <span>G</span>raphics in <span>S</span>tatistics</em>. Springer-Verlag.
</div>
<div id="ref-CMS99" class="csl-entry" role="listitem">
Card, S. K., Mackinlay, J. D., &amp; Schneiderman, B. (1999). <em>Readings in information visualization</em>. Morgan Kaufmann Publishers.
</div>
<div id="ref-CWL96" class="csl-entry" role="listitem">
Carr, D. B., Wegman, E. J., &amp; Luo, Q. (1996). <em>Explor<span>N</span>: <span>D</span>esign <span>C</span>onsiderations <span>P</span>ast and <span>P</span>resent</em> (Technical Report No. 129). Center for Computational Statistics, George Mason University.
</div>
<div id="ref-Ch95" class="csl-entry" role="listitem">
Chatfield, C. (1995). <em>Problem solving: A statistician’s guide</em>. Chapman; Hall/CRC Press.
</div>
<div id="ref-haerdle" class="csl-entry" role="listitem">
Chen, C., Härdle, W., &amp; Unwin, A. (Eds.). (2006). <em>Handbook of computational statistics (volume III) data visualization</em>. Springer.
</div>
<div id="ref-UHC06" class="csl-entry" role="listitem">
Chen, C.-H., Härdle, W., &amp; Unwin, A. (Eds.). (2007). <em>Handbook of <span>D</span>ata <span>V</span>isualization</em>. Springer.
</div>
<div id="ref-CT94" class="csl-entry" role="listitem">
Cheng, B., &amp; Titterington, M. (1994). Neural <span>N</span>etworks: <span>A</span> <span>R</span>eview from a <span>S</span>tatistical <span>P</span>erspective. <em>Statistical Science</em>, <em>9</em>(1), 2–30.
</div>
<div id="ref-R-crosstalk" class="csl-entry" role="listitem">
Cheng, J., &amp; Sievert, C. (2021). <em>Crosstalk: Inter-widget interactivity for HTML widgets</em>. <a href="https://rstudio.github.io/crosstalk/">https://rstudio.github.io/crosstalk/</a>
</div>
<div id="ref-Ch73" class="csl-entry" role="listitem">
Chernoff, H. (1973). The <span>U</span>se of <span>F</span>aces to <span>R</span>epresent <span>P</span>oints in <span class="math inline">\(k\)</span>-dimensional <span>S</span>pace <span>G</span>raphically. <em>Journal of the American Statistical Association</em>, <em>68</em>, 361–368.
</div>
<div id="ref-Cl79" class="csl-entry" role="listitem">
Cleveland, W. S. (1979). Robust <span>L</span>ocaly <span>W</span>eighted <span>R</span>egression and <span>S</span>moothing <span>S</span>catterplots. <em>Journal of American Statistics Association</em>, <em>74</em>, 829–836.
</div>
<div id="ref-Cl93" class="csl-entry" role="listitem">
Cleveland, W. S. (1993). <em>Visualizing <span>D</span>ata</em>. Hobart Press.
</div>
<div id="ref-CM88" class="csl-entry" role="listitem">
Cleveland, W. S., &amp; McGill, M. E. (Eds.). (1988). <em>Dynamic graphics for statistics</em>. Wadsworth.
</div>
<div id="ref-CB95" class="csl-entry" role="listitem">
Cook, D., &amp; Buja, A. (1997). <span>M</span>anual <span>C</span>ontrols <span>F</span>or <span>H</span>igh-<span>D</span>imensional <span>D</span>ata <span>P</span>rojections. <em>Journal of Computational and Graphical Statistics</em>, <em>6</em>(4), 464–480.
</div>
<div id="ref-CBC93" class="csl-entry" role="listitem">
Cook, D., Buja, A., &amp; Cabrera, J. (1993). <span>P</span>rojection <span>P</span>ursuit <span>I</span>ndexes <span>B</span>ased on <span>O</span>rthonormal <span>F</span>unction <span>E</span>xpansions. <em>Journal of Computational and Graphical Statistics</em>, <em>2</em>(3), 225–250.
</div>
<div id="ref-CBCH95" class="csl-entry" role="listitem">
Cook, D., Buja, A., Cabrera, J., &amp; Hurley, C. (1995b). Grand <span>T</span>our and <span>P</span>rojection <span>P</span>ursuit. <em>Journal of Computational and Graphical Statistics</em>, <em>4</em>(3), 155–172.
</div>
<div id="ref-CBCH94" class="csl-entry" role="listitem">
Cook, D., Buja, A., Cabrera, J., &amp; Hurley, C. (1995a). Grand <span>T</span>our and <span>P</span>rojection <span>P</span>ursuit. <em>Journal of Computational and Graphical Statistics</em>, <em>4</em>(3), 155–172.
</div>
<div id="ref-CHLYNW06" class="csl-entry" role="listitem">
Cook, D., Hofmann, H., Lee, E.-K., Yang, H., Nikolau, B., &amp; Wurtele, E. (2007). Exploring <span>G</span>ene <span>E</span>xpression <span>D</span>ata, <span>U</span>sing <span>P</span>lots. <em>Journal of Data Science</em>, <em>5</em>(2), 151–182.
</div>
<div id="ref-R-mulgar" class="csl-entry" role="listitem">
Cook, D., &amp; Laa, U. (2023). <em>Mulgar: Functions for pre-processing data for multivariate data visualisation using tours</em>. <a href="https://dicook.github.io/mulgar/">https://dicook.github.io/mulgar/</a>
</div>
<div id="ref-CLBW06" class="csl-entry" role="listitem">
Cook, D., Lee, E.-K., Buja, A., &amp; Wickham, H. (2006). <span>G</span>rand <span>T</span>ours, <span>P</span>rojection <span>P</span>ursuit <span>G</span>uided <span>T</span>ours and <span>M</span>anual <span>C</span>ontrols. In C.-H. Chen, W. Härdle, &amp; A. Unwin (Eds.), <em>Handbook of <span>D</span>ata <span>V</span>isualization</em>. Springer.
</div>
<div id="ref-CMSC95" class="csl-entry" role="listitem">
Cook, D., Majure, J. J., Symanzik, J., &amp; Cressie, N. (1996). <span>D</span>ynamic <span>G</span>raphics in a <span>GIS</span>: <span>E</span>xploring and <span>A</span>nalyzing <span>M</span>ultivariate <span>S</span>patial <span>D</span>ata using <span>L</span>inked <span>S</span>oftware. <em>Computational Statistics: Special Issue on Computer Aided Analyses of Spatial Data</em>, <em>11</em>(4), 467–480.
</div>
<div id="ref-CS07" class="csl-entry" role="listitem">
Cook, D., &amp; Swayne, D. F. (2007). <em>Interactive and dynamic graphics for data analysis: With <span>R</span> and <span>GGobi</span></em>. Springer-Verlag. <a href="https://doi.org/10.1007/978-0-387-71762-3">https://doi.org/10.1007/978-0-387-71762-3</a>
</div>
<div id="ref-CPV2003" class="csl-entry" role="listitem">
Cortes, C., Pregibon, D., &amp; Volinsky, C. (2003). Computational <span>M</span>ethods for <span>D</span>ynamic <span>G</span>raphs. <em>Journal of Computational &amp; Graphical Statistics</em>, <em>12</em>(4), 950–970.
</div>
<div id="ref-CV95" class="csl-entry" role="listitem">
Cortes, C., &amp; Vapnik, V. N. (1995). Support-<span>V</span>ector <span>N</span>etworks. <em>Machine Learning</em>, <em>20</em>(3), 273–297.
</div>
<div id="ref-Oc1885" class="csl-entry" role="listitem">
d’Ocagne, M. (1885). <em><span>C</span>oordonnées <span>P</span>arallèles et <span>A</span>xiales: <span>M</span>éthode de transformation géométrique et procédé nouveau de calcul graphique déduits de la considération des coordonnées paralléles</em>. Gauthier-Villars.
</div>
<div id="ref-Dalgaard2002" class="csl-entry" role="listitem">
Dalgaard, P. (2002). <em>Introductory statistics with <span>R</span></em>. Springer.
</div>
<div id="ref-DSP05" class="csl-entry" role="listitem">
Dasu, T., Swayne, D. F., &amp; Poole, D. (2005). Grouping <span>M</span>ultivariate <span>T</span>ime <span>S</span>eries: A <span>C</span>ase <span>S</span>tudy. <em>Proceedings of the IEEE Workshop on <span>T</span>emporal <span>D</span>ata <span>M</span>ining: <span>A</span>lgorithms, <span>T</span>heory and <span>A</span>pplications, in Conjunction with the Conference on Data Mining, Houston, November 27, 2005</em>, 25–32.
</div>
<div id="ref-R-ggdendro" class="csl-entry" role="listitem">
de Vries, A., &amp; Ripley, B. D. (2022). <em>Ggdendro: Create dendrograms and tree diagrams using ggplot2</em>. <a href="https://github.com/andrie/ggdendro">https://github.com/andrie/ggdendro</a>
</div>
<div id="ref-fireorigin" class="csl-entry" role="listitem">
Department of Environment, Land, Water &amp; Planning. (2019). <em><span class="nocase">Fire Origins - Current and Historical</span></em>. <a href="https://discover.data.vic.gov.au/dataset/fire-origins-current-and-historical">https://discover.data.vic.gov.au/dataset/fire-origins-current-and-historical</a>
</div>
<div id="ref-cfa" class="csl-entry" role="listitem">
Department of Environment, Land, Water &amp; Planning. (2020a). <em><span>CFA - Fire Station</span></em>. <a href="https://discover.data.vic.gov.au/dataset/cfa-fire-station-vmfeat-geomark_point">https://discover.data.vic.gov.au/dataset/cfa-fire-station-vmfeat-geomark_point</a>
</div>
<div id="ref-recreation" class="csl-entry" role="listitem">
Department of Environment, Land, Water &amp; Planning. (2020b). <em><span>Recreation Sites</span></em>. <a href="https://discover.data.vic.gov.au/dataset/recreation-sites">https://discover.data.vic.gov.au/dataset/recreation-sites</a>
</div>
<div id="ref-DF84" class="csl-entry" role="listitem">
Diaconis, P., &amp; Freedman, D. (1984a). <span>A</span>symptotics of <span>G</span>raphical <span>P</span>rojection <span>P</span>ursuit. <em>Annals of Statistics</em>, <em>12</em>, 793–815.
</div>
<div id="ref-diaconis1984" class="csl-entry" role="listitem">
Diaconis, P., &amp; Freedman, D. (1984b). Asymptotics of graphical projection pursuit. <em>Annals of Statistics</em>, <em>12</em>(3), 793–815. <a href="https://doi.org/10.1214/aos/1176346703">https://doi.org/10.1214/aos/1176346703</a>
</div>
<div id="ref-DMK05" class="csl-entry" role="listitem">
Dykes, J., MacEachren, A. M., &amp; Kraak, M.-J. (2005). <em>Exploring geovisualization</em>. Elsevier.
</div>
<div id="ref-Ev01" class="csl-entry" role="listitem">
Everitt, B. S., Landau, S., &amp; Leese, M. (2001). <em>Cluster <span>A</span>nalysis (4th ed)</em>. Edward Arnold.
</div>
<div id="ref-Fi79" class="csl-entry" role="listitem">
Fienberg, S. E. (1979). Graphical <span>M</span>ethods in <span>S</span>tatistics. <em>Journal of American Statistical Association</em>, <em>33</em>(4), 165–178.
</div>
<div id="ref-Fi36" class="csl-entry" role="listitem">
Fisher, R. A. (1936a). The <span>U</span>se of <span>M</span>ultiple <span>M</span>easurements in <span>T</span>axonomic <span>P</span>roblems. <em>Annals of Eugenics</em>, <em>7</em>, 179–188.
</div>
<div id="ref-fisher1936" class="csl-entry" role="listitem">
Fisher, R. A. (1936b). The use of multiple measurements in taxonomic problems. <em>Annals of Eugenics</em>, <em>7</em>(2), 179–188. <a href="https://doi.org/10.1111/j.1469-1809.1936.tb02137.x">https://doi.org/10.1111/j.1469-1809.1936.tb02137.x</a>
</div>
<div id="ref-Fi38" class="csl-entry" role="listitem">
Fisher, R. A. (1938). The <span>S</span>tatistical <span>U</span>tilization of <span>M</span>ultiple <span>M</span>easurements. <em>Annals of Eugenics</em>, <em>8</em>, 376–386.
</div>
<div id="ref-PRIM9-video" class="csl-entry" role="listitem">
Fisherkeller, M. A., Friedman, J. H., &amp; Tukey, J. W. (1973). <em><span>PRIM-9</span>, an interactive multidimensional data display and analysis system</em>. <a href="https://www.youtube.com/watch?v=B7XoW2qiFUA">https://www.youtube.com/watch?v=B7XoW2qiFUA</a>
</div>
<div id="ref-tukey" class="csl-entry" role="listitem">
Fisherkeller, M. A., Friedman, J. H., &amp; Tukey, J. W. (1974). <span>PRIM-9</span>, an interactive multidimensional data display and analysis system. In W. S. Cleveland (Ed.), <em>The collected works of john w. Tukey: Graphics 1965-1985, volume v</em> (pp. 340–346).
</div>
<div id="ref-eechidna" class="csl-entry" role="listitem">
Forbes, J., Cook, D., &amp; Hyndman, R. J. (2020). Spatial modelling of the two-party preferred vote in australian federal elections: 2001–2016. <em>Australian &amp; New Zealand Journal of Statistics</em>, <em>62</em>(2), 168–185. https://doi.org/<a href="https://doi.org/10.1111/anzs.12292">https://doi.org/10.1111/anzs.12292</a>
</div>
<div id="ref-Ford92" class="csl-entry" role="listitem">
Ford, B. J. (1992). <em>Images of science: A history of scientific illustration</em>. The British Library.
</div>
<div id="ref-forgy65" class="csl-entry" role="listitem">
Forgy, E. (1965). Cluster analysis of multivariate data: Efficiency versus interpretability of classification. <em>Biometrics</em>, <em>21</em>(3), 768–769.
</div>
<div id="ref-FR02" class="csl-entry" role="listitem">
Fraley, C., &amp; Raftery, A. E. (2002). Model-based <span>C</span>lustering, <span>D</span>iscriminant <span>A</span>nalysis, <span>D</span>ensity <span>E</span>stimation. <em>Journal of the American Statistical Association</em>, <em>97</em>, 611–631.
</div>
<div id="ref-R-mclust" class="csl-entry" role="listitem">
Fraley, C., Raftery, A. E., &amp; Scrucca, L. (2022). <em>Mclust: Gaussian mixture modelling for model-based clustering, classification, and density estimation</em>. <a href="https://mclust-org.github.io/mclust/">https://mclust-org.github.io/mclust/</a>
</div>
<div id="ref-Fr87" class="csl-entry" role="listitem">
Friedman, J. H. (1987). <span>E</span>xploratory <span>P</span>rojection <span>P</span>ursuit. <em>Journal of American Statistical Association</em>, <em>82</em>, 249–266.
</div>
<div id="ref-FT74" class="csl-entry" role="listitem">
Friedman, J. H., &amp; Tukey, J. W. (1974). <span>A</span> <span>P</span>rojection <span>P</span>ursuit <span>A</span>lgorithm for <span>E</span>xploratory <span>D</span>ata <span>A</span>nalysis. <em>IEEE Transactions on Computing C</em>, <em>23</em>, 881–889.
</div>
<div id="ref-FD04" class="csl-entry" role="listitem">
Friendly, M., &amp; Denis, D. J. (2004). <em>Milestones in the history of thematic cartography, statistical graphics, and data visualization</em>. http://www.math.yorku.ca/SCS/Gallery/milestone/.
</div>
<div id="ref-neuralnet" class="csl-entry" role="listitem">
Fritsch, S., Guenther, F., &amp; Wright, M. N. (2019). <em>Neuralnet: Training of neural networks</em>. <a href="https://CRAN.R-project.org/package=neuralnet">https://CRAN.R-project.org/package=neuralnet</a>
</div>
<div id="ref-FB94" class="csl-entry" role="listitem">
Furnas, G. W., &amp; Buja, A. (1994). Prosection <span>V</span>iews: <span>D</span>imensional <span>I</span>nference <span>T</span>hrough <span>S</span>ections and <span>P</span>rojections. <em>Journal of Computational and Graphical Statistics</em>, <em>3</em>(4), 323–385.
</div>
<div id="ref-Ga71" class="csl-entry" role="listitem">
Gabriel, K. R. (1971). The <span>B</span>iplot <span>G</span>raphical <span>D</span>isplay of <span>M</span>atrices with <span>A</span>pplications to <span>P</span>rincipal <span>C</span>omponent <span>A</span>nalysis. <em>Biometrika</em>, <em>58</em>, 453–467.
</div>
<div id="ref-Gentle04" class="csl-entry" role="listitem">
Gentle, J. E., Härdle, W., &amp; Mori, Y. (Eds.). (2004). <em>Handbook of computational statistics: Concepts and methods</em>. Springer.
</div>
<div id="ref-giordani" class="csl-entry" role="listitem">
Giordani, P., Ferraro, M. B., &amp; Martella, F. (2020). <em>An introduction to clustering with r</em>. Springer Singapore. <a href="https://doi.org/10.1007/978-981-13-0553-5">https://doi.org/10.1007/978-981-13-0553-5</a>
</div>
<div id="ref-GH92" class="csl-entry" role="listitem">
Glover, D. M., &amp; Hopke, P. K. (1992). <span>E</span>xploration of <span>M</span>ultivariate <span>C</span>hemical <span>D</span>ata by <span>P</span>rojection <span>P</span>ursuit. <em>Chemometrics and Intelligent Laboratory Systems</em>, <em>16</em>, 45–59.
</div>
<div id="ref-Good05" class="csl-entry" role="listitem">
Good, P. (2005). <em>Permutation, <span>P</span>arametric, and <span>B</span>ootstrap <span>T</span>ests of <span>H</span>ypotheses</em>. Springer.
</div>
<div id="ref-GH96" class="csl-entry" role="listitem">
Gower, J. C., &amp; Hand, D. J. (1996). <em>Biplots</em>. Chapman; Hall.
</div>
<div id="ref-risk-survey" class="csl-entry" role="listitem">
Hajibaba, H., Karlsson, L., &amp; Dolnicar, S. (2016). Residents open their homes to tourists when disaster strikes. <em>Journal of Travel Research</em>, <em>56</em>(8), 1065–1078.
</div>
<div id="ref-Hansen2004" class="csl-entry" role="listitem">
Hansen, C., &amp; Johnson, C. R. (2004). <em>Visualization handbook</em>. Academic Press.
</div>
<div id="ref-R-langevitour" class="csl-entry" role="listitem">
Harrison, P. (2023a). <em>Langevitour: Langevin tour</em>. <a href="https://logarithmic.net/langevitour/">https://logarithmic.net/langevitour/</a>
</div>
<div id="ref-langevitour2023" class="csl-entry" role="listitem">
Harrison, P. (2023b). Langevitour: Smooth interactive touring of high dimensions, demonstrated with scRNA-seq data. <em>The R Journal</em>, <em>15</em>(2), 206–219. <a href="https://doi.org/10.32614/RJ-2023-046">https://doi.org/10.32614/RJ-2023-046</a>
</div>
<div id="ref-R-detourr" class="csl-entry" role="listitem">
Hart, C., &amp; Wang, E. (2022). <em>Detourr: Portable and performant tour animations</em>. <a href="https://casperhart.github.io/detourr/">https://casperhart.github.io/detourr/</a>
</div>
<div id="ref-HK81" class="csl-entry" role="listitem">
Hartigan, J. A., &amp; Kleiner, B. (1981). Mosaics for <span>C</span>ontingency <span>T</span>ables. <em>Computer Science and Statistics: Proceedings of the 13th Symposium on the Interface</em>, 268–273.
</div>
<div id="ref-HK84" class="csl-entry" role="listitem">
Hartigan, J., &amp; Kleiner, B. (1984). A <span>M</span>osaic of <span>T</span>elevision <span>R</span>atings. <em>The American Statistician</em>, <em>38</em>, 32–35.
</div>
<div id="ref-HBCUW91" class="csl-entry" role="listitem">
Haslett, J., Bradley, R., Craig, P., Unwin, A., &amp; Wills, G. (1991). <span>D</span>ynamic <span>G</span>raphics for <span>E</span>xploring <span>S</span>patial <span>D</span>ata with <span>A</span>pplication to <span>L</span>ocating <span>G</span>lobal and <span>L</span>ocal <span>A</span>nomalies. <em>The American Statistician</em>, <em>45</em>(3), 234–242.
</div>
<div id="ref-HTF01" class="csl-entry" role="listitem">
Hastie, T., Tibshirani, R., &amp; Friedman, J. (2001). <em>The <span>E</span>lements of <span>S</span>tatistical <span>L</span>earning</em>. Springer.
</div>
<div id="ref-hennig" class="csl-entry" role="listitem">
Hennig, C., Meila, M., Murtagh, F., &amp; Rocci, R. (Eds.). (2015). <em>Handbook of cluster analysis</em>. Chapman; Hall/<span>CRC</span>. <a href="https://doi.org/10.1201/b19706">https://doi.org/10.1201/b19706</a>
</div>
<div id="ref-Ho01" class="csl-entry" role="listitem">
Hofmann, H. (2001). <em>Graphical <span>T</span>ools for the <span>E</span>xploration of <span>M</span>ultivariate <span>C</span>ategorical <span>D</span>ata</em>. Books on Demand.
</div>
<div id="ref-Ho03" class="csl-entry" role="listitem">
Hofmann, H. (2003). Constructing and <span>R</span>eading <span>M</span>osaicplots. <em>Computational Statistics and Data Analysis</em>, <em>43</em>(4), 565–580.
</div>
<div id="ref-HT98" class="csl-entry" role="listitem">
Hofmann, H., &amp; Theus, M. (1998). Selection <span>S</span>equences in <span>MANET</span>. <em>Computational Statistics</em>, <em>13</em>(1), 77–87.
</div>
<div id="ref-ggfortify2018" class="csl-entry" role="listitem">
Horikoshi, M., &amp; Tang, Y. (2018). <em>Ggfortify: Data visualization tools for statistical analysis results</em>. <a href="https://CRAN.R-project.org/package=ggfortify">https://CRAN.R-project.org/package=ggfortify</a>
</div>
<div id="ref-R-ggfortify" class="csl-entry" role="listitem">
Horikoshi, M., &amp; Tang, Y. (2023). <em>Ggfortify: Data visualization tools for statistical analysis results</em>. <a href="https://github.com/sinhrks/ggfortify">https://github.com/sinhrks/ggfortify</a>
</div>
<div id="ref-R-palmerpenguins" class="csl-entry" role="listitem">
Horst, A., Hill, A., &amp; Gorman, K. (2022). <em>Palmerpenguins: Palmer archipelago (antarctica) penguin data</em>. <a href="https://CRAN.R-project.org/package=palmerpenguins">https://CRAN.R-project.org/package=palmerpenguins</a>
</div>
<div id="ref-hotelling-pca" class="csl-entry" role="listitem">
Hotelling, H. (1933). Analysis of a complex of statistical variables into principal components. <em>Journal of Educational Psychology</em>, <em>24</em>(6), 417--441. <a href="https://doi.org/10.1037/h0071325">https://doi.org/10.1037/h0071325</a>
</div>
<div id="ref-Hu85" class="csl-entry" role="listitem">
Huber, P. J. (1985). <span>P</span>rojection <span>P</span>ursuit (with discussion). <em>Annals of Statistics</em>, <em>13</em>, 435–525.
</div>
<div id="ref-Hurley87" class="csl-entry" role="listitem">
Hurley, C. (1987). <em>The data viewer: An interactive program for data analysis</em> [PhD thesis]. University of Washington.
</div>
<div id="ref-R-gt" class="csl-entry" role="listitem">
Iannone, R., Cheng, J., Schloerke, B., Hughes, E., Lauer, A., &amp; Seo, J. (2023). <em>Gt: Easily create presentation-ready display tables</em>. <a href="https://CRAN.R-project.org/package=gt">https://CRAN.R-project.org/package=gt</a>
</div>
<div id="ref-R1996" class="csl-entry" role="listitem">
Ihaka, R., &amp; Gentleman, R. (1996). R: <span>A</span> <span>L</span>anguage for <span>D</span>ata <span>A</span>nalysis and <span>G</span>raphics. <em>Journal of Computational and Graphical Statistics</em>, <em>5</em>, 299–314.
</div>
<div id="ref-R-colorspace" class="csl-entry" role="listitem">
Ihaka, R., Murrell, P., Hornik, K., Fisher, J. C., Stauffer, R., Wilke, C. O., McWhite, C. D., &amp; Zeileis, A. (2023). <em>Colorspace: A toolbox for manipulating and assessing colors and palettes</em>. <a href="https://CRAN.R-project.org/package=colorspace">https://CRAN.R-project.org/package=colorspace</a>
</div>
<div id="ref-In85" class="csl-entry" role="listitem">
Inselberg, A. (1985). <span>T</span>he <span>P</span>lane with <span>P</span>arallel <span>C</span>oordinates. <em>The Visual Computer</em>, <em>1</em>, 69–91.
</div>
<div id="ref-windasos" class="csl-entry" role="listitem">
Iowa State University. (2020). <em>ASOS-AWOS-METAR data download</em>. <a href="https://mesonet.agron.iastate.edu/request/download.phtml?network=AU__ASOS">https://mesonet.agron.iastate.edu/request/download.phtml?network=AU__ASOS</a>
</div>
<div id="ref-JT07" class="csl-entry" role="listitem">
Johnson, D., &amp; Travis, J. (2007). <em>Flatland: The movie</em>. https://round-drum-w7xh.squarespace.com/our-story.
</div>
<div id="ref-JW02" class="csl-entry" role="listitem">
Johnson, R. A., &amp; Wichern, D. W. (2002). <em>Applied multivariate statistical analysis (5th ed)</em>. Prentice-Hall.
</div>
<div id="ref-joliffe2016" class="csl-entry" role="listitem">
Jolliffe, I. T., &amp; Cadima, J. (2016). Principal component analysis: A review and recent developments. <em>Phil. Trans. R. Soc. A.</em>, <em>374</em>, 20150202. <a href="https://doi.org/10.1098/rsta.2015.0202">https://doi.org/10.1098/rsta.2015.0202</a>
</div>
<div id="ref-JS87" class="csl-entry" role="listitem">
Jones, M. C., &amp; Sibson, R. (1987). <span>W</span>hat is <span>P</span>rojection <span>P</span>ursuit? (With discussion). <em>Journal of the Royal Statistical Society, Series A</em>, <em>150</em>, 1–36.
</div>
<div id="ref-kassambara" class="csl-entry" role="listitem">
Kassambara, A. (2017). <em>Practical guide to cluster analysis in r: Unsupervised machine learning</em>. STHDA.
</div>
<div id="ref-R-ggpubr" class="csl-entry" role="listitem">
Kassambara, A. (2023). <em>Ggpubr: ggplot2 based publication ready plots</em>. <a href="https://rpkgs.datanovia.com/ggpubr/">https://rpkgs.datanovia.com/ggpubr/</a>
</div>
<div id="ref-Ko01" class="csl-entry" role="listitem">
Kohonen, T. (2001). <em>Self-<span>O</span>rganizing <span>M</span>aps (3rd ed)</em>. Springer.
</div>
<div id="ref-KS96" class="csl-entry" role="listitem">
Koschat, M. A., &amp; Swayne, D. F. (1996). Interactive <span>G</span>raphical <span>M</span>ethods in the <span>A</span>nalysis of <span>C</span>ustomer <span>P</span>anel <span>D</span>ata (with discussion). <em>Journal of Business and Economic Statistics</em>, <em>14</em>(1), 113–132.
</div>
<div id="ref-R-Rtsne" class="csl-entry" role="listitem">
Krijthe, J. (2022). <em>Rtsne: T-distributed stochastic neighbor embedding using a barnes-hut implementation</em>. <a href="https://github.com/jkrijthe/Rtsne">https://github.com/jkrijthe/Rtsne</a>
</div>
<div id="ref-Kr64a" class="csl-entry" role="listitem">
Kruskal, J. B. (1964a). Multidimensional <span>S</span>caling by <span>O</span>ptimizing <span>G</span>oodness of <span>F</span>it to a <span>N</span>onmetric <span>H</span>ypothesis. <em>Psychometrika</em>, <em>29</em>, 1–27.
</div>
<div id="ref-Kr64b" class="csl-entry" role="listitem">
Kruskal, J. B. (1964b). Nonmetric <span>M</span>ultidimensional <span>S</span>caling: A <span>N</span>umerical <span>M</span>ethod. <em>Psychometrika</em>, <em>29</em>, 115–129.
</div>
<div id="ref-KW78" class="csl-entry" role="listitem">
Kruskal, J. B., &amp; Wish, M. (1978). <em>Multidimensional <span>S</span>caling</em>. Sage Publications.
</div>
<div id="ref-tidymodels2020" class="csl-entry" role="listitem">
Kuhn, M., &amp; Wickham, H. (2020). <em>Tidymodels: A collection of packages for modeling and machine learning using tidyverse principles.</em> <a href="https://www.tidymodels.org">https://www.tidymodels.org</a>
</div>
<div id="ref-R-tidymodels" class="csl-entry" role="listitem">
Kuhn, M., &amp; Wickham, H. (2023). <em>Tidymodels: Easily install and load the tidymodels packages</em>. <a href="https://CRAN.R-project.org/package=tidymodels">https://CRAN.R-project.org/package=tidymodels</a>
</div>
<div id="ref-sagetour" class="csl-entry" role="listitem">
Laa, U., Cook, D., &amp; Lee, S. (2022). Burning sage: Reversing the curse of dimensionality in the visualization of high-dimensional data. <em>Journal of Computational and Graphical Statistics</em>, <em>31</em>(1), 40–49. <a href="https://doi.org/10.1080/10618600.2021.1963264">https://doi.org/10.1080/10618600.2021.1963264</a>
</div>
<div id="ref-slicetour" class="csl-entry" role="listitem">
Laa, U., Cook, D., &amp; Valencia, G. (2020a). A slice tour for finding hollowness in high-dimensional data. <em>Journal of Computational and Graphical Statistics</em>, <em>29</em>(3), 681–687. <a href="https://doi.org/10.1080/10618600.2020.1777140">https://doi.org/10.1080/10618600.2020.1777140</a>
</div>
<div id="ref-tourr2020" class="csl-entry" role="listitem">
Laa, U., Cook, D., &amp; Valencia, G. (2020b). A slice tour for finding hollowness in high-dimensional data. <em>Journal of Computational and Graphical Statistics</em>, <em>29</em>(3), 681–687. <a href="https://doi.org/10.1080/10618600.2020.1777140">https://doi.org/10.1080/10618600.2020.1777140</a>
</div>
<div id="ref-helmert" class="csl-entry" role="listitem">
Lancaster, H. O. (1965). The helmert matrices. <em>The American Mathematical Monthly</em>, <em>72</em>(1), 4–12.
</div>
<div id="ref-R-cxhull" class="csl-entry" role="listitem">
Laurent, S. (2023). <em>Cxhull: Convex hull</em>. <a href="https://github.com/stla/cxhull">https://github.com/stla/cxhull</a>
</div>
<div id="ref-Lee2018" class="csl-entry" role="listitem">
Lee, E.-K. (2018). PPtreeViz: An r package for visualizing projection pursuit classification trees. <em>Journal of Statistical Software</em>, <em>83</em>(8), 1–30. <a href="https://doi.org/10.18637/jss.v083.i08">https://doi.org/10.18637/jss.v083.i08</a>
</div>
<div id="ref-Lee2009" class="csl-entry" role="listitem">
Lee, E.-K., &amp; Cook, D. (2009). A projection pursuit index for large <span class="math inline">\(p\)</span> small <span class="math inline">\(n\)</span> data. <em>Statistics and Computing</em>, <em>20</em>, 381–392. <a href="https://doi.org/10.1007/s11222-009-9131-1">https://doi.org/10.1007/s11222-009-9131-1</a>
</div>
<div id="ref-LCKL05" class="csl-entry" role="listitem">
Lee, E.-K., Cook, D., Klinke, S., &amp; Lumley, T. (2005). <span>P</span>rojection <span>P</span>ursuit for <span>E</span>xploratory <span>S</span>upervised <span>C</span>lassification. <em>Journal of Computational and Graphical Statistics</em>, <em>14</em>(4), 831–846.
</div>
<div id="ref-R-liminal" class="csl-entry" role="listitem">
Lee, S. (2021). <em>Liminal: Multivariate data visualization with tours and embeddings</em>. <a href="https://CRAN.R-project.org/package=liminal">https://CRAN.R-project.org/package=liminal</a>
</div>
<div id="ref-tours2022" class="csl-entry" role="listitem">
Lee, S., Cook, D., Silva, N. da, Laa, U., Spyrison, N., Wang, E., &amp; Zhang, H. S. (2022). The state-of-the-art on tours for dynamic visualization of high-dimensional data. <em>WIREs Computational Statistics</em>, <em>14</em>(4), e1573. <a href="https://doi.org/10.1002/wics.1573">https://doi.org/10.1002/wics.1573</a>
</div>
<div id="ref-Lee2013" class="csl-entry" role="listitem">
Lee, Y. D., Cook, D., Park, J., &amp; Lee, E.-K. (2013). <span class="nocase">PPtree: Projection pursuit classification tree</span>. <em>Electronic Journal of Statistics</em>, <em>7</em>(none), 1369–1386. <a href="https://doi.org/10.1214/13-EJS810">https://doi.org/10.1214/13-EJS810</a>
</div>
<div id="ref-ctv-clustering" class="csl-entry" role="listitem">
Leisch, F., &amp; Gruen, B. (2023). <em>CRAN task view: Cluster analysis &amp; finite mixture models</em>. https://cran.r-project.org/web/views/Cluster.html.
</div>
<div id="ref-R-MSA" class="csl-entry" role="listitem">
Leisch, F., &amp; Grün, B. (2020). <em>MSA: Market segmentation analysis</em>.
</div>
<div id="ref-li2020visualizing" class="csl-entry" role="listitem">
Li, M., Zhao, Z., &amp; Scheidegger, C. (2020). Visualizing neural networks with the grand tour. <em>Distill</em>. <a href="https://doi.org/10.23915/distill.00025">https://doi.org/10.23915/distill.00025</a>
</div>
<div id="ref-randomForest2002" class="csl-entry" role="listitem">
Liaw, A., &amp; Wiener, M. (2002). Classification and regression by randomForest. <em>R News</em>, <em>2</em>(3), 18–22. <a href="https://CRAN.R-project.org/doc/Rnews/">https://CRAN.R-project.org/doc/Rnews/</a>
</div>
<div id="ref-LSDB92" class="csl-entry" role="listitem">
Littman, M. L., Swayne, D. F., Dean, N., &amp; Buja, A. (1992). Visualizing the <span>E</span>mbedding of <span>O</span>bjects in <span>E</span>uclidean <span>S</span>pace. <em>Computing Science and Statistics: Proceedings of the 24th Symposium on the Interface</em>, 208–217.
</div>
<div id="ref-Lloyd1982" class="csl-entry" role="listitem">
Lloyd, S. (1982). Least squares quantization in PCM. <em>IEEE Transactions on Information Theory</em>, <em>28</em>(2), 129–137. <a href="https://doi.org/10.1109/TIT.1982.1056489">https://doi.org/10.1109/TIT.1982.1056489</a>
</div>
<div id="ref-LMGR05" class="csl-entry" role="listitem">
Longley, P. A., Maguire, D. J., Goodchild, M. F., &amp; Rhind, D. W. (2005). <em>Geographic information systems and science</em>. John Wiley &amp; Sons.
</div>
<div id="ref-loperfido" class="csl-entry" role="listitem">
Loperfido, N. (2018). Skewness-based projection pursuit: A computational approach. <em>Computational Statistics &amp; Data Analysis</em>, <em>120</em>, 42–57. https://doi.org/<a href="https://doi.org/10.1016/j.csda.2017.11.001">https://doi.org/10.1016/j.csda.2017.11.001</a>
</div>
<div id="ref-Maaten2008" class="csl-entry" role="listitem">
Maaten, L. van der, &amp; Hinton, G. (2008). Visualizing data using <span class="nocase">t-SNE</span>. <em>J. Mach. Learn. Res.</em>, <em>9</em>(Nov), 2579–2605. <a href="http://www.jmlr.org/papers/v9/vandermaaten08a.html">http://www.jmlr.org/papers/v9/vandermaaten08a.html</a>
</div>
<div id="ref-MacQueen1967" class="csl-entry" role="listitem">
MacQueen, J. B. (1967). Some methods for classification and analysis of multivariate observations. In L. M. L. Cam &amp; J. Neyman (Eds.), <em>Proc. Of the fifth berkeley symposium on mathematical statistics and probability</em> (Vol. 1, pp. 281–297). University of California Press.
</div>
<div id="ref-MB03" class="csl-entry" role="listitem">
Maindonald, J., &amp; Braun, J. (2003). <em>Data analysis and graphics using r - an example-based approach</em>. Cambridge University Press.
</div>
<div id="ref-Ma65" class="csl-entry" role="listitem">
Martin, E. (1965). <em>Flatland</em>. http://www.der.org/films/flatland.html.
</div>
<div id="ref-kernelshap" class="csl-entry" role="listitem">
Mayer, M., &amp; Watson, D. (2023). <em>Kernelshap: Kernel SHAP</em>. <a href="https://CRAN.R-project.org/package=kernelshap">https://CRAN.R-project.org/package=kernelshap</a>
</div>
<div id="ref-MY94" class="csl-entry" role="listitem">
McFarlane, M., &amp; Young, F. W. (1994). Graphical <span>S</span>ensitivity <span>A</span>nalysis for <span>M</span>ultidimensional <span>S</span>caling. <em>Journal of Computational and Graphical Statistics</em>, <em>3</em>, 23–33.
</div>
<div id="ref-McInnes2018" class="csl-entry" role="listitem">
McInnes, L., Healy, J., &amp; Melville, J. (2018). <em><span>UMAP</span>: Uniform manifold approximation and projection for dimension reduction</em>. <a href="http://arxiv.org/abs/1802.03426">http://arxiv.org/abs/1802.03426</a>
</div>
<div id="ref-Mc77" class="csl-entry" role="listitem">
McNeil, D. (1977). <em>Interactive <span>D</span>ata <span>A</span>nalysis</em>. John Wiley &amp; Sons.
</div>
<div id="ref-wind" class="csl-entry" role="listitem">
McVicar, T. (2011). <em>Near-surface wind speed. v10. CSIRO. Data collection.</em> <a href="https://doi.org/10.25919/5c5106acbcb02">https://doi.org/10.25919/5c5106acbcb02</a>
</div>
<div id="ref-R-e1071" class="csl-entry" role="listitem">
Meyer, D., Dimitriadou, E., Hornik, K., Weingessel, A., &amp; Leisch, F. (2023). <em>e1071: Misc functions of the department of statistics, probability theory group (formerly: E1071), TU wien</em>. <a href="https://CRAN.R-project.org/package=e1071">https://CRAN.R-project.org/package=e1071</a>
</div>
<div id="ref-R-rpart.plot" class="csl-entry" role="listitem">
Milborrow, S. (2022). <em>Rpart.plot: Plot rpart models: An enhanced version of plot.rpart</em>. <a href="http://www.milbo.org/rpart-plot/index.html">http://www.milbo.org/rpart-plot/index.html</a>
</div>
<div id="ref-R-gtExtras" class="csl-entry" role="listitem">
Mock, T. (2022). <em>gtExtras: Extending gt for beautiful HTML tables</em>. <a href="https://CRAN.R-project.org/package=gtExtras">https://CRAN.R-project.org/package=gtExtras</a>
</div>
<div id="ref-iml" class="csl-entry" role="listitem">
Molnar, C. (2022). <em>Interpretable machine learning: A guide for making black box models explainable (2nd ed)</em>. https://christophm.github.io/interpretable-ml-book/.
</div>
<div id="ref-Moon2019" class="csl-entry" role="listitem">
Moon, K. R., Dijk, D. van, Wang, Z., Gigante, S., Burkhardt, D. B., Chen, W. S., Yim, K., Elzen, A. van den, Hirn, M. J., Coifman, R. R., Ivanova, N. B., Wolf, G., &amp; Krishnaswamy, S. (2019). Visualizing structure and transitions for biological data exploration. <em>Nature Biotechnology</em>, <em>37</em>, 1482–1492. <a href="https://doi.org/10.1038/s41587-019-0336-3">https://doi.org/10.1038/s41587-019-0336-3</a>
</div>
<div id="ref-Murrell2005" class="csl-entry" role="listitem">
Murrell, P. (2005). <em>R graphics</em>. Chapman &amp; Hall/CRC.
</div>
<div id="ref-OpenStreetMap" class="csl-entry" role="listitem">
OpenStreetMap contributors. (2020). <em><span class="nocase">Planet dump retrieved from https://planet.osm.org </span></em>.<a href="https://www.openstreetmap.org%20" class="uri"> https://www.openstreetmap.org</a>.
</div>
<div id="ref-pearson-pca" class="csl-entry" role="listitem">
Pearson, K. (1901). LIII. On lines and planes of closest fit to systems of points in space. <em>The London, Edinburgh, and Dublin Philosophical Magazine and Journal of Science</em>, <em>2</em>(11), 559–572. <a href="https://doi.org/10.1080/14786440109462720">https://doi.org/10.1080/14786440109462720</a>
</div>
<div id="ref-R-patchwork" class="csl-entry" role="listitem">
Pedersen, T. L. (2023). <em>Patchwork: The composer of plots</em>. <a href="https://CRAN.R-project.org/package=patchwork">https://CRAN.R-project.org/package=patchwork</a>
</div>
<div id="ref-perisic2009" class="csl-entry" role="listitem">
Perisic, I., &amp; Posse, C. (2005). Projection pursuit indices based on the empirical distribution function. <em>Journal of Computational and Graphical Statistics</em>, <em>14</em>(3), 700–715. <a href="https://doi.org/10.1198/106186005X69440">https://doi.org/10.1198/106186005X69440</a>
</div>
<div id="ref-Po95" class="csl-entry" role="listitem">
Polzehl, J. (1995). Projection <span>P</span>ursuit <span>D</span>iscriminant <span>A</span>nalysis. <em>Computational Statistics and Data Analysis</em>, <em>20</em>, 141–157.
</div>
<div id="ref-Po92" class="csl-entry" role="listitem">
Posse, C. (1992). Projection <span>P</span>ursuit <span>D</span>iscriminant <span>A</span>nalysis for <span>T</span>wo <span>G</span>roups. <em>Communications in Statistics, Part A – Theory and Methods</em>, <em>21</em>, 1–19.
</div>
<div id="ref-Po94" class="csl-entry" role="listitem">
Posse, C. (1995). Tools for <span>T</span>wo-dimensional <span>P</span>rojection <span>P</span>ursuit. <em>Journal of Computational and Graphical Statistics</em>, <em>4</em>(2), 83–100.
</div>
<div id="ref-jaxa" class="csl-entry" role="listitem">
P-Tree System. (2020). <em><span class="nocase">JAXA Himawari Monitor - User’s Guide</span></em>. <a href="https://www.eorc.jaxa.jp/ptree/userguide.html">https://www.eorc.jaxa.jp/ptree/userguide.html</a>
</div>
<div id="ref-R" class="csl-entry" role="listitem">
R Core Team. (2023). <em>R: A language and environment for statistical computing</em>. R Foundation for Statistical Computing. <a href="https://www.R-project.org/">https://www.R-project.org/</a>
</div>
<div id="ref-Rao48" class="csl-entry" role="listitem">
Rao, C. R. (1948). The <span>U</span>tilization of <span>M</span>ultiple <span>M</span>easurements in <span>P</span>roblems of <span>B</span>iological <span>C</span>lassification (with discussion). <em>Journal of the Royal Statistical Society, Series B</em>, <em>10</em>, 159–203.
</div>
<div id="ref-Rao93" class="csl-entry" role="listitem">
Rao, C. R. (Ed.). (1993). <em><span>H</span>andbook of <span>S</span>tatistics, <span>V</span>ol.&nbsp;9</em>. Elsevier Science Publishers.
</div>
<div id="ref-W05" class="csl-entry" role="listitem">
Rao, C. R., Wegman, E. J., &amp; Solka, J. L. (Eds.). (2006). <em>Handbook of <span>S</span>tatistics: <span>D</span>ata <span>M</span>ining and <span>V</span>isualization</em>. Elsevier/North-Holland.
</div>
<div id="ref-Ri96" class="csl-entry" role="listitem">
Ripley, B. (1996). <em>Pattern <span>R</span>ecognition and <span>N</span>eural <span>N</span>etworks</em>. Cambridge University Press.
</div>
<div id="ref-R-MASS" class="csl-entry" role="listitem">
Ripley, B. (2023a). <em>MASS: Support functions and datasets for venables and ripley’s MASS</em>. <a href="http://www.stats.ox.ac.uk/pub/MASS4/">http://www.stats.ox.ac.uk/pub/MASS4/</a>
</div>
<div id="ref-R-nnet" class="csl-entry" role="listitem">
Ripley, B. (2023b). <em>Nnet: Feed-forward neural networks and multinomial log-linear models</em>. <a href="http://www.stats.ox.ac.uk/pub/MASS4/">http://www.stats.ox.ac.uk/pub/MASS4/</a>
</div>
<div id="ref-Ro1957" class="csl-entry" role="listitem">
Rothkopf, E. Z. (1957). A <span>M</span>easure of <span>S</span>timulus <span>S</span>imilarity and <span>E</span>rrors in <span>S</span>ome <span>P</span>aired-associate <span>L</span>earning <span>T</span>asks. <em>Journal of Experimental Psychology</em>, <em>53</em>, 94–101.
</div>
<div id="ref-R-geozoo" class="csl-entry" role="listitem">
Schloerke, B. (2016). <em>Geozoo: Zoo of geometric objects</em>. <a href="https://CRAN.R-project.org/package=geozoo">https://CRAN.R-project.org/package=geozoo</a>
</div>
<div id="ref-R-GGally" class="csl-entry" role="listitem">
Schloerke, B., Cook, D., Larmarange, J., Briatte, F., Marbach, M., Thoen, E., Elberg, A., &amp; Crowley, J. (2023). <em>GGally: Extension to ggplot2</em>. <a href="https://ggobi.github.io/ggally/">https://ggobi.github.io/ggally/</a>
</div>
<div id="ref-mclust2016" class="csl-entry" role="listitem">
Scrucca, L., Fop, M., Murphy, T. B., &amp; Raftery, A. E. (2016). <span class="nocase">mclust</span> 5: Clustering, classification and density estimation using <span>G</span>aussian finite mixture models. <em>The <span>R</span> Journal</em>, <em>8</em>(1), 289–317. <a href="https://doi.org/10.32614/RJ-2016-021">https://doi.org/10.32614/RJ-2016-021</a>
</div>
<div id="ref-Sh62" class="csl-entry" role="listitem">
Shepard, R. N. (1962). The <span>A</span>nalysis of <span>P</span>roximities: <span>M</span>ultidimensional <span>S</span>caling with an <span>U</span>nknown <span>D</span>istance <span>F</span>unction, <span>I</span> and <span>II</span>. <em>Psychometrika</em>, <em>27</em>, 125-139 and 219-246.
</div>
<div id="ref-plotly2020" class="csl-entry" role="listitem">
Sievert, C. (2020). <em>Interactive web-based data visualization with r, plotly, and shiny</em>. Chapman; Hall/CRC. <a href="https://plotly-r.com">https://plotly-r.com</a>
</div>
<div id="ref-R-plotly" class="csl-entry" role="listitem">
Sievert, C., Parmer, C., Hocking, T., Chamberlain, S., Ram, K., Corvellec, M., &amp; Despouy, P. (2023). <em>Plotly: Create interactive web graphics via plotly.js</em>. <a href="https://CRAN.R-project.org/package=plotly">https://CRAN.R-project.org/package=plotly</a>
</div>
<div id="ref-R-gtsummary" class="csl-entry" role="listitem">
Sjoberg, D. D., Larmarange, J., Curry, M., Lavery, J., Whiting, K., &amp; Zabor, E. C. (2023). <em>Gtsummary: Presentation-ready data summary and analytic result tables</em>. <a href="https://CRAN.R-project.org/package=gtsummary">https://CRAN.R-project.org/package=gtsummary</a>
</div>
<div id="ref-gtsummary" class="csl-entry" role="listitem">
Sjoberg, D. D., Whiting, K., Curry, M., Lavery, J. A., &amp; Larmarange, J. (2021). Reproducible summary tables with the gtsummary package. <em><span>The R Journal</span></em>, <em>13</em>, 570–580. <a href="https://doi.org/10.32614/RJ-2021-053">https://doi.org/10.32614/RJ-2021-053</a>
</div>
<div id="ref-R-ggrepel" class="csl-entry" role="listitem">
Slowikowski, K. (2023). <em>Ggrepel: Automatically position non-overlapping text labels with ggplot2</em>. <a href="https://github.com/slowkow/ggrepel">https://github.com/slowkow/ggrepel</a>
</div>
<div id="ref-R-bomrang" class="csl-entry" role="listitem">
Sparks, A. H., Carroll, J., Goldie, J., Marchiori, D., Melloy, P., Padgham, M., Parsonage, H., &amp; Pembleton, K. (2020). <em><span class="nocase">bomrang</span>: Australian government bureau of meteorology (BOM) data client</em>. <a href="https://CRAN.R-project.org/package=bomrang">https://CRAN.R-project.org/package=bomrang</a>
</div>
<div id="ref-Spence07" class="csl-entry" role="listitem">
Spence, R. (2007). <em>Information visualization: Design for interaction</em>. Prentice Hall.
</div>
<div id="ref-colorspace2009b" class="csl-entry" role="listitem">
Stauffer, R., Mayr, G. J., Dabernig, M., &amp; Zeileis, A. (2009). Somewhere over the rainbow: How to make effective use of colors in meteorological visualizations. <em>Bulletin of the American Meteorological Society</em>, <em>96</em>(2), 203–216. <a href="https://doi.org/10.1175/BAMS-D-13-00155.1">https://doi.org/10.1175/BAMS-D-13-00155.1</a>
</div>
<div id="ref-SRLLDCC00" class="csl-entry" role="listitem">
Sutherland, P., Rossini, A., Lumley, T., Lewin-Koh, N., Dickerson, J., Cox, Z., &amp; Cook, D. (2000a). Orca: <span>A</span> <span>V</span>isualization <span>T</span>oolkit for <span>H</span>igh-<span>D</span>imensional <span>D</span>ata. <em>Journal of Computational and Graphical Statistics</em>, <em>9</em>(3), 509–529.
</div>
<div id="ref-orca" class="csl-entry" role="listitem">
Sutherland, P., Rossini, A., Lumley, T., Lewin-Koh, N., Dickerson, J., Cox, Z., &amp; Cook, D. (2000b). Orca: A visualization toolkit for high-dimensional data. <em>Journal of Computational and Graphical Statistics</em>, <em>9</em>(3), 509–529. <a href="https://doi.org/10.1080/10618600.2000.10474896">https://doi.org/10.1080/10618600.2000.10474896</a>
</div>
<div id="ref-SBTL04" class="csl-entry" role="listitem">
Swayne, D. F., Buja, A., &amp; Temple Lang, D. (2004). Exploratory visual analysis of graphs in <span>GG</span>obi. In J. Antoch (Ed.), <em>CompStat: Proceedings in computational statistics, 16th symposium</em>. Physica-Verlag.
</div>
<div id="ref-SCB91" class="csl-entry" role="listitem">
Swayne, D. F., Cook, D., &amp; Buja, A. (1992). <span>XG</span>obi: <span>I</span>nteractive <span>D</span>ynamic <span>G</span>raphics in the <span>X</span> <span>W</span>indow <span>S</span>ystem with a <span>L</span>ink to <span>S</span>. <em>American Statistical Association 1991 Proceedings of the Section on Statistical Graphics</em>, 1–8.
</div>
<div id="ref-xgobi" class="csl-entry" role="listitem">
Swayne, D. F., Cook, D., &amp; Buja, A. (1998). XGobi: Interactive dynamic data visualization in the x window system. <em>Journal of Computational and Graphical Statistics</em>, <em>7</em>(1), 113–130. <a href="https://doi.org/10.1080/10618600.1998.10474764">https://doi.org/10.1080/10618600.1998.10474764</a>
</div>
<div id="ref-SK98" class="csl-entry" role="listitem">
Swayne, D. F., &amp; Klinke, S. (1998). Editorial commentary. <em>Computational Statistics: Special Issue on The Use of Interactive Graphics</em>, <em>14</em>(1).
</div>
<div id="ref-STLBC03" class="csl-entry" role="listitem">
Swayne, D. F., Temple Lang, D., Buja, A., &amp; Cook, D. (2003). <span>GG</span>obi: <span>E</span>volving from <span>XG</span>obi into an <span>E</span>xtensible <span>F</span>ramework for <span>I</span>nteractive <span>D</span>ata <span>V</span>isualization. <em>Computational Statistics &amp; Data Analysis</em>, <em>43</em>, 423–444.
</div>
<div id="ref-SB98" class="csl-entry" role="listitem">
Swayne, D., &amp; Buja, A. (1998). <span>M</span>issing <span>D</span>ata in <span>I</span>nteractive <span>H</span>igh-<span>D</span>imensional <span>D</span>ata <span>V</span>isualization. <em>Computational Statistics</em>, <em>13</em>(1), 15–26.
</div>
<div id="ref-Symanzik2002NewAO" class="csl-entry" role="listitem">
Symanzik, J. (2002). New applications of the image grand tour. <em>Computing Science and Statistics</em>, <em>34</em>, 500--512. <a href="https://math.usu.edu/symanzik/papers/2002_interface.pdf">https://math.usu.edu/symanzik/papers/2002_interface.pdf</a>
</div>
<div id="ref-Symanzik04" class="csl-entry" role="listitem">
Symanzik, J. (2004). Interactive and <span>D</span>ynamic <span>G</span>raphics. In J. E. Gentle, W. Härdle, &amp; Y. Mori (Eds.), <em>Handbook of computational statistics: Concepts and methods</em> (pp. 293–336). Springer.
</div>
<div id="ref-TaGa02" class="csl-entry" role="listitem">
Takatsuka, M., &amp; Gahegan, M. (2002). <span>GeoVISTA Studio</span>: A <span>C</span>odeless <span>V</span>isual <span>P</span>rogramming <span>E</span>nvironment for <span>G</span>eoscientific <span>D</span>ata <span>A</span>nalysis and <span>V</span>isualization. <em>The Journal of Computers and Geosciences</em>, <em>28</em>(10), 1131–1144.
</div>
<div id="ref-ggfortify2016" class="csl-entry" role="listitem">
Tang, Y., Horikoshi, M., &amp; Li, W. (2016). Ggfortify: Unified interface to visualize statistical result of popular r packages. <em>The R Journal</em>, <em>8</em>(2), 474–485. <a href="https://doi.org/10.32614/RJ-2016-060">https://doi.org/10.32614/RJ-2016-060</a>
</div>
<div id="ref-TF95" class="csl-entry" role="listitem">
Tarpey, T., Li, L., &amp; Flury, B. (1995). Principal points and self–consistent points of elliptical distributions. <em>The Annals of Statistics</em>, <em>23</em>, 103–112.
</div>
<div id="ref-LSWL06:rggobi" class="csl-entry" role="listitem">
Temple Lang, D., Swayne, D., Wickham, H., &amp; Lawrence, M. (2006). <em><span><code>rggobi</code></span>: <span>A</span>n <span>I</span>nterface between <span>R</span> and <span>GG</span>obi</em>.<span><code> http://www.R-project.org</code></span>.
</div>
<div id="ref-R-rpart" class="csl-entry" role="listitem">
Therneau, T., &amp; Atkinson, B. (2023). <em>Rpart: Recursive partitioning and regression trees</em>. <a href="https://CRAN.R-project.org/package=rpart">https://CRAN.R-project.org/package=rpart</a>
</div>
<div id="ref-Theus02" class="csl-entry" role="listitem">
Theus, M. (2002). Interactive <span>D</span>ata <span>V</span>isualization <span>U</span>sing <span>M</span>ondrian. <em>Journal of Statistical Software</em>, <em>7</em>(11), http://www.jstatsoft.org.
</div>
<div id="ref-THW98" class="csl-entry" role="listitem">
Theus, M., Hofmann, H., &amp; Wilhelm, A. F. X. (1998). Selection <span>S</span>equences – <span>I</span>nteractive <span>A</span>nalysis of <span>M</span>assive <span>D</span>ata <span>S</span>ets. <em>Computing Science and Statistics</em>, <em>29</em>(1), 439–444.
</div>
<div id="ref-GThom93" class="csl-entry" role="listitem">
Thompson, G. L. (1993). Generalized <span>P</span>ermutation <span>P</span>olytopes and <span>E</span>xploratory <span>G</span>raphical <span>M</span>ethods for <span>R</span>anked <span>D</span>ata. <em>The Annals of Statistics</em>, <em>21</em>, 1401–1430.
</div>
<div id="ref-Ti91" class="csl-entry" role="listitem">
Tierney, L. (1991). <em><span>L</span>isp<span>S</span>tat: <span>A</span>n <span>O</span>bject-<span>O</span>rientated <span>E</span>nvironment for <span>S</span>tatistical <span>C</span>omputing and <span>D</span>ynamic <span>G</span>raphics</em>. John Wiley &amp; Sons.
</div>
<div id="ref-naniar" class="csl-entry" role="listitem">
Tierney, N., &amp; Cook, D. (2023a). Expanding tidy data principles to facilitate missing data exploration, visualization and assessment of imputations. <em>Journal of Statistical Software</em>, <em>105</em>(7), 1–31. <a href="https://doi.org/10.18637/jss.v105.i07">https://doi.org/10.18637/jss.v105.i07</a>
</div>
<div id="ref-naniar2023" class="csl-entry" role="listitem">
Tierney, N., &amp; Cook, D. (2023b). Expanding tidy data principles to facilitate missing data exploration, visualization and assessment of imputations. <em>Journal of Statistical Software</em>, <em>105</em>(7), 1–31. <a href="https://doi.org/10.18637/jss.v105.i07">https://doi.org/10.18637/jss.v105.i07</a>
</div>
<div id="ref-R-naniar" class="csl-entry" role="listitem">
Tierney, N., Cook, D., McBain, M., &amp; Fay, C. (2023). <em>Naniar: Data structures, summaries, and visualisations for missing data</em>. <a href="https://github.com/njtierney/naniar">https://github.com/njtierney/naniar</a>
</div>
<div id="ref-To52" class="csl-entry" role="listitem">
Torgerson, W. S. (1952). Multidimensional <span>S</span>caling. 1. <span>T</span>heory and <span>M</span>ethod. <em>Psychometrika</em>, <em>17</em>, 401–419.
</div>
<div id="ref-Tufte83" class="csl-entry" role="listitem">
Tufte, E. (1983). <em>The visual display of quantitative information</em>. Graphics Press.
</div>
<div id="ref-Tufte90" class="csl-entry" role="listitem">
Tufte, E. (1990). <em>Envisioning information</em>. Graphics Press.
</div>
<div id="ref-Tu65" class="csl-entry" role="listitem">
Tukey, J. W. (1965). The <span>T</span>echnical <span>T</span>ools of <span>S</span>tatistics. <em>The American Statistician</em>, <em>19</em>, 23–28.
</div>
<div id="ref-UHS96" class="csl-entry" role="listitem">
Unwin, A. R., Hawkins, G., Hofmann, H., &amp; Siegl, B. (1996). <span>I</span>nteractive <span>G</span>raphics for <span>D</span>ata <span>S</span>ets with <span>M</span>issing <span>V</span>alues - <span>MANET</span>. <em>Journal of Computational and Graphical Statistics</em>, <em>5</em>(2), 113–122.
</div>
<div id="ref-UHW02" class="csl-entry" role="listitem">
Unwin, A., Hofmann, H., &amp; Wilhelm, A. (2002). Direct <span>M</span>anipulation <span>G</span>raphics for <span>D</span>ata <span>M</span>ining. <em>Journal of Image and Graphics</em>, <em>2</em>(1), 49–65.
</div>
<div id="ref-Unwin06" class="csl-entry" role="listitem">
Unwin, A., Theus, M., &amp; Hofmann, H. (2006). <em>Graphics of <span>L</span>arge <span>D</span>atasets: <span>V</span>isualizing a <span>M</span>illion</em>. Springer.
</div>
<div id="ref-UVW03" class="csl-entry" role="listitem">
Unwin, A., Volinsky, C., &amp; Winkler, S. (2003). Parallel <span>C</span>oordinates for <span>E</span>xploratory <span>M</span>odelling <span>A</span>nalysis. <em>Comput. Stat. Data Anal.</em>, <em>43</em>(4), 553–564. https://doi.org/<a href="{\tt http://dx.doi.org/10.1016/S0167-9473(02)00292-X}">{\tt http://dx.doi.org/10.1016/S0167-9473(02)00292-X}</a>
</div>
<div id="ref-iPlots03" class="csl-entry" role="listitem">
Urbanek, S., &amp; Theus, M. (2003). <span class="nocase">iPlots</span>: <span>H</span>igh <span>I</span>nteraction <span>G</span>raphics for <span>R</span>. In K. Hornik, F. Leisch, &amp; A. Zeileis (Eds.), <em>Proceedings of the 3rd international workshop on distributed statistical computing (DSC 2003)</em>.
</div>
<div id="ref-R-htmlwidgets" class="csl-entry" role="listitem">
Vaidyanathan, R., Xie, Y., Allaire, J., Cheng, J., Sievert, C., &amp; Russell, K. (2023). <em>Htmlwidgets: HTML widgets for r</em>. <a href="https://github.com/ramnathv/htmlwidgets">https://github.com/ramnathv/htmlwidgets</a>
</div>
<div id="ref-Rtsne2014" class="csl-entry" role="listitem">
van der Maaten, L. J. P. (2014). Accelerating t-SNE using tree-based algorithms. <em>Journal of Machine Learning Research</em>, <em>15</em>, 3221–3245.
</div>
<div id="ref-Rtsne2008" class="csl-entry" role="listitem">
van der Maaten, L. J. P., &amp; Hinton, G. E. (2008). Visualizing high-dimensional data using t-SNE. <em>Journal of Machine Learning Research</em>, <em>9</em>, 2579–2605.
</div>
<div id="ref-Va99" class="csl-entry" role="listitem">
Vapnik, V. N. (1999). <em>The <span>N</span>ature of <span>S</span>tatistical <span>L</span>earning <span>T</span>heory</em>. Springer.
</div>
<div id="ref-VeVe85" class="csl-entry" role="listitem">
Velleman, P. F., &amp; Velleman, A. Y. (1985). <em>Data desk handbook</em>. Data Description, Inc.
</div>
<div id="ref-VR02" class="csl-entry" role="listitem">
Venables, W. N., &amp; Ripley, B. (2002a). <em>Modern <span>A</span>pplied <span>S</span>tatistics with <span>S</span></em>. Springer-Verlag.
</div>
<div id="ref-MASS2002" class="csl-entry" role="listitem">
Venables, W. N., &amp; Ripley, B. D. (2002b). <em>Modern applied statistics with s</em> (Fourth). Springer. <a href="https://www.stats.ox.ac.uk/pub/MASS4/">https://www.stats.ox.ac.uk/pub/MASS4/</a>
</div>
<div id="ref-nnet2002" class="csl-entry" role="listitem">
Venables, W. N., &amp; Ripley, B. D. (2002c). <em>Modern applied statistics with s</em> (Fourth). Springer. <a href="https://www.stats.ox.ac.uk/pub/MASS4/">https://www.stats.ox.ac.uk/pub/MASS4/</a>
</div>
<div id="ref-Wa00" class="csl-entry" role="listitem">
Wainer, H. (2000). <em>Visual <span>R</span>evelations (2nd ed)</em>. LEA, Inc.
</div>
<div id="ref-WS05a" class="csl-entry" role="listitem">
Wainer, H., &amp; Spence, I. (eds). (2005a). <em>The <span>C</span>ommercial and <span>P</span>olitical <span>A</span>tlas, <span>R</span>epresenting, by means of <span>S</span>tained <span>C</span>opper-<span>P</span>late <span>C</span>harts, <span>T</span>he <span>P</span>rogress of the <span>C</span>ommerce, <span>R</span>evenues, <span>E</span>xpenditure, and <span>D</span>ebts of <span>E</span>ngland, during the whole of the <span>E</span>ighteenth <span>C</span>entury, by <span>W</span>illiam <span>P</span>layfair</em>. Cambridge University Press.
</div>
<div id="ref-WS05b" class="csl-entry" role="listitem">
Wainer, H., &amp; Spence, I. (eds). (2005b). <em>The <span>S</span>tatistical <span>B</span>reviary; <span>S</span>hewing on a <span>P</span>rinciple entirely new, the resources of every state and kingdom in <span>E</span>urope; illustrated with <span>S</span>tained <span>C</span>opper-<span>P</span>late <span>C</span>harts, representing the physical powers of each distinct nation with ease and perspicuity by <span>W</span>illiam <span>P</span>layfair</em>. Cambridge University Press.
</div>
<div id="ref-Wa78" class="csl-entry" role="listitem">
Wang, P. C. C. (Ed.). (1978). <em><span>G</span>raphical <span>R</span>epresentation of <span>M</span>ultivariate <span>D</span>ata</em>. Academic Press.
</div>
<div id="ref-We90" class="csl-entry" role="listitem">
Wegman, E. (1990). <span>H</span>yperdimensional <span>D</span>ata <span>A</span>nalysis <span>U</span>sing <span>P</span>arallel <span>C</span>oordinates. <em>Journal of American Statistics Association</em>, <em>85</em>, 664–675.
</div>
<div id="ref-We91" class="csl-entry" role="listitem">
Wegman, E. J. (1991). <em>The <span>G</span>rand <span>T</span>our in <span class="math inline">\(k\)</span>-<span>D</span>imensions</em> (Technical Report No. 68). Center for Computational Statistics, George Mason University.
</div>
<div id="ref-WC93" class="csl-entry" role="listitem">
Wegman, E. J., &amp; Carr, D. B. (1993). <em><span>S</span>tatistical <span>G</span>raphics and <span>V</span>isualization</em> (C. R. Rao, Ed.; pp. 857–958). Elsevier Science Publishers.
</div>
<div id="ref-WPS98" class="csl-entry" role="listitem">
Wegman, E. J., Poston, W. L., &amp; Solka, J. L. (1998). Image <span>G</span>rand <span>T</span>our. <em>Automatic Target Recognition VIII - Proceedings of SPIE, 3371</em>, 286–294.
</div>
<div id="ref-kohonen2007" class="csl-entry" role="listitem">
Wehrens, R., &amp; Buydens, L. M. C. (2007). Self- and super-organizing maps in <span>R</span>: The <span class="nocase">kohonen</span> package. <em>Journal of Statistical Software</em>, <em>21</em>(5), 1–19. <a href="https://doi.org/10.18637/jss.v021.i05">https://doi.org/10.18637/jss.v021.i05</a>
</div>
<div id="ref-kohonen2018" class="csl-entry" role="listitem">
Wehrens, R., &amp; Kruisselbrink, J. (2018). Flexible self-organizing maps in <span class="nocase">kohonen</span> 3.0. <em>Journal of Statistical Software</em>, <em>87</em>(7), 1–18. <a href="https://doi.org/10.18637/jss.v087.i07">https://doi.org/10.18637/jss.v087.i07</a>
</div>
<div id="ref-R-kohonen" class="csl-entry" role="listitem">
Wehrens, R., &amp; Kruisselbrink, J. (2023). <em>Kohonen: Supervised and unsupervised self-organising maps</em>. <a href="https://CRAN.R-project.org/package=kohonen">https://CRAN.R-project.org/package=kohonen</a>
</div>
<div id="ref-ggplot22016" class="csl-entry" role="listitem">
Wickham, H. (2016). <em>ggplot2: Elegant graphics for data analysis</em>. Springer-Verlag New York. <a href="https://ggplot2.tidyverse.org">https://ggplot2.tidyverse.org</a>
</div>
<div id="ref-R-classifly" class="csl-entry" role="listitem">
Wickham, H. (2022). <em>Classifly: Explore classification models in high dimensions</em>. <a href="http://had.co.nz/classifly">http://had.co.nz/classifly</a>
</div>
<div id="ref-R-ggplot2" class="csl-entry" role="listitem">
Wickham, H., Chang, W., Henry, L., Pedersen, T. L., Takahashi, K., Wilke, C., Woo, K., Yutani, H., &amp; Dunnington, D. (2023). <em>ggplot2: Create elegant data visualisations using the grammar of graphics</em>. <a href="https://CRAN.R-project.org/package=ggplot2">https://CRAN.R-project.org/package=ggplot2</a>
</div>
<div id="ref-R-tourr" class="csl-entry" role="listitem">
Wickham, H., &amp; Cook, D. (2024). <em>Tourr: Tour methods for multivariate data visualisation</em>. <a href="https://github.com/ggobi/tourr">https://github.com/ggobi/tourr</a>
</div>
<div id="ref-wickham2015" class="csl-entry" role="listitem">
Wickham, H., Cook, D., &amp; Hofmann, H. (2015). Visualizing statistical models: Removing the blindfold. <em>Statistical Analysis and Data Mining: The ASA Data Science Journal</em>, <em>8</em>(4), 203–225. <a href="https://doi.org/10.1002/sam.11271">https://doi.org/10.1002/sam.11271</a>
</div>
<div id="ref-tourr" class="csl-entry" role="listitem">
Wickham, H., Cook, D., Hofmann, H., &amp; Buja, A. (2011a). Tourr: <span>An</span> <span>R</span> <span>Package</span> for <span>Exploring</span> <span>Multivariate</span> <span>Data</span> with <span>Projections</span>. <em>Journal of Statistical Software</em>, <em>40</em>(2). <a href="https://doi.org/10.18637/jss.v040.i02">https://doi.org/10.18637/jss.v040.i02</a>
</div>
<div id="ref-tourr2011" class="csl-entry" role="listitem">
Wickham, H., Cook, D., Hofmann, H., &amp; Buja, A. (2011b). <span class="nocase">tourr</span>: An <span>R</span> package for exploring multivariate data with projections. <em>Journal of Statistical Software</em>, <em>40</em>(2), 1–18. <a href="https://doi.org/10.18637/jss.v040.i02">https://doi.org/10.18637/jss.v040.i02</a>
</div>
<div id="ref-R-dplyr" class="csl-entry" role="listitem">
Wickham, H., François, R., Henry, L., Müller, K., &amp; Vaughan, D. (2023). <em>Dplyr: A grammar of data manipulation</em>. <a href="https://CRAN.R-project.org/package=dplyr">https://CRAN.R-project.org/package=dplyr</a>
</div>
<div id="ref-R-readr" class="csl-entry" role="listitem">
Wickham, H., Hester, J., &amp; Bryan, J. (2023). <em>Readr: Read rectangular text data</em>. <a href="https://CRAN.R-project.org/package=readr">https://CRAN.R-project.org/package=readr</a>
</div>
<div id="ref-WWS99" class="csl-entry" role="listitem">
Wilhelm, A. F. X., Wegman, E. J., &amp; Symanzik, J. (1999). Visual <span>C</span>lustering and <span>C</span>lassification: <span>T</span>he <span>O</span>ronsay <span>P</span>article <span>S</span>ize <span>D</span>ata <span>S</span>et <span>R</span>evisited. <em>Computational Statistics: Special Issue on Interactive Graphical Data Analysis</em>, <em>14</em>(1), 109–146.
</div>
<div id="ref-Wilkinson05" class="csl-entry" role="listitem">
Wilkinson, L. (2005). <em>The grammar of graphics</em>. Springer.
</div>
<div id="ref-Wills99" class="csl-entry" role="listitem">
Wills, G. (1999). NicheWorks – <span>I</span>nteractive <span>V</span>isualization of <span>V</span>ery <span>L</span>arge <span>G</span>raphs. <em>Journal of Computational and Graphical Statistics</em>, <em>8</em>(2), 190–212.
</div>
<div id="ref-cranvas" class="csl-entry" role="listitem">
Xie, Y., Hofmann, H., &amp; Cheng, X. (2014). <span class="nocase">Reactive Programming for Interactive Graphics</span>. <em>Statistical Science</em>, <em>29</em>(2), 201–213. <a href="https://doi.org/10.1214/14-STS477">https://doi.org/10.1214/14-STS477</a>
</div>
<div id="ref-Young06" class="csl-entry" role="listitem">
Young, F. W., Valero-Mora, P. M., &amp; Friendly, M. (2006). <em>Visual <span>S</span>tatistics: <span>S</span>eeing <span>D</span>ata with <span>D</span>ynamic <span>I</span>nteractive <span>G</span>raphics</em>. John Wiley &amp; Sons.
</div>
<div id="ref-colorspace2020a" class="csl-entry" role="listitem">
Zeileis, A., Fisher, J. C., Hornik, K., Ihaka, R., McWhite, C. D., Murrell, P., Stauffer, R., &amp; Wilke, C. O. (2020). <span class="nocase">colorspace</span>: A toolbox for manipulating and assessing colors and palettes. <em>Journal of Statistical Software</em>, <em>96</em>(1), 1–49. <a href="https://doi.org/10.18637/jss.v096.i01">https://doi.org/10.18637/jss.v096.i01</a>
</div>
<div id="ref-colorspace2009a" class="csl-entry" role="listitem">
Zeileis, A., Hornik, K., &amp; Murrell, P. (2009). Escaping <span>RGB</span>land: Selecting colors for statistical graphics. <em>Computational Statistics &amp; Data Analysis</em>, <em>53</em>(9), 3259–3270. <a href="https://doi.org/10.1016/j.csda.2008.11.033">https://doi.org/10.1016/j.csda.2008.11.033</a>
</div>
<div id="ref-zhang2023" class="csl-entry" role="listitem">
Zhang, C., Ye, J., &amp; Wang, X. (2023). A computational perspective on projection pursuit in high dimensions: Feasible or infeasible feature extraction. <em>International Statistical Review</em>, <em>91</em>(1), 140–161. <a href="https://doi.org/10.1111/insr.12517">https://doi.org/10.1111/insr.12517</a>
</div>
<div id="ref-ferrn2021" class="csl-entry" role="listitem">
Zhang, H. S., Cook, D., Laa, U., Langrené, N., &amp; Menéndez, P. (2021). Visual diagnostics for constrained optimisation with application to guided tours. <em>The R Journal</em>, <em>13</em>(2), 624–641. <a href="https://doi.org/10.32614/RJ-2021-105">https://doi.org/10.32614/RJ-2021-105</a>
</div>
<div id="ref-R-ferrn" class="csl-entry" role="listitem">
Zhang, H. S., Cook, D., Laa, U., Langrené, N., &amp; Menéndez, P. (2022). <em>Ferrn: Facilitate exploration of touRR optimisatioN</em>. <a href="https://github.com/huizezhang-sherry/ferrn/">https://github.com/huizezhang-sherry/ferrn/</a>
</div>
<div id="ref-R-kableExtra" class="csl-entry" role="listitem">
Zhu, H. (2021). <em>kableExtra: Construct complex table with kable and pipe syntax</em>. <a href="https://CRAN.R-project.org/package=kableExtra">https://CRAN.R-project.org/package=kableExtra</a>
</div>
</div>
</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./16-svm.html" class="pagination-link  aria-label=" vector="" machines="">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Support vector machines</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./18-summary-class.html" class="pagination-link" aria-label="<span class='chapter-number'>18</span>&nbsp; <span class='chapter-title'>Exploring misclassifications</span>">
        <span class="nav-page-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Exploring misclassifications</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
<p>Interactively exploring high-dimensional data and models in R was written by Dianne Cook and Ursula Laa.</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/dicook/mulgar_book/edit/master/17-nn.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/dicook/mulgar_book/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>


</body></html>